{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebookd is to format the domain annotation output for plasmid regions detected using KMER spectrum analysis. HMMER and the pfam database each of the plasmid proteomes was annotated. We used the NCBI annotation of the genome when possible. Otherwise, PATRIC annotations or the getORF tool ( from the EMBOSS suite ) were used.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from matplotlib import pyplot as plt\n",
    "import uniprot\n",
    "import numpy as np\n",
    "import itertools\n",
    "import glob\n",
    "import multiprocessing as mp\n",
    "import subprocess\n",
    "import shlex\n",
    "import Bio.SeqFeature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "hnatans = pd.read_table( './926690.3.PATRIC.features.tab')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      genome_id                  genome_name     accession annotation  \\\n",
      "2707   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "2708   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "2709   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "2710   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "2711   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "...         ...                          ...           ...        ...   \n",
      "3405   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "3406   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "3407   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "3408   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "3409   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "\n",
      "     feature_type              patric_id  refseq_locus_tag    start      end  \\\n",
      "2707          CDS  fig|926690.3.peg.2651               NaN  1496172  1497365   \n",
      "2708          CDS  fig|926690.3.peg.2652               NaN  1498710  1498907   \n",
      "2709          CDS  fig|926690.3.peg.2653               NaN  1500237  1501934   \n",
      "2710          CDS  fig|926690.3.peg.2654               NaN  1503001  1503423   \n",
      "2711          CDS  fig|926690.3.peg.2655               NaN  1503499  1505130   \n",
      "...           ...                    ...               ...      ...      ...   \n",
      "3405          CDS  fig|926690.3.peg.3332               NaN  2165873  2166127   \n",
      "3406          CDS  fig|926690.3.peg.3333               NaN  2166128  2167183   \n",
      "3407          CDS  fig|926690.3.peg.3334               NaN  2167248  2167445   \n",
      "3408          CDS  fig|926690.3.peg.3335               NaN  2167649  2167960   \n",
      "3409         rRNA    fig|926690.3.rna.50               NaN  2168757  2169903   \n",
      "\n",
      "     strand  na_length  gene  \\\n",
      "2707      -       1194   NaN   \n",
      "2708      -        198   NaN   \n",
      "2709      -       1698   NaN   \n",
      "2710      +        423   NaN   \n",
      "2711      -       1632   NaN   \n",
      "...     ...        ...   ...   \n",
      "3405      +        255   NaN   \n",
      "3406      -       1056   NaN   \n",
      "3407      -        198   NaN   \n",
      "3408      +        312   NaN   \n",
      "3409      +       1147   NaN   \n",
      "\n",
      "                                                product  plfam_id  \\\n",
      "2707                               hypothetical protein       NaN   \n",
      "2708                               hypothetical protein       NaN   \n",
      "2709                               hypothetical protein       NaN   \n",
      "2710     hypothetical protein; probable calcium binding       NaN   \n",
      "2711                               hypothetical protein       NaN   \n",
      "...                                                 ...       ...   \n",
      "3405                               hypothetical protein       NaN   \n",
      "3406  Dihydroorotate dehydrogenase (quinone) (EC 1.3...       NaN   \n",
      "3407                               hypothetical protein       NaN   \n",
      "3408                            Chromosomal protein MC1       NaN   \n",
      "3409      Small Subunit Ribosomal RNA; ssuRNA; SSU rRNA       NaN   \n",
      "\n",
      "          pgfam_id  \n",
      "2707           NaN  \n",
      "2708           NaN  \n",
      "2709  PGF_06091862  \n",
      "2710           NaN  \n",
      "2711           NaN  \n",
      "...            ...  \n",
      "3405           NaN  \n",
      "3406  PGF_01033770  \n",
      "3407  PGF_00303083  \n",
      "3408  PGF_00417877  \n",
      "3409           NaN  \n",
      "\n",
      "[703 rows x 15 columns]\n",
      "703\n"
     ]
    }
   ],
   "source": [
    "sub = hnatans[ hnatans.accession == 'ATYM01000002']\n",
    "sub = sub[sub.start > 1493414]\n",
    "sub = sub[sub.end > 1495372]\n",
    "print(sub)\n",
    "print(len(sub))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "684\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "684"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_seq_file = \"./926690.3.PATRIC.faa\"\n",
    "parsed = SeqIO.parse( in_seq_file , 'fasta')\n",
    "keep = []\n",
    "for i,seq in enumerate(parsed):\n",
    "    if seq.id in sub.patric_id.unique():\n",
    "        seqlist.add(seq.id)\n",
    "        keep.append(seq)\n",
    "outfile = './926690.3.PATRIC.proteo.fasta'\n",
    "SeqIO.write( keep,outfile, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub['protein_id'] = sub.patric_id\n",
    "sub.index = sub.patric_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: JNCS01000001.1\n",
      "Name: JNCS01000001\n",
      "Description: Natrinema altunense strain AJ2 N_altunense_AJ2_contig_1, whole genome shotgun sequence\n",
      "Database cross-references: BioProject:PRJNA248700, BioSample:SAMN02800870\n",
      "Number of features: 1\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=BCT\n",
      "/date=21-JUL-2014\n",
      "/accessions=['JNCS01000001', 'REGION:', '496500..593500']\n",
      "/sequence_version=1\n",
      "/keywords=['WGS']\n",
      "/source=Natrinema altunense\n",
      "/organism=Natrinema altunense\n",
      "/taxonomy=['Archaea', 'Euryarchaeota', 'Stenosarchaea group', 'Halobacteria', 'Natrialbales', 'Natrialbaceae', 'Natrinema']\n",
      "/references=[Reference(title='Natrinema altunense sp. nov., an extremely halophilic archaeon isolated from a salt lake in Altun Mountain in Xinjiang, China', ...), Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=OrderedDict([('Genome-Assembly-Data', OrderedDict([('Assembly Method', 'Newbler v. Feb-2013'), ('Genome Representation', 'Full'), ('Expected Final Version', 'No'), ('Genome Coverage', '120.0x'), ('Sequencing Technology', '454')]))])\n",
      "Seq('GTACTCGTGCCCTCTATCCACGCGGGTGTCGTCGTGTTATTTGTTTCCGGACAC...GGT', IUPACAmbiguousDNA())\n",
      "                          protein_id  start   stop  \\\n",
      "JNCS01000001.1_1    JNCS01000001.1_1    708   1748   \n",
      "JNCS01000001.1_2    JNCS01000001.1_2   3911   4552   \n",
      "JNCS01000001.1_3    JNCS01000001.1_3   4354   5184   \n",
      "JNCS01000001.1_4    JNCS01000001.1_4   8492   9190   \n",
      "JNCS01000001.1_5    JNCS01000001.1_5  12484  13356   \n",
      "...                              ...    ...    ...   \n",
      "JNCS01000001.1_86  JNCS01000001.1_86   8063   7323   \n",
      "JNCS01000001.1_87  JNCS01000001.1_87   7281   5257   \n",
      "JNCS01000001.1_88  JNCS01000001.1_88   5247   4204   \n",
      "JNCS01000001.1_89  JNCS01000001.1_89   3508   2351   \n",
      "JNCS01000001.1_90  JNCS01000001.1_90    688     80   \n",
      "\n",
      "                                                       genome xrefs  \n",
      "JNCS01000001.1_1   [BioProject:PRJNA248700, BioSample:SAMN02800870]  \n",
      "JNCS01000001.1_2   [BioProject:PRJNA248700, BioSample:SAMN02800870]  \n",
      "JNCS01000001.1_3   [BioProject:PRJNA248700, BioSample:SAMN02800870]  \n",
      "JNCS01000001.1_4   [BioProject:PRJNA248700, BioSample:SAMN02800870]  \n",
      "JNCS01000001.1_5   [BioProject:PRJNA248700, BioSample:SAMN02800870]  \n",
      "...                                                             ...  \n",
      "JNCS01000001.1_86  [BioProject:PRJNA248700, BioSample:SAMN02800870]  \n",
      "JNCS01000001.1_87  [BioProject:PRJNA248700, BioSample:SAMN02800870]  \n",
      "JNCS01000001.1_88  [BioProject:PRJNA248700, BioSample:SAMN02800870]  \n",
      "JNCS01000001.1_89  [BioProject:PRJNA248700, BioSample:SAMN02800870]  \n",
      "JNCS01000001.1_90  [BioProject:PRJNA248700, BioSample:SAMN02800870]  \n",
      "\n",
      "[90 rows x 4 columns]\n",
      "ID: AOIK01000043.1\n",
      "Name: AOIK01000043\n",
      "Description: Natrinema altunense JCM 12890 contig_43, whole genome shotgun sequence\n",
      "Database cross-references: BioProject:PRJNA174934, BioSample:SAMN02471672\n",
      "Number of features: 209\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=BCT\n",
      "/date=04-FEB-2013\n",
      "/accessions=['AOIK01000043', 'REGION:', '116500..213000']\n",
      "/sequence_version=1\n",
      "/keywords=['WGS']\n",
      "/source=Natrinema altunense JCM 12890\n",
      "/organism=Natrinema altunense JCM 12890\n",
      "/taxonomy=['Archaea', 'Euryarchaeota', 'Stenosarchaea group', 'Halobacteria', 'Natrialbales', 'Natrialbaceae', 'Natrinema']\n",
      "/references=[Reference(title='Phylogenetically driven sequencing of extremely halophilic archaea reveals strategies for static and dynamic osmo-response', ...), Reference(title='Direct Submission', ...)]\n",
      "/comment=Annotation was added by the NCBI Prokaryotic Genomes Automatic\n",
      "Annotation Pipeline Group. Information about the Pipeline can be\n",
      "found here:\n",
      "http://www.ncbi.nlm.nih.gov/genomes/static/Pipeline.html. Please be\n",
      "aware that the annotation is done automatically with little or no\n",
      "manual curation.\n",
      "/structured_comment=OrderedDict([('Genome-Assembly-Data', OrderedDict([('Assembly Method', 'A5 v. Nov-2011'), ('Genome Coverage', '200x'), ('Sequencing Technology', 'Illumina GAIIX; HiSeq')]))])\n",
      "Seq('CATCCGCGCCAGCAACAGTTTCGACTACAGCGCCCGCTGGCTCGAATGCACGGG...GCT', IUPACAmbiguousDNA())\n",
      "             locus_tag codon_start transl_table  \\\n",
      "ELY83594.1  C485_17617           1           11   \n",
      "ELY83595.1  C485_17622           1           11   \n",
      "ELY83596.1  C485_17627           1           11   \n",
      "ELY83597.1  C485_17632           1           11   \n",
      "ELY83598.1  C485_17637           1           11   \n",
      "...                ...         ...          ...   \n",
      "ELY83692.1  C485_18107           1           11   \n",
      "ELY83693.1  C485_18112           1           11   \n",
      "ELY83694.1  C485_18117           1           11   \n",
      "ELY83695.1  C485_18122           1           11   \n",
      "ELY83696.1  C485_18127           1           11   \n",
      "\n",
      "                                           product  protein_id  \\\n",
      "ELY83594.1                    hypothetical protein  ELY83594.1   \n",
      "ELY83595.1                    hypothetical protein  ELY83595.1   \n",
      "ELY83596.1          site-specific recombinase xerd  ELY83596.1   \n",
      "ELY83597.1                    hypothetical protein  ELY83597.1   \n",
      "ELY83598.1  transposase, IS605 OrfB family protein  ELY83598.1   \n",
      "...                                            ...         ...   \n",
      "ELY83692.1                    hypothetical protein  ELY83692.1   \n",
      "ELY83693.1                    hypothetical protein  ELY83693.1   \n",
      "ELY83694.1                    hypothetical protein  ELY83694.1   \n",
      "ELY83695.1                    hypothetical protein  ELY83695.1   \n",
      "ELY83696.1                    hypothetical protein  ELY83696.1   \n",
      "\n",
      "                                                  translation  \\\n",
      "ELY83594.1  MTELARYKYQEAEYFLNRLRQADNEDHVRYNLDAFLAAARAVTETL...   \n",
      "ELY83595.1  MTDNDEYVKLSEDRCAVKIYPENEVAEQWREEADDQDISLSRYLQN...   \n",
      "ELY83596.1  MASKKYDVLLDADEWERRVFENKHKCINDWLRSKDSTNKSSSRRTL...   \n",
      "ELY83597.1  MVDWFKWKTIGIGGVGIGLVLTGIQLGLLVILQFFDWSTIPALSDG...   \n",
      "ELY83598.1  MHNRLPELKDEWQELKQVHSKVLQRVVTQLYDNLKSLSEQKKNGYK...   \n",
      "...                                                       ...   \n",
      "ELY83692.1  MHHFDAEGVRIDCERNIAFTVPIQLVKNLLDLLFVLLPGRVATRVI...   \n",
      "ELY83693.1  MIESEENSSNTGYIILDENALISWFDRWIATRPTIGNHSDLFKIVF...   \n",
      "ELY83694.1  MVDDDLYDELEPSENGDRYSWSDAVEDLEEPRRKKLERKRRKLSRK...   \n",
      "ELY83695.1  MKRKYLKQDVDKEDLTEAQEEIIEIAIEKPELTHQEIADQTKWSRS...   \n",
      "ELY83696.1  MSTQPLQPPTDAEPTRTTTRTRPTPSSDVADPRPAADRAQRMFSHG...   \n",
      "\n",
      "                                                genome xrefs  \\\n",
      "ELY83594.1  [BioProject:PRJNA174934, BioSample:SAMN02471672]   \n",
      "ELY83595.1  [BioProject:PRJNA174934, BioSample:SAMN02471672]   \n",
      "ELY83596.1  [BioProject:PRJNA174934, BioSample:SAMN02471672]   \n",
      "ELY83597.1  [BioProject:PRJNA174934, BioSample:SAMN02471672]   \n",
      "ELY83598.1  [BioProject:PRJNA174934, BioSample:SAMN02471672]   \n",
      "...                                                      ...   \n",
      "ELY83692.1  [BioProject:PRJNA174934, BioSample:SAMN02471672]   \n",
      "ELY83693.1  [BioProject:PRJNA174934, BioSample:SAMN02471672]   \n",
      "ELY83694.1  [BioProject:PRJNA174934, BioSample:SAMN02471672]   \n",
      "ELY83695.1  [BioProject:PRJNA174934, BioSample:SAMN02471672]   \n",
      "ELY83696.1  [BioProject:PRJNA174934, BioSample:SAMN02471672]   \n",
      "\n",
      "                                                       note  \n",
      "ELY83594.1                                              NaN  \n",
      "ELY83595.1                                              NaN  \n",
      "ELY83596.1                                COG0582 Integrase  \n",
      "ELY83597.1                                              NaN  \n",
      "ELY83598.1  COG0675 Transposase and inactivated derivatives  \n",
      "...                                                     ...  \n",
      "ELY83692.1                                              NaN  \n",
      "ELY83693.1                                              NaN  \n",
      "ELY83694.1                                              NaN  \n",
      "ELY83695.1                                              NaN  \n",
      "ELY83696.1                                              NaN  \n",
      "\n",
      "[103 rows x 8 columns]\n",
      "ID: NZ_CP071597.1\n",
      "Name: NZ_CP071597\n",
      "Description: Halovivax sp. KZCA124 chromosome 1, complete sequence\n",
      "Database cross-references: BioProject:PRJNA224116, BioSample:SAMN18203374, Assembly:GCF_017357405.1\n",
      "Number of features: 229\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=CON\n",
      "/date=18-MAR-2021\n",
      "/accessions=['NZ_CP071597', 'REGION:', '3085000..3179500']\n",
      "/sequence_version=1\n",
      "/keywords=['RefSeq']\n",
      "/source=Halovivax sp. KZCA124\n",
      "/organism=Halovivax sp. KZCA124\n",
      "/taxonomy=['Archaea', 'Euryarchaeota', 'Stenosarchaea group', 'Halobacteria', 'Natrialbales', 'Natrialbaceae', 'Halovivax', 'unclassified Halovivax']\n",
      "/references=[Reference(title='Natronomonas halophila sp. nov. and Natronomonas salina sp. nov., two new members of the genus Natronomonas', ...), Reference(title='Direct Submission', ...)]\n",
      "/comment=REFSEQ INFORMATION: The reference sequence is identical to\n",
      "CP071597.1.\n",
      "The annotation was added by the NCBI Prokaryotic Genome Annotation\n",
      "Pipeline (PGAP). Information about PGAP can be found here:\n",
      "https://www.ncbi.nlm.nih.gov/genome/annotation_prok/\n",
      "COMPLETENESS: full length.\n",
      "/structured_comment=OrderedDict([('Genome-Assembly-Data', OrderedDict([('Assembly Method', 'HGAP v. 4'), ('Genome Representation', 'Full'), ('Expected Final Version', 'Yes'), ('Genome Coverage', '100.0x'), ('Sequencing Technology', 'PacBio')])), ('Genome-Annotation-Data', OrderedDict([('Annotation Provider', 'NCBI RefSeq'), ('Annotation Date', '03/17/2021 08:53:43'), ('Annotation Pipeline', 'NCBI Prokaryotic Genome Annotation Pipeline (PGAP)'), ('Annotation Method', 'Best-placed reference protein set; GeneMarkS-2+'), ('Annotation Software revision', '5.1'), ('Features Annotated', 'Gene; CDS; rRNA; tRNA; ncRNA; repeat_region'), ('Genes (total)', '4,707'), ('CDSs (total)', '4,652'), ('Genes (coding)', '4,562'), ('CDSs (with protein)', '4,562'), ('Genes (RNA)', '55'), ('rRNAs', '2, 2, 2 (5S, 16S, 23S)'), ('complete rRNAs', '2, 2, 2 (5S, 16S, 23S)'), ('tRNAs', '47'), ('ncRNAs', '2'), ('Pseudo Genes (total)', '90'), ('CDSs (without protein)', '90'), ('Pseudo Genes (ambiguous residues)', '0 of 90'), ('Pseudo Genes (frameshifted)', '28 of 90'), ('Pseudo Genes (incomplete)', '53 of 90'), ('Pseudo Genes (internal stop)', '22 of 90'), ('Pseudo Genes (multiple problems)', '12 of 90')]))])\n",
      "Seq('GACGAGTTTGGATTTCGAAACGCGGTGTGGCATACGTATACGAGGCGCTCGAGG...GAG', IUPACAmbiguousDNA())\n",
      "                    locus_tag  \\\n",
      "WP_207587062.1  J1N62_RS14465   \n",
      "WP_207587064.1  J1N62_RS14470   \n",
      "WP_207587066.1  J1N62_RS14475   \n",
      "WP_207587068.1  J1N62_RS14480   \n",
      "WP_207587070.1  J1N62_RS14485   \n",
      "...                       ...   \n",
      "WP_207587205.1  J1N62_RS15005   \n",
      "WP_207587207.1  J1N62_RS15010   \n",
      "WP_207587209.1  J1N62_RS15015   \n",
      "WP_207592150.1  J1N62_RS15020   \n",
      "WP_207587211.1  J1N62_RS15025   \n",
      "\n",
      "                                                        inference  \\\n",
      "WP_207587062.1  COORDINATES: similar to AA sequence:RefSeq:WP_...   \n",
      "WP_207587064.1     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "WP_207587066.1     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "WP_207587068.1     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "WP_207587070.1     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "...                                                           ...   \n",
      "WP_207587205.1     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "WP_207587207.1     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "WP_207587209.1     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "WP_207592150.1  COORDINATES: similar to AA sequence:RefSeq:WP_...   \n",
      "WP_207587211.1     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "\n",
      "                                                             note codon_start  \\\n",
      "WP_207587062.1  Derived by automated computational analysis us...           1   \n",
      "WP_207587064.1  Derived by automated computational analysis us...           1   \n",
      "WP_207587066.1  Derived by automated computational analysis us...           1   \n",
      "WP_207587068.1  Derived by automated computational analysis us...           1   \n",
      "WP_207587070.1  Derived by automated computational analysis us...           1   \n",
      "...                                                           ...         ...   \n",
      "WP_207587205.1  Derived by automated computational analysis us...           1   \n",
      "WP_207587207.1  Derived by automated computational analysis us...           1   \n",
      "WP_207587209.1  Derived by automated computational analysis us...           1   \n",
      "WP_207592150.1  Derived by automated computational analysis us...           1   \n",
      "WP_207587211.1  Derived by automated computational analysis us...           1   \n",
      "\n",
      "               transl_table                 product      protein_id  \\\n",
      "WP_207587062.1           11        HNH endonuclease  WP_207587062.1   \n",
      "WP_207587064.1           11    hypothetical protein  WP_207587064.1   \n",
      "WP_207587066.1           11    hypothetical protein  WP_207587066.1   \n",
      "WP_207587068.1           11    hypothetical protein  WP_207587068.1   \n",
      "WP_207587070.1           11    hypothetical protein  WP_207587070.1   \n",
      "...                     ...                     ...             ...   \n",
      "WP_207587205.1           11    hypothetical protein  WP_207587205.1   \n",
      "WP_207587207.1           11    hypothetical protein  WP_207587207.1   \n",
      "WP_207587209.1           11    hypothetical protein  WP_207587209.1   \n",
      "WP_207592150.1           11  DUF4385 family protein  WP_207592150.1   \n",
      "WP_207587211.1           11    hypothetical protein  WP_207587211.1   \n",
      "\n",
      "                        db_xref  \\\n",
      "WP_207587062.1  GeneID:63213810   \n",
      "WP_207587064.1  GeneID:63213811   \n",
      "WP_207587066.1  GeneID:63213812   \n",
      "WP_207587068.1  GeneID:63213813   \n",
      "WP_207587070.1  GeneID:63213814   \n",
      "...                         ...   \n",
      "WP_207587205.1  GeneID:63213918   \n",
      "WP_207587207.1  GeneID:63213919   \n",
      "WP_207587209.1  GeneID:63213920   \n",
      "WP_207592150.1  GeneID:63213921   \n",
      "WP_207587211.1  GeneID:63213922   \n",
      "\n",
      "                                                      translation  \\\n",
      "WP_207587062.1  MVVLTDSHCFICSEGNDCKLQKHHIVPRRYGGSDSEENLVTLCANC...   \n",
      "WP_207587064.1  MTVYGGESEDDLDFERLLTDAVAREREELLRQHALKEEQLEELVEQ...   \n",
      "WP_207587066.1  MSESQTDPDTRQVPPRRDRPAINAIWDGHDEWWIFSEAETPIEDGR...   \n",
      "WP_207587068.1  MSDVELARDELRAFEQSDDPRLTQLTEPVDPDPDYSQGIPVARDRR...   \n",
      "WP_207587070.1  MTDDTQIIVGVGAIIFLGVLAGTFLDILSIRAAYFGLGAGGLGLIL...   \n",
      "...                                                           ...   \n",
      "WP_207587205.1  MGTETDRLFWWIYWFLFEWDLNAVKWKAGITIATIWIADTLFFTGL...   \n",
      "WP_207587207.1  MPTFPTSDLPVPESWAEFEAVVADVIKRKWNDPYVTRNGRQGQAQN...   \n",
      "WP_207587209.1  MAIALKIASPVSTLITVCGALLGIHDWFSKRSREPEYSHQKYEQYP...   \n",
      "WP_207592150.1  MDYRAHPERYEIGRGEEGVFKVQPYKSELLTLWGYADRETADEGGR...   \n",
      "WP_207587211.1  MDRSRSSDRDDPTLVAAECEGSTELQIDVGVLVAHAPETDPGQLRR...   \n",
      "\n",
      "                                                     genome xrefs EC_number  \n",
      "WP_207587062.1  [BioProject:PRJNA224116, BioSample:SAMN1820337...       NaN  \n",
      "WP_207587064.1  [BioProject:PRJNA224116, BioSample:SAMN1820337...       NaN  \n",
      "WP_207587066.1  [BioProject:PRJNA224116, BioSample:SAMN1820337...       NaN  \n",
      "WP_207587068.1  [BioProject:PRJNA224116, BioSample:SAMN1820337...       NaN  \n",
      "WP_207587070.1  [BioProject:PRJNA224116, BioSample:SAMN1820337...       NaN  \n",
      "...                                                           ...       ...  \n",
      "WP_207587205.1  [BioProject:PRJNA224116, BioSample:SAMN1820337...       NaN  \n",
      "WP_207587207.1  [BioProject:PRJNA224116, BioSample:SAMN1820337...       NaN  \n",
      "WP_207587209.1  [BioProject:PRJNA224116, BioSample:SAMN1820337...       NaN  \n",
      "WP_207592150.1  [BioProject:PRJNA224116, BioSample:SAMN1820337...       NaN  \n",
      "WP_207587211.1  [BioProject:PRJNA224116, BioSample:SAMN1820337...       NaN  \n",
      "\n",
      "[112 rows x 11 columns]\n",
      "ID: CP058579.1\n",
      "Name: CP058579\n",
      "Description: Halobonum sp. NJ-3-1 chromosome, complete genome\n",
      "Database cross-references: BioProject:PRJNA638230, BioSample:SAMN15183787\n",
      "Number of features: 363\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=BCT\n",
      "/date=12-JUL-2020\n",
      "/accessions=['CP058579', 'REGION:', '1918000..2078500']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Halobonum sp. NJ-3-1\n",
      "/organism=Halobonum sp. NJ-3-1\n",
      "/taxonomy=['Archaea', 'Euryarchaeota', 'Stenosarchaea group', 'Halobacteria', 'Haloferacales', 'Halorubraceae', 'Candidatus Halobonum']\n",
      "/references=[Reference(title='NJ-3-1, isolated from saline soil', ...), Reference(title='Direct Submission', ...)]\n",
      "/comment=Bacteria and source DNA available from Jiangsu University.\n",
      "The annotation was added by the NCBI Prokaryotic Genome Annotation\n",
      "Pipeline (PGAP). Information about PGAP can be found here:\n",
      "https://www.ncbi.nlm.nih.gov/genome/annotation_prok/\n",
      "/structured_comment=OrderedDict([('Genome-Assembly-Data', OrderedDict([('Assembly Method', 'SOAPdenovo v. v. JULY-2018'), ('Genome Representation', 'Full'), ('Expected Final Version', 'No'), ('Genome Coverage', '100.0x'), ('Sequencing Technology', 'Illumina HiSeq')])), ('Genome-Annotation-Data', OrderedDict([('Annotation Provider', 'NCBI'), ('Annotation Date', '07/06/2020 20:05:22'), ('Annotation Pipeline', 'NCBI Prokaryotic Genome Annotation Pipeline (PGAP)'), ('Annotation Method', 'Best-placed reference protein set; GeneMarkS-2+'), ('Annotation Software revision', '4.12'), ('Features Annotated', 'Gene; CDS; rRNA; tRNA; ncRNA; repeat_region'), ('Genes (total)', '4,275'), ('CDSs (total)', '4,222'), ('Genes (coding)', '4,125'), ('CDSs (with protein)', '4,125'), ('Genes (RNA)', '53'), ('rRNAs', '1, 1, 1 (5S, 16S, 23S)'), ('complete rRNAs', '1, 1, 1 (5S, 16S, 23S)'), ('tRNAs', '48'), ('ncRNAs', '2'), ('Pseudo Genes (total)', '97'), ('CDSs (without protein)', '97'), ('Pseudo Genes (ambiguous residues)', '0 of 97'), ('Pseudo Genes (frameshifted)', '53 of 97'), ('Pseudo Genes (incomplete)', '43 of 97'), ('Pseudo Genes (internal stop)', '10 of 97'), ('Pseudo Genes (multiple problems)', '9 of 97')]))])\n",
      "Seq('TTCCCATCTCGTCGCTCGCGACCTCCGGGGGACCGCTCTACTTCGGCGTCGCGT...GAT', IUPACAmbiguousDNA())\n",
      "              locus_tag                                          inference  \\\n",
      "QLG61896.1  HUG12_09255          COORDINATES: protein motif:HMM:NF021465.1   \n",
      "QLG61897.1  HUG12_09260     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QLG61898.1  HUG12_09265  COORDINATES: similar to AA sequence:RefSeq:WP_...   \n",
      "QLG61899.1  HUG12_09270     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QLG64035.1  HUG12_09275  COORDINATES: similar to AA sequence:RefSeq:WP_...   \n",
      "...                 ...                                                ...   \n",
      "QLG62066.1  HUG12_10135     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QLG62067.1  HUG12_10140     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QLG62068.1  HUG12_10145     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QLG62069.1  HUG12_10150     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QLG62070.1  HUG12_10155          COORDINATES: protein motif:HMM:NF014044.1   \n",
      "\n",
      "                                                         note codon_start  \\\n",
      "QLG61896.1  Derived by automated computational analysis us...           3   \n",
      "QLG61897.1  Derived by automated computational analysis us...           1   \n",
      "QLG61898.1  Derived by automated computational analysis us...           1   \n",
      "QLG61899.1  Derived by automated computational analysis us...           1   \n",
      "QLG64035.1  Derived by automated computational analysis us...           1   \n",
      "...                                                       ...         ...   \n",
      "QLG62066.1  Derived by automated computational analysis us...           1   \n",
      "QLG62067.1  Derived by automated computational analysis us...           1   \n",
      "QLG62068.1  Derived by automated computational analysis us...           1   \n",
      "QLG62069.1  Derived by automated computational analysis us...           1   \n",
      "QLG62070.1  Derived by automated computational analysis us...           1   \n",
      "\n",
      "           transl_table                                  product  protein_id  \\\n",
      "QLG61896.1           11        DUF2182 domain-containing protein  QLG61896.1   \n",
      "QLG61897.1           11                     hypothetical protein  QLG61897.1   \n",
      "QLG61898.1           11                   IS5 family transposase  QLG61898.1   \n",
      "QLG61899.1           11                     hypothetical protein  QLG61899.1   \n",
      "QLG64035.1           11  class I SAM-dependent methyltransferase  QLG64035.1   \n",
      "...                 ...                                      ...         ...   \n",
      "QLG62066.1           11                     hypothetical protein  QLG62066.1   \n",
      "QLG62067.1           11                     hypothetical protein  QLG62067.1   \n",
      "QLG62068.1           11                     hypothetical protein  QLG62068.1   \n",
      "QLG62069.1           11                     hypothetical protein  QLG62069.1   \n",
      "QLG62070.1           11                      ATP-binding protein  QLG62070.1   \n",
      "\n",
      "                                                  translation  \\\n",
      "QLG61896.1  MTVLERWTRRHRDGYPTVTLLAWAALVLHDPIEQAVFGHDHEPLGL...   \n",
      "QLG61897.1  MKNIGIVPVVKNRGISPVVLAIAGFAAIVVGIHQGVLHVAPGYQGT...   \n",
      "QLG61898.1  MDALSKSQLLRFVERAMVLARRAVARFSTRYSRKRFTLRQRVVSLC...   \n",
      "QLG61899.1  MFLSKGFNKAVLLETALGLLRTFATELARLEERSDTESEPTLYPYS...   \n",
      "QLG64035.1  MHTCVDRRSSEPAIGLEIAVEDAGAEALPFADESFDVVIASLVFCT...   \n",
      "...                                                       ...   \n",
      "QLG62066.1  MAWHVPREELPEWLDRMDVPFDGHSNEEKHERMQDFAVYLADQAAA...   \n",
      "QLG62067.1  MIVQALVAAVELENVFVGVWDHKQTGPIPDSWQVVIETETVAGAEV...   \n",
      "QLG62068.1  MVLINRRQLMHQRWTAHILMAVLILVGGGVLFWAAISGGSDGLVVL...   \n",
      "QLG62069.1  MRIDTKLSEDDSERVKTLADETGLQKPAMYTRLVRTGLDIEERRER...   \n",
      "QLG62070.1  MPEHSNSGRSTAERSSLTDQEELEALQDDMKAVGEPDEKIANVSRV...   \n",
      "\n",
      "                                                genome xrefs EC_number gene  \n",
      "QLG61896.1  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN  \n",
      "QLG61897.1  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN  \n",
      "QLG61898.1  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN  \n",
      "QLG61899.1  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN  \n",
      "QLG64035.1  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN  \n",
      "...                                                      ...       ...  ...  \n",
      "QLG62066.1  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN  \n",
      "QLG62067.1  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN  \n",
      "QLG62068.1  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN  \n",
      "QLG62069.1  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN  \n",
      "QLG62070.1  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN  \n",
      "\n",
      "[179 rows x 11 columns]\n",
      "ID: LOEP01000011.1\n",
      "Name: LOEP01000011\n",
      "Description: Haloferax sp. Q22 scaffold11, whole genome shotgun sequence\n",
      "Database cross-references: BioProject:PRJNA294240, BioSample:SAMN04018175\n",
      "Number of features: 1\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=BCT\n",
      "/date=23-DEC-2015\n",
      "/accessions=['LOEP01000011', 'REGION:', '21500..36558']\n",
      "/sequence_version=1\n",
      "/keywords=['WGS']\n",
      "/source=Haloferax sp. Q22\n",
      "/organism=Haloferax sp. Q22\n",
      "/taxonomy=['Archaea', 'Euryarchaeota', 'Halobacteria', 'Haloferacales', 'Haloferacaceae', 'Haloferax']\n",
      "/references=[Reference(title='Genome sequence of Haloferax sp. strain Q22', ...), Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=OrderedDict([('Genome-Assembly-Data', OrderedDict([('Assembly Date', 'APR-2015'), ('Assembly Method', 'Velvet v. May-2015'), ('Genome Representation', 'Full'), ('Expected Final Version', 'Yes'), ('Genome Coverage', '417.9x'), ('Sequencing Technology', 'Illumina MiSeq')]))])\n",
      "Seq('GAAGTGGTCGATCTGGCCTTCGAGATGGCGATGCCAAGCCCTGACGACGCCGGG...AGG', IUPACAmbiguousDNA())\n",
      "                        protein_id  start   stop  \\\n",
      "LOEP01000011.1_1  LOEP01000011.1_1    497   1303   \n",
      "LOEP01000011.1_2  LOEP01000011.1_2   3372   4208   \n",
      "LOEP01000011.1_3  LOEP01000011.1_3   4638   6287   \n",
      "LOEP01000011.1_4  LOEP01000011.1_4   6242   7819   \n",
      "LOEP01000011.1_5  LOEP01000011.1_5   9085  10650   \n",
      "LOEP01000011.1_6  LOEP01000011.1_6  14723  13455   \n",
      "LOEP01000011.1_7  LOEP01000011.1_7  13216  12233   \n",
      "LOEP01000011.1_8  LOEP01000011.1_8   3352   2303   \n",
      "LOEP01000011.1_9  LOEP01000011.1_9   2263   1310   \n",
      "\n",
      "                                                      genome xrefs  \n",
      "LOEP01000011.1_1  [BioProject:PRJNA294240, BioSample:SAMN04018175]  \n",
      "LOEP01000011.1_2  [BioProject:PRJNA294240, BioSample:SAMN04018175]  \n",
      "LOEP01000011.1_3  [BioProject:PRJNA294240, BioSample:SAMN04018175]  \n",
      "LOEP01000011.1_4  [BioProject:PRJNA294240, BioSample:SAMN04018175]  \n",
      "LOEP01000011.1_5  [BioProject:PRJNA294240, BioSample:SAMN04018175]  \n",
      "LOEP01000011.1_6  [BioProject:PRJNA294240, BioSample:SAMN04018175]  \n",
      "LOEP01000011.1_7  [BioProject:PRJNA294240, BioSample:SAMN04018175]  \n",
      "LOEP01000011.1_8  [BioProject:PRJNA294240, BioSample:SAMN04018175]  \n",
      "LOEP01000011.1_9  [BioProject:PRJNA294240, BioSample:SAMN04018175]  \n",
      "ID: CP048739.1\n",
      "Name: CP048739\n",
      "Description: Halogeometricum borinquense strain wsp4 chromosome, complete genome\n",
      "Database cross-references: BioProject:PRJNA605299, BioSample:SAMN14053718\n",
      "Number of features: 75\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=BCT\n",
      "/date=18-FEB-2020\n",
      "/accessions=['CP048739', 'REGION:', '2797000..2827000']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Halogeometricum borinquense\n",
      "/organism=Halogeometricum borinquense\n",
      "/taxonomy=['Archaea', 'Euryarchaeota', 'Stenosarchaea group', 'Halobacteria', 'Haloferacales', 'Haloferacaceae', 'Halogeometricum']\n",
      "/references=[Reference(title='Whole genome sequence of Halogeometricum borinquense strain wsp4', ...), Reference(title='Direct Submission', ...)]\n",
      "/comment=The annotation was added by the NCBI Prokaryotic Genome Annotation\n",
      "Pipeline (PGAP). Information about PGAP can be found here:\n",
      "https://www.ncbi.nlm.nih.gov/genome/annotation_prok/\n",
      "/structured_comment=OrderedDict([('Genome-Assembly-Data', OrderedDict([('Assembly Method', 'CLC NGS Cell v. ver 9(CGWB)'), ('Genome Representation', 'Full'), ('Expected Final Version', 'Yes'), ('Genome Coverage', '1.0x'), ('Sequencing Technology', 'Illumina NextSeq')])), ('Genome-Annotation-Data', OrderedDict([('Annotation Provider', 'NCBI'), ('Annotation Date', '02/11/2020 16:53:36'), ('Annotation Pipeline', 'NCBI Prokaryotic Genome Annotation Pipeline (PGAP)'), ('Annotation Method', 'Best-placed reference protein set; GeneMarkS-2+'), ('Annotation Software revision', '4.11'), ('Features Annotated', 'Gene; CDS; rRNA; tRNA; ncRNA; repeat_region'), ('Genes (total)', '4,007'), ('CDSs (total)', '3,950'), ('Genes (coding)', '3,809'), ('CDSs (with protein)', '3,809'), ('Genes (RNA)', '57'), ('rRNAs', '1, 1, 1 (5S, 16S, 23S)'), ('complete rRNAs', '1, 1, 1 (5S, 16S, 23S)'), ('tRNAs', '52'), ('ncRNAs', '2'), ('Pseudo Genes (total)', '141'), ('CDSs (without protein)', '141'), ('Pseudo Genes (ambiguous residues)', '0 of 141'), ('Pseudo Genes (frameshifted)', '59 of 141'), ('Pseudo Genes (incomplete)', '96 of 141'), ('Pseudo Genes (internal stop)', '21 of 141'), ('Pseudo Genes (multiple problems)', '29 of 141'), ('CRISPR Arrays', '3')]))])\n",
      "Seq('CCGGACGGAGCGATACACCGTGATGGTGCGGACCGAAAGCGGCCGGATGGAGCG...CCG', IUPACAmbiguousDNA())\n",
      "              locus_tag                                          inference  \\\n",
      "QIB75350.1  G3I44_14265     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75351.1  G3I44_14270     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75352.1  G3I44_14275     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75353.1  G3I44_14280     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75354.1  G3I44_14285     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75355.1  G3I44_14290     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75356.1  G3I44_14295     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75357.1  G3I44_14300     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75358.1  G3I44_14305     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75359.1  G3I44_14310     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75360.1  G3I44_14315  COORDINATES: similar to AA sequence:RefSeq:WP_...   \n",
      "QIB75361.1  G3I44_14320     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75362.1  G3I44_14325     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75363.1  G3I44_14330     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75364.1  G3I44_14335     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75365.1  G3I44_14340          COORDINATES: protein motif:HMM:NF012920.1   \n",
      "QIB75366.1  G3I44_14345     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75367.1  G3I44_14350          COORDINATES: protein motif:HMM:NF024777.1   \n",
      "QIB75368.1  G3I44_14355          COORDINATES: protein motif:HMM:NF016908.1   \n",
      "QIB75369.1  G3I44_14360     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75370.1  G3I44_14365     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75371.1  G3I44_14370     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75372.1  G3I44_14375     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75373.1  G3I44_14380     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75374.1  G3I44_14385          COORDINATES: protein motif:HMM:NF014044.1   \n",
      "QIB75375.1  G3I44_14390     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75376.1  G3I44_14395     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75377.1  G3I44_14400     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75378.1  G3I44_14405     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75379.1  G3I44_14410     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75380.1  G3I44_14415     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75381.1  G3I44_14420     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75382.1  G3I44_14425     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75383.1  G3I44_14430     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75384.1  G3I44_14435     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75385.1  G3I44_14440     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "QIB75386.1  G3I44_14445          COORDINATES: protein motif:HMM:NF025817.1   \n",
      "\n",
      "                                                         note codon_start  \\\n",
      "QIB75350.1  Derived by automated computational analysis us...           2   \n",
      "QIB75351.1  Derived by automated computational analysis us...           1   \n",
      "QIB75352.1  Derived by automated computational analysis us...           1   \n",
      "QIB75353.1  Derived by automated computational analysis us...           1   \n",
      "QIB75354.1  Derived by automated computational analysis us...           1   \n",
      "QIB75355.1  Derived by automated computational analysis us...           1   \n",
      "QIB75356.1  Derived by automated computational analysis us...           1   \n",
      "QIB75357.1  Derived by automated computational analysis us...           1   \n",
      "QIB75358.1  Derived by automated computational analysis us...           1   \n",
      "QIB75359.1  Derived by automated computational analysis us...           1   \n",
      "QIB75360.1  Derived by automated computational analysis us...           1   \n",
      "QIB75361.1  Derived by automated computational analysis us...           1   \n",
      "QIB75362.1  Derived by automated computational analysis us...           1   \n",
      "QIB75363.1  Derived by automated computational analysis us...           1   \n",
      "QIB75364.1  Derived by automated computational analysis us...           1   \n",
      "QIB75365.1  Derived by automated computational analysis us...           1   \n",
      "QIB75366.1  Derived by automated computational analysis us...           1   \n",
      "QIB75367.1  Derived by automated computational analysis us...           1   \n",
      "QIB75368.1  Derived by automated computational analysis us...           1   \n",
      "QIB75369.1  Derived by automated computational analysis us...           1   \n",
      "QIB75370.1  Derived by automated computational analysis us...           1   \n",
      "QIB75371.1  Derived by automated computational analysis us...           1   \n",
      "QIB75372.1  Derived by automated computational analysis us...           1   \n",
      "QIB75373.1  Derived by automated computational analysis us...           1   \n",
      "QIB75374.1  Derived by automated computational analysis us...           1   \n",
      "QIB75375.1  Derived by automated computational analysis us...           1   \n",
      "QIB75376.1  Derived by automated computational analysis us...           1   \n",
      "QIB75377.1  Derived by automated computational analysis us...           1   \n",
      "QIB75378.1  Derived by automated computational analysis us...           1   \n",
      "QIB75379.1  Derived by automated computational analysis us...           1   \n",
      "QIB75380.1  Derived by automated computational analysis us...           1   \n",
      "QIB75381.1  Derived by automated computational analysis us...           1   \n",
      "QIB75382.1  Derived by automated computational analysis us...           1   \n",
      "QIB75383.1  Derived by automated computational analysis us...           1   \n",
      "QIB75384.1  Derived by automated computational analysis us...           1   \n",
      "QIB75385.1  Derived by automated computational analysis us...           1   \n",
      "QIB75386.1  Derived by automated computational analysis us...           1   \n",
      "\n",
      "           transl_table                            product  protein_id  \\\n",
      "QIB75350.1           11               hypothetical protein  QIB75350.1   \n",
      "QIB75351.1           11               hypothetical protein  QIB75351.1   \n",
      "QIB75352.1           11  DUF4258 domain-containing protein  QIB75352.1   \n",
      "QIB75353.1           11               hypothetical protein  QIB75353.1   \n",
      "QIB75354.1           11            site-specific integrase  QIB75354.1   \n",
      "QIB75355.1           11               hypothetical protein  QIB75355.1   \n",
      "QIB75356.1           11               hypothetical protein  QIB75356.1   \n",
      "QIB75357.1           11               hypothetical protein  QIB75357.1   \n",
      "QIB75358.1           11               hypothetical protein  QIB75358.1   \n",
      "QIB75359.1           11               hypothetical protein  QIB75359.1   \n",
      "QIB75360.1           11               hypothetical protein  QIB75360.1   \n",
      "QIB75361.1           11               hypothetical protein  QIB75361.1   \n",
      "QIB75362.1           11               hypothetical protein  QIB75362.1   \n",
      "QIB75363.1           11               hypothetical protein  QIB75363.1   \n",
      "QIB75364.1           11               hypothetical protein  QIB75364.1   \n",
      "QIB75365.1           11        S26 family signal peptidase  QIB75365.1   \n",
      "QIB75366.1           11               hypothetical protein  QIB75366.1   \n",
      "QIB75367.1           11     LamG domain-containing protein  QIB75367.1   \n",
      "QIB75368.1           11      FkbM family methyltransferase  QIB75368.1   \n",
      "QIB75369.1           11               hypothetical protein  QIB75369.1   \n",
      "QIB75370.1           11               hypothetical protein  QIB75370.1   \n",
      "QIB75371.1           11               hypothetical protein  QIB75371.1   \n",
      "QIB75372.1           11               hypothetical protein  QIB75372.1   \n",
      "QIB75373.1           11               hypothetical protein  QIB75373.1   \n",
      "QIB75374.1           11                ATP-binding protein  QIB75374.1   \n",
      "QIB75375.1           11               hypothetical protein  QIB75375.1   \n",
      "QIB75376.1           11               hypothetical protein  QIB75376.1   \n",
      "QIB75377.1           11               hypothetical protein  QIB75377.1   \n",
      "QIB75378.1           11               hypothetical protein  QIB75378.1   \n",
      "QIB75379.1           11               hypothetical protein  QIB75379.1   \n",
      "QIB75380.1           11               hypothetical protein  QIB75380.1   \n",
      "QIB75381.1           11               hypothetical protein  QIB75381.1   \n",
      "QIB75382.1           11               hypothetical protein  QIB75382.1   \n",
      "QIB75383.1           11               hypothetical protein  QIB75383.1   \n",
      "QIB75384.1           11               hypothetical protein  QIB75384.1   \n",
      "QIB75385.1           11               hypothetical protein  QIB75385.1   \n",
      "QIB75386.1           11               hypothetical protein  QIB75386.1   \n",
      "\n",
      "                                                  translation  \\\n",
      "QIB75350.1  MADTQRPTEIPSERRFDGKKFHFHTYRYGSGARQELQSLGERKKSW...   \n",
      "QIB75351.1  MEQLLDESEYREAWFDGHGDLVFSDEPVYEVSGEPLPLEKGGTSTV...   \n",
      "QIB75352.1  MSTELQQEPDSDQPGIHQLLGIQYIRPSDHGNARAQERGIPDDYIA...   \n",
      "QIB75353.1  MPDDFGGSEFFGEDEDERSPSDHFESRHSPRLGYLFIEKSLSSLDA...   \n",
      "QIB75354.1  MSTPRTIDEDARKFETWLNRETSCGGGTPGMYAAYIKRLDTPIDLN...   \n",
      "QIB75355.1  MSAIKIMLGMGVSGASMGLVAFDYLREANGWIYSDLNTAYLAWAAI...   \n",
      "QIB75356.1  MSEVPVTRSSYRIKVCPDCGSQKCARDGMKWRCWECDNRFEEFQTV...   \n",
      "QIB75357.1  MIELTDGDKWAIVENLIANHHSYQQRIENGDFPGLLSEEVCREGMQ...   \n",
      "QIB75358.1  MVPIETDDVTVKPAEPIDPFEVLTAFKRRGMKKNCHKDVDASKDSA...   \n",
      "QIB75359.1  MTEPNLEDRPDLSKLNYSHEKLLELWEGNQNTRGIRDILRDEGRLC...   \n",
      "QIB75360.1  MTRQWGIDPELLCRQHLLGEHREMHQEVGHINADNLATVKGHARCG...   \n",
      "QIB75361.1  MSIEDVFRVFIRTGRALGSLMILYSMYAIGKDEAAQAAAKLFKDETTSR   \n",
      "QIB75362.1  MAKTGINRYDELPERVGHINVDQHRDALGQALVNLEQATDLHADLQ...   \n",
      "QIB75363.1  MTDINLTPSQILPLFIEWVNDRGTVRIDSSSAAGNGKRYFINRDRG...   \n",
      "QIB75364.1  MSHPFEDEELNRALRSFLEDDSISYQPYEEDERPGPPRIQQRHLID...   \n",
      "QIB75365.1  MSWSMVLNLRVLAVLSILLSFSAGWIVSDMANPYTVTVQQPTASTD...   \n",
      "QIB75366.1  MFTILMTVVIWLMSIGMVLFGYVTNDAGLAFTGVVFAVALPLFTTV...   \n",
      "QIB75367.1  MDTRTKALLITVLFLVALPLASAKASGLEGFSAEKVNESFHADHIR...   \n",
      "QIB75368.1  MNLAKKVHWLYQEYIFPRFYHLRGVTRLSKVLRSLAVHDFDPEEDG...   \n",
      "QIB75369.1  MTITDADRETFQSTVEDFQPQITEDMCLPTALKNVLDEFAERHGAD...   \n",
      "QIB75370.1  MASADRGTDNRLRFRSLEEIRTGRAETELKHRYGRGIYVNEIHEAS...   \n",
      "QIB75371.1  MKRLAAYLAAIALLVITGFFVATAGPLSSVSQLCTGGADEVFIDNN...   \n",
      "QIB75372.1  MGFSSKQVLGIVALVVVGMVSTGFVMSGASVGDVLPFSSAAQNQQP...   \n",
      "QIB75373.1  MATGLVSAVVGAFDPLSVAVLLVIGLAIGLKTQGYDIRVHKGFWAV...   \n",
      "QIB75374.1  MYRYGKFAKAGDQTHQRYVPDEKDLAPPPDYEKAIPIGTDINTGEL...   \n",
      "QIB75375.1  MKLRHLAISLFGFIGFVLVNARYELMPWYYFLFVGAVVALTVVVHL...   \n",
      "QIB75376.1  MLDIALPHMMAAWSLSAVFLSGLVVLGENVGGSASSKLTYLVIGNV...   \n",
      "QIB75377.1  MAVFIKQLYDAKRGDFVTVDDVPKGRLEYDQTVTHVDEFGRRHKYR...   \n",
      "QIB75378.1  MAATQYEPDLVQKILNETDLSDADRDRQRDIEKMDEERLVAQKLNK...   \n",
      "QIB75379.1  MVKEDFTFDAFYGAYHGGFTQDYNQTSRGKHPFIRLMGLPRKKEFP...   \n",
      "QIB75380.1  MKRLTLLSACLFLFITTTAALNVSINQSTTTLNPGTTAQLGEILLV...   \n",
      "QIB75381.1  MSEEFETTVDDVNGNLADLIQHNYSGRHLLERLLIRLIPFVNQTVT...   \n",
      "QIB75382.1      MVDILVSGCSEEEAKNLRVAKAAEGYTSWREMMLDLAGVESDGDE   \n",
      "QIB75383.1  MSIGRALDALKALYDVLKVVLGGTYSGYRYVRGLGLVSKITGAVTF...   \n",
      "QIB75384.1  MDRSPEGFEVLADIFSALGNRTRLAVLYGLYEGDSMPEVAEFLEVE...   \n",
      "QIB75385.1  MSQTVYIDGESVSIDEHTTGKELKAVAGADEDAELIHHDGDETVRI...   \n",
      "QIB75386.1  MNLPRVQTELAALRQRYSVAVNLNNEFPYVIVKDVDFGPDWRPRQG...   \n",
      "\n",
      "                                                genome xrefs  \n",
      "QIB75350.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75351.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75352.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75353.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75354.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75355.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75356.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75357.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75358.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75359.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75360.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75361.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75362.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75363.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75364.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75365.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75366.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75367.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75368.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75369.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75370.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75371.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75372.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75373.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75374.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75375.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75376.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75377.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75378.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75379.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75380.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75381.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75382.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75383.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75384.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75385.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "QIB75386.1  [BioProject:PRJNA605299, BioSample:SAMN14053718]  \n",
      "ID: JABURA010000001.1\n",
      "Name: JABURA010000001\n",
      "Description: Haloterrigena sp. SYSU A121-1 Contig00001, whole genome shotgun sequence\n",
      "Database cross-references: BioProject:PRJNA637493, BioSample:SAMN15143775\n",
      "Number of features: 225\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=BCT\n",
      "/date=14-JUN-2020\n",
      "/accessions=['JABURA010000001', 'REGION:', '1761500..1860500']\n",
      "/sequence_version=1\n",
      "/keywords=['WGS']\n",
      "/source=Haloterrigena sp. SYSU A121-1\n",
      "/organism=Haloterrigena sp. SYSU A121-1\n",
      "/taxonomy=['Archaea', 'Euryarchaeota', 'Stenosarchaea group', 'Halobacteria', 'Natrialbales', 'Natrialbaceae', 'Haloterrigena', 'unclassified Haloterrigena']\n",
      "/references=[Reference(title='Haloterrigena sp. nov., an extremely halophilic archaeon isolated from a saline sediment', ...), Reference(title='Direct Submission', ...)]\n",
      "/comment=The annotation was added by the NCBI Prokaryotic Genome Annotation\n",
      "Pipeline (PGAP). Information about PGAP can be found here:\n",
      "https://www.ncbi.nlm.nih.gov/genome/annotation_prok/\n",
      "/structured_comment=OrderedDict([('Genome-Assembly-Data', OrderedDict([('Assembly Date', '06-JUN-2020'), ('Assembly Method', 'SOAPdenovo v. 3.6.0'), ('Genome Representation', 'Full'), ('Expected Final Version', 'Yes'), ('Genome Coverage', '100x'), ('Sequencing Technology', 'Illumina MiSeq')])), ('Genome-Annotation-Data', OrderedDict([('Annotation Provider', 'NCBI'), ('Annotation Date', '06/09/2020 15:35:07'), ('Annotation Pipeline', 'NCBI Prokaryotic Genome Annotation Pipeline (PGAP)'), ('Annotation Method', 'Best-placed reference protein set; GeneMarkS-2+'), ('Annotation Software revision', '4.11'), ('Features Annotated', 'Gene; CDS; rRNA; tRNA; ncRNA; repeat_region'), ('Genes (total)', '4,773'), ('CDSs (total)', '4,703'), ('Genes (coding)', '4,602'), ('CDSs (with protein)', '4,602'), ('Genes (RNA)', '70'), ('rRNAs', '5, 4, 4 (5S, 16S, 23S)'), ('complete rRNAs', '5, 4, 4 (5S, 16S, 23S)'), ('tRNAs', '55'), ('ncRNAs', '2'), ('Pseudo Genes (total)', '101'), ('CDSs (without protein)', '101'), ('Pseudo Genes (ambiguous residues)', '0 of 101'), ('Pseudo Genes (frameshifted)', '44 of 101'), ('Pseudo Genes (incomplete)', '59 of 101'), ('Pseudo Genes (internal stop)', '19 of 101'), ('Pseudo Genes (multiple problems)', '21 of 101')]))])\n",
      "UnknownSeq(99001, alphabet=IUPACAmbiguousDNA(), character='N')\n",
      "              locus_tag                                          inference  \\\n",
      "NUB91080.1  HT576_08605  COORDINATES: similar to AA sequence:RefSeq:WP_...   \n",
      "NUB91081.1  HT576_08615  COORDINATES: similar to AA sequence:RefSeq:WP_...   \n",
      "NUB91082.1  HT576_08620     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "NUB91083.1  HT576_08625     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "NUB91084.1  HT576_08630     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "...                 ...                                                ...   \n",
      "NUB91184.1  HT576_09140     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "NUB91185.1  HT576_09145          COORDINATES: protein motif:HMM:NF014044.1   \n",
      "NUB91186.1  HT576_09150  COORDINATES: protein motif:HMM:NF012722.1,HMM:...   \n",
      "NUB91187.1  HT576_09155     COORDINATES: ab initio prediction:GeneMarkS-2+   \n",
      "NUB91188.1  HT576_09160  COORDINATES: similar to AA sequence:RefSeq:WP_...   \n",
      "\n",
      "                                                         note codon_start  \\\n",
      "NUB91080.1  Derived by automated computational analysis us...           2   \n",
      "NUB91081.1  Derived by automated computational analysis us...           1   \n",
      "NUB91082.1  Derived by automated computational analysis us...           1   \n",
      "NUB91083.1  Derived by automated computational analysis us...           1   \n",
      "NUB91084.1  Derived by automated computational analysis us...           1   \n",
      "...                                                       ...         ...   \n",
      "NUB91184.1  Derived by automated computational analysis us...           1   \n",
      "NUB91185.1  Derived by automated computational analysis us...           1   \n",
      "NUB91186.1  Derived by automated computational analysis us...           1   \n",
      "NUB91187.1  Derived by automated computational analysis us...           1   \n",
      "NUB91188.1  Derived by automated computational analysis us...           1   \n",
      "\n",
      "           transl_table                                            product  \\\n",
      "NUB91080.1           11  twin-arginine translocase TatA/TatE family sub...   \n",
      "NUB91081.1           11                            site-specific integrase   \n",
      "NUB91082.1           11                               hypothetical protein   \n",
      "NUB91083.1           11                               hypothetical protein   \n",
      "NUB91084.1           11                               hypothetical protein   \n",
      "...                 ...                                                ...   \n",
      "NUB91184.1           11          TFIIB-type zinc ribbon-containing protein   \n",
      "NUB91185.1           11                                ATP-binding protein   \n",
      "NUB91186.1           11                      PAS domain-containing protein   \n",
      "NUB91187.1           11                               hypothetical protein   \n",
      "NUB91188.1           11                tyrosine-type recombinase/integrase   \n",
      "\n",
      "            protein_id                                        translation  \\\n",
      "NUB91080.1  NUB91080.1  MAIETAPLFVPTPGAPELLIILFVAILLFGANKIPKLARSTGEAMG...   \n",
      "NUB91081.1  NUB91081.1  MDGIDNMTIEQLQQVLSEASPSLQPLSPEEGVGRYLKRREGEITEN...   \n",
      "NUB91082.1  NUB91082.1  MDLTWLILEETFILMVAIWLGVLLFAPLEHANAFWLTGAIHMITAL...   \n",
      "NUB91083.1  NUB91083.1  MVDDDTTPGIDIGTGFEKEETEHFAERGTDIVHHKVKTVDPATGRI...   \n",
      "NUB91084.1  NUB91084.1  MVSTDTIPDALKDRDQWICWTTEDRDGKPTKVPIDPVNDGYASVSD...   \n",
      "...                ...                                                ...   \n",
      "NUB91184.1  NUB91184.1  MRIETQLSEDDSDLVQKLADRRGLQKDRAYTELIRTGVDVGMRRER...   \n",
      "NUB91185.1  NUB91185.1  MAWIGDDEDTPSREELAELQEMLELVGLSAEKIKQILSMAAMDRDA...   \n",
      "NUB91186.1  NUB91186.1  MSPEVDPDLLHIGFHALPFEVALVDETGQIVFANEAWKAFGKANGA...   \n",
      "NUB91187.1  NUB91187.1  MIVPVSVAGCSDTVLSNGSEGEGDTIEIMVENRTDDRAQIGVRVED...   \n",
      "NUB91188.1  NUB91188.1  MTDLEPIDPETAIDLYLTDRDSEVTAATLQSHRSRLGHFVRWCDQQ...   \n",
      "\n",
      "                                                genome xrefs EC_number  \n",
      "NUB91080.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN  \n",
      "NUB91081.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN  \n",
      "NUB91082.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN  \n",
      "NUB91083.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN  \n",
      "NUB91084.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN  \n",
      "...                                                      ...       ...  \n",
      "NUB91184.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN  \n",
      "NUB91185.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN  \n",
      "NUB91186.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN  \n",
      "NUB91187.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN  \n",
      "NUB91188.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN  \n",
      "\n",
      "[109 rows x 10 columns]\n"
     ]
    }
   ],
   "source": [
    "featureDfs = {}\n",
    "\n",
    "def rungetORF( geno , proteome , minsize = 600 ):\n",
    "    args = 'getorf -sequence '+ geno + ' -outseq ' + proteome +' -minsize ' + str(minsize)\n",
    "    p = subprocess.run( shlex.split(args) )\n",
    "    return p, proteome\n",
    "\n",
    "for filename in glob.glob('*.gb'):\n",
    "    parsed = SeqIO.parse( filename , 'gb')\n",
    "    featuredict={}\n",
    "    fastastr = ''\n",
    "    for i,seq in enumerate(parsed):\n",
    "            print(seq)\n",
    "            xrefs = {'genome xrefs' : seq.dbxrefs }\n",
    "            for feature in seq.features:\n",
    "                if feature.type == 'CDS':\n",
    "                    if 'protein_id' in feature.qualifiers:\n",
    "                        featuredict[feature.qualifiers['protein_id'][0]]  = {q:feature.qualifiers[q][0] for q in feature.qualifiers }\n",
    "                        featuredict[feature.qualifiers['protein_id'][0]].update(xrefs)\n",
    "                        fastastr+= '>'+ feature.qualifiers['protein_id'][0] + '\\n' + feature.qualifiers['translation'][0] +'\\n'\n",
    "    \n",
    "    if len(fastastr)==0:\n",
    "        #fall back on getORF if NCBI annot unavailable\n",
    "        geno = seq\n",
    "        with open( filename+'.convert.fasta' , 'w') as fastout:\n",
    "            SeqIO.write( geno , fastout, 'fasta')\n",
    "        #predict orfs of all genos w hap2\n",
    "        rungetORF(   filename+'.convert.fasta' , filename+'.convert.proteo.fasta')\n",
    "        \n",
    "        #rungetorf\n",
    "        parsed = SeqIO.parse( filename+'.convert.proteo.fasta' , 'fasta')\n",
    "        for orf in parsed:\n",
    "            #JNCS01000001.1_1 [708 - 1748] Natrinema altunense strain AJ2 N_altunense_AJ2_contig_1, whole genome shotgun sequence\n",
    "            featuredict[orf.id] = dict( zip (['protein_id', 'start' , 'dash' ,'stop'] , orf.description.split()[0:4] )) \n",
    "            del featuredict[orf.id]['dash']\n",
    "            featuredict[orf.id]['start'] = featuredict[orf.id]['start'].replace('[','')\n",
    "            featuredict[orf.id]['stop']= featuredict[orf.id]['stop'].replace(']','')\n",
    "            featuredict[orf.id].update(xrefs)\n",
    "            #parse output\n",
    "    featureDf = pd.DataFrame.from_dict( featuredict , orient = 'index')\n",
    "    featureDfs[filename] = featureDf\n",
    "    print(featureDf)\n",
    "    with open(filename+'.proteo.fasta', 'w') as fastout:\n",
    "        fastout.write( fastastr )\n",
    "        \n",
    "#add patric annot for hnatans\n",
    "featureDfs['hnatans'] = sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['./Naltunense1.gb.fastahmmerscan.csv'\n",
      " './halogeometricum.gb.proteo.fastahmmerscan.csv'\n",
      " './haloterigena_reg.gb.proteo.fastahmmerscan.csv'\n",
      " './Nal.gb.convert.proteo.fastahmmerscan.csv'\n",
      " './halovivax.gb.proteo.fastahmmerscan.csv'\n",
      " './haloterigena_reg.gb.fastahmmerscan.csv'\n",
      " './halobonum.gb.fastahmmerscan.csv' './halovivax.gb.fastahmmerscan.csv'\n",
      " './Naltunense1.gb.proteo.fastahmmerscan.csv'\n",
      " './haloferax.gb.convert.proteo.fastahmmerscan.csv'\n",
      " './926690.3.PATRIC.proteo.fastahmmerscan.csv'\n",
      " './halogeometricum.gb.fastahmmerscan.csv'\n",
      " './halobonum.gb.proteo.fastahmmerscan.csv']\n",
      "          targetname   accession                queryid accession2 E-value  \\\n",
      "3                AAA  PF00004.31  fig|926690.3.peg.2651          -  0.0016   \n",
      "4             DUF853  PF05872.14  fig|926690.3.peg.2651          -  0.0068   \n",
      "5              DUF87  PF01935.19  fig|926690.3.peg.2651          -  0.0085   \n",
      "6           SHNi-TPR  PF10516.11  fig|926690.3.peg.2651          -    0.13   \n",
      "7             DUF309  PF03745.16  fig|926690.3.peg.2651          -    0.29   \n",
      "...              ...         ...                    ...        ...     ...   \n",
      "3360             NMO  PF03060.17  fig|926690.3.peg.3333          -  0.0002   \n",
      "3361      SymE_toxin  PF08845.12  fig|926690.3.peg.3334          -   0.002   \n",
      "3362  MazE_antitoxin  PF04014.20  fig|926690.3.peg.3334          -   0.012   \n",
      "3363            DtxR   PF18357.3  fig|926690.3.peg.3334          -   0.049   \n",
      "3364             MC1  PF05854.13  fig|926690.3.peg.3335          -   4e-41   \n",
      "\n",
      "      score bias domain-E-value domain-score domain-bias  ... reg clu ov env  \\\n",
      "3      18.9  0.0           0.69         10.4         0.0  ...   2   0  0   2   \n",
      "4      15.2  0.0           0.01         14.6         0.0  ...   1   0  0   1   \n",
      "5      16.3  0.0          0.014         15.6         0.0  ...   1   1  0   1   \n",
      "6      11.8  0.2           0.34         10.5         0.2  ...   1   0  0   1   \n",
      "7      11.0  1.7            0.6          9.9         1.7  ...   1   0  0   1   \n",
      "...     ...  ...            ...          ...         ...  ...  ..  .. ..  ..   \n",
      "3360   21.0  0.3        0.00088         18.9         0.2  ...   2   0  0   2   \n",
      "3361   18.3  0.0          0.002         18.2         0.0  ...   1   0  0   1   \n",
      "3362   15.6  0.3          0.018         15.0         0.3  ...   1   0  0   1   \n",
      "3363   14.0  0.1          0.063         13.7         0.1  ...   1   0  0   1   \n",
      "3364  139.3  0.4        4.8e-41        139.0         0.4  ...   1   0  0   1   \n",
      "\n",
      "     dom rep inc                                        description  \\\n",
      "3      2   2   2  ATPase family associated with various cellular...   \n",
      "4      1   1   1             Bacterial protein of unknown function    \n",
      "5      1   1   1                            Helicase HerA, central    \n",
      "6      1   1   0                                                      \n",
      "7      1   1   0                        Domain of unknown function    \n",
      "...   ..  ..  ..                                                ...   \n",
      "3360   2   2   1                                         Nitronate    \n",
      "3361   1   1   1                Toxin SymE, type I toxin-antitoxin    \n",
      "3362   1   1   0        Antidote-toxin recognition MazE, bacterial    \n",
      "3363   1   1   0                     Diphteria toxin repressor SH3    \n",
      "3364   1   1   1                   Non-histone chromosomal protein    \n",
      "\n",
      "                                           infile domain  \n",
      "3     ./926690.3.PATRIC.proteo.fastahmmerscan.csv      4  \n",
      "4     ./926690.3.PATRIC.proteo.fastahmmerscan.csv   5872  \n",
      "5     ./926690.3.PATRIC.proteo.fastahmmerscan.csv   1935  \n",
      "6     ./926690.3.PATRIC.proteo.fastahmmerscan.csv  10516  \n",
      "7     ./926690.3.PATRIC.proteo.fastahmmerscan.csv   3745  \n",
      "...                                           ...    ...  \n",
      "3360  ./926690.3.PATRIC.proteo.fastahmmerscan.csv   3060  \n",
      "3361  ./926690.3.PATRIC.proteo.fastahmmerscan.csv   8845  \n",
      "3362  ./926690.3.PATRIC.proteo.fastahmmerscan.csv   4014  \n",
      "3363  ./926690.3.PATRIC.proteo.fastahmmerscan.csv  18357  \n",
      "3364  ./926690.3.PATRIC.proteo.fastahmmerscan.csv   5854  \n",
      "\n",
      "[3362 rows x 21 columns]\n"
     ]
    }
   ],
   "source": [
    "dfs = []\n",
    "annotations = glob.glob( './*hmmerscan.csv')\n",
    "header = 'targetname accession queryid accession2    E-value  score  bias   domain-E-value  domain-score  domain-bias   exp reg clu  ov env dom rep inc description'.split()\n",
    "for filename in  annotations:\n",
    "    framedict = {}\n",
    "    with open(filename , 'r') as infile:\n",
    "        for i,line in enumerate(infile):\n",
    "            if line[0] != '#':\n",
    "                words = line.split()\n",
    "                words[18] = ''.join([ w + ' ' for w in words[18:-1]])\n",
    "                #print(words)\n",
    "                framedict[i] = dict(zip( header , words[0:19]))\n",
    "        df = pd.DataFrame.from_dict(framedict , orient='index')\n",
    "        df['infile'] = filename\n",
    "        #df.to_csv(filename + '.reformat.csv')\n",
    "        dfs.append(df)\n",
    "globaldf = pd.concat(dfs)\n",
    "globaldf['domain'] = globaldf.accession.map(lambda x : str(int(x.split('.')[0].replace('PF',''))) )\n",
    "\n",
    "print(globaldf.infile.unique())\n",
    "print(globaldf[globaldf.infile == './926690.3.PATRIC.proteo.fastahmmerscan.csv'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['targetname', 'accession', 'queryid', 'accession2', 'E-value', 'score',\n",
      "       'bias', 'domain-E-value', 'domain-score', 'domain-bias', 'exp', 'reg',\n",
      "       'clu', 'ov', 'env', 'dom', 'rep', 'inc', 'description', 'infile',\n",
      "       'domain'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(globaldf.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "globaldf['E-value'] =globaldf['E-value'].map(lambda x : float(x)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          targetname  accession     queryid accession2       E-value  score  \\\n",
      "16   Phage_integrase  pfam00589  ELY83596.1          -  7.800000e-18   64.9   \n",
      "17   Phage_int_SAM_4  pfam13495  ELY83596.1          -  2.000000e-11   44.3   \n",
      "18   Phage_int_SAM_1  pfam02899  ELY83596.1          -  6.900000e-05   23.1   \n",
      "19   Phage_int_SAM_5  pfam13102  ELY83596.1          -  8.800000e-04   19.7   \n",
      "22        OrfB_IS605  pfam01385  ELY83598.1          -  8.900000e-31  106.7   \n",
      "..               ...        ...         ...        ...           ...    ...   \n",
      "504            PQQ_2  pfam13360  QLG62040.1          -  1.200000e-48  166.0   \n",
      "513           HalOD1  pfam18545  QLG62045.1          -  7.000000e-10   39.1   \n",
      "514           TrkA_C  pfam02080  QLG62049.1          -  2.900000e-11   43.2   \n",
      "540            DUF87  pfam01935  QLG62070.1          -  3.400000e-06   27.4   \n",
      "541     FtsK_SpoIIIE  pfam01580  QLG62070.1          -  7.400000e-04   19.0   \n",
      "\n",
      "     bias domain-E-value domain-score domain-bias  ... reg clu ov env dom rep  \\\n",
      "16    0.0        1.4e-17         64.1         0.0  ...   1   0  0   1   1   1   \n",
      "17    0.1        4.2e-11         43.3         0.1  ...   1   0  0   1   1   1   \n",
      "18    0.1        0.00014         22.1         0.1  ...   1   0  0   1   1   1   \n",
      "19    0.3         0.0024         18.3         0.1  ...   2   0  0   2   2   2   \n",
      "22    1.8        8.9e-31        106.7         1.8  ...   2   0  0   2   2   2   \n",
      "..    ...            ...          ...         ...  ...  ..  .. ..  ..  ..  ..   \n",
      "504  19.5        1.4e-19         70.8         4.3  ...   2   1  2   4   4   4   \n",
      "513   0.8        8.1e-10         38.9         0.8  ...   1   0  0   1   1   1   \n",
      "514   0.0        5.7e-11         42.2         0.0  ...   1   0  0   1   1   1   \n",
      "540   0.0        2.3e-05         24.7         0.0  ...   1   1  0   1   1   1   \n",
      "541   0.0         0.0024         17.3         0.0  ...   2   0  0   2   2   2   \n",
      "\n",
      "    inc                            description  \\\n",
      "16    1                       Phage integrase    \n",
      "17    1  Phage integrase, N-terminal SAM-like    \n",
      "18    1  Phage integrase, N-terminal SAM-like    \n",
      "19    1              Phage integrase SAM-like    \n",
      "22    1                              Probable    \n",
      "..   ..                                    ...   \n",
      "504   3                              PQQ-like    \n",
      "513   1           Halobacterial output domain    \n",
      "514   1                                TrkA-C    \n",
      "540   1                Helicase HerA, central    \n",
      "541   1                          FtsK/SpoIIIE    \n",
      "\n",
      "                                       infile domain  \n",
      "16        ./Naltunense1.gb.fastahmmerscan.csv    589  \n",
      "17        ./Naltunense1.gb.fastahmmerscan.csv  13495  \n",
      "18        ./Naltunense1.gb.fastahmmerscan.csv   2899  \n",
      "19        ./Naltunense1.gb.fastahmmerscan.csv  13102  \n",
      "22        ./Naltunense1.gb.fastahmmerscan.csv   1385  \n",
      "..                                        ...    ...  \n",
      "504  ./halobonum.gb.proteo.fastahmmerscan.csv  13360  \n",
      "513  ./halobonum.gb.proteo.fastahmmerscan.csv  18545  \n",
      "514  ./halobonum.gb.proteo.fastahmmerscan.csv   2080  \n",
      "540  ./halobonum.gb.proteo.fastahmmerscan.csv   1935  \n",
      "541  ./halobonum.gb.proteo.fastahmmerscan.csv   1580  \n",
      "\n",
      "[2149 rows x 21 columns]\n"
     ]
    }
   ],
   "source": [
    "globaldf=globaldf[globaldf['E-value'] < .001]\n",
    "globaldf.accession = globaldf.accession.map( lambda x : x.split('.')[0].lower().replace('pf', 'pfam'))\n",
    "print(globaldf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13443\n",
      "5232\n",
      "         arCOG-id functional-class gene-name  \\\n",
      "0      arCOG00001                K         -   \n",
      "1      arCOG00002                K         -   \n",
      "2      arCOG00004                K         -   \n",
      "3      arCOG00005                K         -   \n",
      "4      arCOG00006                K         -   \n",
      "...           ...              ...       ...   \n",
      "13280  arCOG15117                S         -   \n",
      "13290  arCOG15127                S         -   \n",
      "13318  arCOG15155                S         -   \n",
      "13331  arCOG15168                P         -   \n",
      "13417  arCOG15257                P         -   \n",
      "\n",
      "                             arCOG-annotation supercluster profile-pfam  \\\n",
      "0      Transcriptional regulator, PadR family     COG01695    pfam03551   \n",
      "1      Transcriptional regulator, PadR family     COG01695    pfam03551   \n",
      "2      Transcriptional regulator, PadR family     COG01695    pfam03551   \n",
      "3      Transcriptional regulator, PadR family     COG01695    pfam03551   \n",
      "4      Transcriptional regulator, PadR family     COG01695    pfam03551   \n",
      "...                                       ...          ...          ...   \n",
      "13280                 Uncharacterized protein          NaN    pfam13613   \n",
      "13290        Uncharacterized membrane protein     SC.01024    pfam13829   \n",
      "13318        Uncharacterized membrane protein          NaN    pfam04156   \n",
      "13331            Sulfurtransferase TusA famiy     COG00425    pfam01206   \n",
      "13417           Heavy-metal-associated domain     COG02608    pfam00403   \n",
      "\n",
      "      profile-cdd profile-tigrfam  \n",
      "0             NaN             NaN  \n",
      "1             NaN       TIGR03433  \n",
      "2             NaN             NaN  \n",
      "3             NaN             NaN  \n",
      "4             NaN       TIGR03433  \n",
      "...           ...             ...  \n",
      "13280         NaN             NaN  \n",
      "13290         NaN             NaN  \n",
      "13318         NaN             NaN  \n",
      "13331     cd00291             NaN  \n",
      "13417     cd00371       TIGR00003  \n",
      "\n",
      "[5232 rows x 8 columns]\n"
     ]
    }
   ],
   "source": [
    "#map to arcog\n",
    "header = '<arCOG-id> <functional-class> <gene-name> <arCOG-annotation> <supercluster> <profile-pfam> <profile-cdd> <profile-tigrfam>'\n",
    "header = header.replace('<', '').replace('>' , '').split()\n",
    "\n",
    "arcogDF = pd.read_table( 'ar14.arCOGdef.tab'  ,engine='python' )\n",
    "arcogDF.columns = header\n",
    "print(len(arcogDF))\n",
    "arcogDF = arcogDF[~arcogDF['profile-pfam'].isna()]\n",
    "print(len(arcogDF))\n",
    "print(arcogDF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "arcogmap = dict(zip( arcogDF['profile-pfam'] , arcogDF['arCOG-id']))\n",
    "globaldf['arCOG-id'] = globaldf.accession.map(arcogmap)\n",
    "globaldf = globaldf.merge( arcogDF , left_on = 'arCOG-id'  , right_on = 'arCOG-id' , how = 'left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           targetname  accession     queryid accession2       E-value  score  \\\n",
      "0     Phage_integrase  pfam00589  ELY83596.1          -  7.800000e-18   64.9   \n",
      "1     Phage_int_SAM_4  pfam13495  ELY83596.1          -  2.000000e-11   44.3   \n",
      "2     Phage_int_SAM_1  pfam02899  ELY83596.1          -  6.900000e-05   23.1   \n",
      "3     Phage_int_SAM_5  pfam13102  ELY83596.1          -  8.800000e-04   19.7   \n",
      "4          OrfB_IS605  pfam01385  ELY83598.1          -  8.900000e-31  106.7   \n",
      "...               ...        ...         ...        ...           ...    ...   \n",
      "2144            PQQ_2  pfam13360  QLG62040.1          -  1.200000e-48  166.0   \n",
      "2145           HalOD1  pfam18545  QLG62045.1          -  7.000000e-10   39.1   \n",
      "2146           TrkA_C  pfam02080  QLG62049.1          -  2.900000e-11   43.2   \n",
      "2147            DUF87  pfam01935  QLG62070.1          -  3.400000e-06   27.4   \n",
      "2148     FtsK_SpoIIIE  pfam01580  QLG62070.1          -  7.400000e-04   19.0   \n",
      "\n",
      "      bias domain-E-value domain-score domain-bias  ...  \\\n",
      "0      0.0        1.4e-17         64.1         0.0  ...   \n",
      "1      0.1        4.2e-11         43.3         0.1  ...   \n",
      "2      0.1        0.00014         22.1         0.1  ...   \n",
      "3      0.3         0.0024         18.3         0.1  ...   \n",
      "4      1.8        8.9e-31        106.7         1.8  ...   \n",
      "...    ...            ...          ...         ...  ...   \n",
      "2144  19.5        1.4e-19         70.8         4.3  ...   \n",
      "2145   0.8        8.1e-10         38.9         0.8  ...   \n",
      "2146   0.0        5.7e-11         42.2         0.0  ...   \n",
      "2147   0.0        2.3e-05         24.7         0.0  ...   \n",
      "2148   0.0         0.0024         17.3         0.0  ...   \n",
      "\n",
      "                                        infile domain    arCOG-id  \\\n",
      "0          ./Naltunense1.gb.fastahmmerscan.csv    589  arCOG08889   \n",
      "1          ./Naltunense1.gb.fastahmmerscan.csv  13495         NaN   \n",
      "2          ./Naltunense1.gb.fastahmmerscan.csv   2899         NaN   \n",
      "3          ./Naltunense1.gb.fastahmmerscan.csv  13102         NaN   \n",
      "4          ./Naltunense1.gb.fastahmmerscan.csv   1385         NaN   \n",
      "...                                        ...    ...         ...   \n",
      "2144  ./halobonum.gb.proteo.fastahmmerscan.csv  13360  arCOG05863   \n",
      "2145  ./halobonum.gb.proteo.fastahmmerscan.csv  18545         NaN   \n",
      "2146  ./halobonum.gb.proteo.fastahmmerscan.csv   2080  arCOG05758   \n",
      "2147  ./halobonum.gb.proteo.fastahmmerscan.csv   1935  arCOG00283   \n",
      "2148  ./halobonum.gb.proteo.fastahmmerscan.csv   1580         NaN   \n",
      "\n",
      "     functional-class gene-name  \\\n",
      "0                   X      XerC   \n",
      "1                 NaN       NaN   \n",
      "2                 NaN       NaN   \n",
      "3                 NaN       NaN   \n",
      "4                 NaN       NaN   \n",
      "...               ...       ...   \n",
      "2144                M         -   \n",
      "2145              NaN       NaN   \n",
      "2146                P         -   \n",
      "2147                L         -   \n",
      "2148              NaN       NaN   \n",
      "\n",
      "                                       arCOG-annotation supercluster  \\\n",
      "0                            XerD/XerC family integrase     COG00582   \n",
      "1                                                   NaN          NaN   \n",
      "2                                                   NaN          NaN   \n",
      "3                                                   NaN          NaN   \n",
      "4                                                   NaN          NaN   \n",
      "...                                                 ...          ...   \n",
      "2144  WD40/PQQ-like beta propeller repeat containing...     COG01520   \n",
      "2145                                                NaN          NaN   \n",
      "2146                    TrkA-C domain contaning protein     COG00490   \n",
      "2147                                      HerA helicase     COG00433   \n",
      "2148                                                NaN          NaN   \n",
      "\n",
      "     profile-pfam profile-cdd profile-tigrfam  \n",
      "0       pfam00589     cd00801       TIGR02225  \n",
      "1             NaN         NaN             NaN  \n",
      "2             NaN         NaN             NaN  \n",
      "3             NaN         NaN             NaN  \n",
      "4             NaN         NaN             NaN  \n",
      "...           ...         ...             ...  \n",
      "2144    pfam13360     cd10276       TIGR03300  \n",
      "2145          NaN         NaN             NaN  \n",
      "2146    pfam02080         NaN             NaN  \n",
      "2147    pfam01935         NaN       TIGR02525  \n",
      "2148          NaN         NaN             NaN  \n",
      "\n",
      "[2149 rows x 29 columns]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(globaldf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           targetname  accession     queryid accession2       E-value  score  \\\n",
      "0     Phage_integrase  pfam00589  ELY83596.1          -  7.800000e-18   64.9   \n",
      "1     Phage_int_SAM_4  pfam13495  ELY83596.1          -  2.000000e-11   44.3   \n",
      "2     Phage_int_SAM_1  pfam02899  ELY83596.1          -  6.900000e-05   23.1   \n",
      "3     Phage_int_SAM_5  pfam13102  ELY83596.1          -  8.800000e-04   19.7   \n",
      "4          OrfB_IS605  pfam01385  ELY83598.1          -  8.900000e-31  106.7   \n",
      "...               ...        ...         ...        ...           ...    ...   \n",
      "2144            PQQ_2  pfam13360  QLG62040.1          -  1.200000e-48  166.0   \n",
      "2145           HalOD1  pfam18545  QLG62045.1          -  7.000000e-10   39.1   \n",
      "2146           TrkA_C  pfam02080  QLG62049.1          -  2.900000e-11   43.2   \n",
      "2147            DUF87  pfam01935  QLG62070.1          -  3.400000e-06   27.4   \n",
      "2148     FtsK_SpoIIIE  pfam01580  QLG62070.1          -  7.400000e-04   19.0   \n",
      "\n",
      "      bias domain-E-value domain-score domain-bias  ...  \\\n",
      "0      0.0        1.4e-17         64.1         0.0  ...   \n",
      "1      0.1        4.2e-11         43.3         0.1  ...   \n",
      "2      0.1        0.00014         22.1         0.1  ...   \n",
      "3      0.3         0.0024         18.3         0.1  ...   \n",
      "4      1.8        8.9e-31        106.7         1.8  ...   \n",
      "...    ...            ...          ...         ...  ...   \n",
      "2144  19.5        1.4e-19         70.8         4.3  ...   \n",
      "2145   0.8        8.1e-10         38.9         0.8  ...   \n",
      "2146   0.0        5.7e-11         42.2         0.0  ...   \n",
      "2147   0.0        2.3e-05         24.7         0.0  ...   \n",
      "2148   0.0         0.0024         17.3         0.0  ...   \n",
      "\n",
      "                                        infile domain    arCOG-id  \\\n",
      "0          ./Naltunense1.gb.fastahmmerscan.csv    589  arCOG08889   \n",
      "1          ./Naltunense1.gb.fastahmmerscan.csv  13495         NaN   \n",
      "2          ./Naltunense1.gb.fastahmmerscan.csv   2899         NaN   \n",
      "3          ./Naltunense1.gb.fastahmmerscan.csv  13102         NaN   \n",
      "4          ./Naltunense1.gb.fastahmmerscan.csv   1385         NaN   \n",
      "...                                        ...    ...         ...   \n",
      "2144  ./halobonum.gb.proteo.fastahmmerscan.csv  13360  arCOG05863   \n",
      "2145  ./halobonum.gb.proteo.fastahmmerscan.csv  18545         NaN   \n",
      "2146  ./halobonum.gb.proteo.fastahmmerscan.csv   2080  arCOG05758   \n",
      "2147  ./halobonum.gb.proteo.fastahmmerscan.csv   1935  arCOG00283   \n",
      "2148  ./halobonum.gb.proteo.fastahmmerscan.csv   1580         NaN   \n",
      "\n",
      "     functional-class gene-name  \\\n",
      "0                   X      XerC   \n",
      "1                 NaN       NaN   \n",
      "2                 NaN       NaN   \n",
      "3                 NaN       NaN   \n",
      "4                 NaN       NaN   \n",
      "...               ...       ...   \n",
      "2144                M         -   \n",
      "2145              NaN       NaN   \n",
      "2146                P         -   \n",
      "2147                L         -   \n",
      "2148              NaN       NaN   \n",
      "\n",
      "                                       arCOG-annotation supercluster  \\\n",
      "0                            XerD/XerC family integrase     COG00582   \n",
      "1                                                   NaN          NaN   \n",
      "2                                                   NaN          NaN   \n",
      "3                                                   NaN          NaN   \n",
      "4                                                   NaN          NaN   \n",
      "...                                                 ...          ...   \n",
      "2144  WD40/PQQ-like beta propeller repeat containing...     COG01520   \n",
      "2145                                                NaN          NaN   \n",
      "2146                    TrkA-C domain contaning protein     COG00490   \n",
      "2147                                      HerA helicase     COG00433   \n",
      "2148                                                NaN          NaN   \n",
      "\n",
      "     profile-pfam profile-cdd profile-tigrfam  \n",
      "0       pfam00589     cd00801       TIGR02225  \n",
      "1             NaN         NaN             NaN  \n",
      "2             NaN         NaN             NaN  \n",
      "3             NaN         NaN             NaN  \n",
      "4             NaN         NaN             NaN  \n",
      "...           ...         ...             ...  \n",
      "2144    pfam13360     cd10276       TIGR03300  \n",
      "2145          NaN         NaN             NaN  \n",
      "2146    pfam02080         NaN             NaN  \n",
      "2147    pfam01935         NaN       TIGR02525  \n",
      "2148          NaN         NaN             NaN  \n",
      "\n",
      "[2149 rows x 29 columns]\n"
     ]
    }
   ],
   "source": [
    "print(globaldf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       genome_id                  genome_name     accession  \\\n",
      "patric_id                                                                     \n",
      "fig|926690.3.peg.2651   926690.3  Haloplanus natans DSM 17983  ATYM01000002   \n",
      "fig|926690.3.peg.2652   926690.3  Haloplanus natans DSM 17983  ATYM01000002   \n",
      "fig|926690.3.peg.2653   926690.3  Haloplanus natans DSM 17983  ATYM01000002   \n",
      "fig|926690.3.peg.2654   926690.3  Haloplanus natans DSM 17983  ATYM01000002   \n",
      "fig|926690.3.peg.2655   926690.3  Haloplanus natans DSM 17983  ATYM01000002   \n",
      "...                          ...                          ...           ...   \n",
      "fig|926690.3.peg.3332   926690.3  Haloplanus natans DSM 17983  ATYM01000002   \n",
      "fig|926690.3.peg.3333   926690.3  Haloplanus natans DSM 17983  ATYM01000002   \n",
      "fig|926690.3.peg.3334   926690.3  Haloplanus natans DSM 17983  ATYM01000002   \n",
      "fig|926690.3.peg.3335   926690.3  Haloplanus natans DSM 17983  ATYM01000002   \n",
      "fig|926690.3.rna.50     926690.3  Haloplanus natans DSM 17983  ATYM01000002   \n",
      "\n",
      "                      annotation feature_type              patric_id  \\\n",
      "patric_id                                                              \n",
      "fig|926690.3.peg.2651     PATRIC          CDS  fig|926690.3.peg.2651   \n",
      "fig|926690.3.peg.2652     PATRIC          CDS  fig|926690.3.peg.2652   \n",
      "fig|926690.3.peg.2653     PATRIC          CDS  fig|926690.3.peg.2653   \n",
      "fig|926690.3.peg.2654     PATRIC          CDS  fig|926690.3.peg.2654   \n",
      "fig|926690.3.peg.2655     PATRIC          CDS  fig|926690.3.peg.2655   \n",
      "...                          ...          ...                    ...   \n",
      "fig|926690.3.peg.3332     PATRIC          CDS  fig|926690.3.peg.3332   \n",
      "fig|926690.3.peg.3333     PATRIC          CDS  fig|926690.3.peg.3333   \n",
      "fig|926690.3.peg.3334     PATRIC          CDS  fig|926690.3.peg.3334   \n",
      "fig|926690.3.peg.3335     PATRIC          CDS  fig|926690.3.peg.3335   \n",
      "fig|926690.3.rna.50       PATRIC         rRNA    fig|926690.3.rna.50   \n",
      "\n",
      "                       refseq_locus_tag    start      end strand  na_length  \\\n",
      "patric_id                                                                     \n",
      "fig|926690.3.peg.2651               NaN  1496172  1497365      -       1194   \n",
      "fig|926690.3.peg.2652               NaN  1498710  1498907      -        198   \n",
      "fig|926690.3.peg.2653               NaN  1500237  1501934      -       1698   \n",
      "fig|926690.3.peg.2654               NaN  1503001  1503423      +        423   \n",
      "fig|926690.3.peg.2655               NaN  1503499  1505130      -       1632   \n",
      "...                                 ...      ...      ...    ...        ...   \n",
      "fig|926690.3.peg.3332               NaN  2165873  2166127      +        255   \n",
      "fig|926690.3.peg.3333               NaN  2166128  2167183      -       1056   \n",
      "fig|926690.3.peg.3334               NaN  2167248  2167445      -        198   \n",
      "fig|926690.3.peg.3335               NaN  2167649  2167960      +        312   \n",
      "fig|926690.3.rna.50                 NaN  2168757  2169903      +       1147   \n",
      "\n",
      "                       gene  \\\n",
      "patric_id                     \n",
      "fig|926690.3.peg.2651   NaN   \n",
      "fig|926690.3.peg.2652   NaN   \n",
      "fig|926690.3.peg.2653   NaN   \n",
      "fig|926690.3.peg.2654   NaN   \n",
      "fig|926690.3.peg.2655   NaN   \n",
      "...                     ...   \n",
      "fig|926690.3.peg.3332   NaN   \n",
      "fig|926690.3.peg.3333   NaN   \n",
      "fig|926690.3.peg.3334   NaN   \n",
      "fig|926690.3.peg.3335   NaN   \n",
      "fig|926690.3.rna.50     NaN   \n",
      "\n",
      "                                                                 product  \\\n",
      "patric_id                                                                  \n",
      "fig|926690.3.peg.2651                               hypothetical protein   \n",
      "fig|926690.3.peg.2652                               hypothetical protein   \n",
      "fig|926690.3.peg.2653                               hypothetical protein   \n",
      "fig|926690.3.peg.2654     hypothetical protein; probable calcium binding   \n",
      "fig|926690.3.peg.2655                               hypothetical protein   \n",
      "...                                                                  ...   \n",
      "fig|926690.3.peg.3332                               hypothetical protein   \n",
      "fig|926690.3.peg.3333  Dihydroorotate dehydrogenase (quinone) (EC 1.3...   \n",
      "fig|926690.3.peg.3334                               hypothetical protein   \n",
      "fig|926690.3.peg.3335                            Chromosomal protein MC1   \n",
      "fig|926690.3.rna.50        Small Subunit Ribosomal RNA; ssuRNA; SSU rRNA   \n",
      "\n",
      "                       plfam_id      pgfam_id             protein_id  \n",
      "patric_id                                                             \n",
      "fig|926690.3.peg.2651       NaN           NaN  fig|926690.3.peg.2651  \n",
      "fig|926690.3.peg.2652       NaN           NaN  fig|926690.3.peg.2652  \n",
      "fig|926690.3.peg.2653       NaN  PGF_06091862  fig|926690.3.peg.2653  \n",
      "fig|926690.3.peg.2654       NaN           NaN  fig|926690.3.peg.2654  \n",
      "fig|926690.3.peg.2655       NaN           NaN  fig|926690.3.peg.2655  \n",
      "...                         ...           ...                    ...  \n",
      "fig|926690.3.peg.3332       NaN           NaN  fig|926690.3.peg.3332  \n",
      "fig|926690.3.peg.3333       NaN  PGF_01033770  fig|926690.3.peg.3333  \n",
      "fig|926690.3.peg.3334       NaN  PGF_00303083  fig|926690.3.peg.3334  \n",
      "fig|926690.3.peg.3335       NaN  PGF_00417877  fig|926690.3.peg.3335  \n",
      "fig|926690.3.rna.50         NaN           NaN    fig|926690.3.rna.50  \n",
      "\n",
      "[703 rows x 16 columns]\n"
     ]
    }
   ],
   "source": [
    "print(featureDfs['hnatans'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nal.gb\n",
      "         protein_id  start   stop  \\\n",
      "0  JNCS01000001.1_1    708   1748   \n",
      "1  JNCS01000001.1_2   3911   4552   \n",
      "2  JNCS01000001.1_3   4354   5184   \n",
      "3  JNCS01000001.1_4   8492   9190   \n",
      "4  JNCS01000001.1_5  12484  13356   \n",
      "\n",
      "                                       genome xrefs targetname  accession  \\\n",
      "0  [BioProject:PRJNA248700, BioSample:SAMN02800870]    Mrr_cat  pfam04471   \n",
      "1  [BioProject:PRJNA248700, BioSample:SAMN02800870]        NaN        NaN   \n",
      "2  [BioProject:PRJNA248700, BioSample:SAMN02800870]        NaN        NaN   \n",
      "3  [BioProject:PRJNA248700, BioSample:SAMN02800870]        NaN        NaN   \n",
      "4  [BioProject:PRJNA248700, BioSample:SAMN02800870]        NaN        NaN   \n",
      "\n",
      "            queryid   E-value  clu   description domain    arCOG-id  \\\n",
      "0  JNCS01000001.1_1  0.000001    0  Restriction    4471  arCOG09571   \n",
      "1  JNCS01000001.1_2       NaN  NaN           NaN    NaN         NaN   \n",
      "2  JNCS01000001.1_3       NaN  NaN           NaN    NaN         NaN   \n",
      "3  JNCS01000001.1_4       NaN  NaN           NaN    NaN         NaN   \n",
      "4  JNCS01000001.1_5       NaN  NaN           NaN    NaN         NaN   \n",
      "\n",
      "  functional-class gene-name                      arCOG-annotation  \\\n",
      "0                V         -  Restriction endonuclease, Mrr family   \n",
      "1              NaN       NaN                                   NaN   \n",
      "2              NaN       NaN                                   NaN   \n",
      "3              NaN       NaN                                   NaN   \n",
      "4              NaN       NaN                                   NaN   \n",
      "\n",
      "  supercluster profile-cdd profile-tigrfam  \n",
      "0          NaN         NaN             NaN  \n",
      "1          NaN         NaN             NaN  \n",
      "2          NaN         NaN             NaN  \n",
      "3          NaN         NaN             NaN  \n",
      "4          NaN         NaN             NaN  \n",
      "Naltunense1.gb\n",
      "    locus_tag codon_start                         product  protein_id  \\\n",
      "0  C485_17617           1            hypothetical protein  ELY83594.1   \n",
      "1  C485_17622           1            hypothetical protein  ELY83595.1   \n",
      "2  C485_17627           1  site-specific recombinase xerd  ELY83596.1   \n",
      "3  C485_17627           1  site-specific recombinase xerd  ELY83596.1   \n",
      "4  C485_17627           1  site-specific recombinase xerd  ELY83596.1   \n",
      "\n",
      "                                       genome xrefs       targetname  \\\n",
      "0  [BioProject:PRJNA174934, BioSample:SAMN02471672]              NaN   \n",
      "1  [BioProject:PRJNA174934, BioSample:SAMN02471672]              NaN   \n",
      "2  [BioProject:PRJNA174934, BioSample:SAMN02471672]  Phage_integrase   \n",
      "3  [BioProject:PRJNA174934, BioSample:SAMN02471672]  Phage_int_SAM_4   \n",
      "4  [BioProject:PRJNA174934, BioSample:SAMN02471672]  Phage_int_SAM_1   \n",
      "\n",
      "   accession     queryid       E-value  clu  \\\n",
      "0        NaN  ELY83594.1           NaN  NaN   \n",
      "1        NaN  ELY83595.1           NaN  NaN   \n",
      "2  pfam00589  ELY83596.1  7.800000e-18    0   \n",
      "3  pfam13495  ELY83596.1  2.000000e-11    0   \n",
      "4  pfam02899  ELY83596.1  6.900000e-05    0   \n",
      "\n",
      "                             description domain    arCOG-id functional-class  \\\n",
      "0                                    NaN    NaN         NaN              NaN   \n",
      "1                                    NaN    NaN         NaN              NaN   \n",
      "2                       Phage integrase     589  arCOG08889                X   \n",
      "3  Phage integrase, N-terminal SAM-like   13495         NaN              NaN   \n",
      "4  Phage integrase, N-terminal SAM-like    2899         NaN              NaN   \n",
      "\n",
      "  gene-name            arCOG-annotation supercluster profile-cdd  \\\n",
      "0       NaN                         NaN          NaN         NaN   \n",
      "1       NaN                         NaN          NaN         NaN   \n",
      "2      XerC  XerD/XerC family integrase     COG00582     cd00801   \n",
      "3       NaN                         NaN          NaN         NaN   \n",
      "4       NaN                         NaN          NaN         NaN   \n",
      "\n",
      "  profile-tigrfam  \n",
      "0             NaN  \n",
      "1             NaN  \n",
      "2       TIGR02225  \n",
      "3             NaN  \n",
      "4             NaN  \n",
      "halovivax.gb\n",
      "       locus_tag codon_start               product      protein_id  \\\n",
      "0  J1N62_RS14465           1      HNH endonuclease  WP_207587062.1   \n",
      "1  J1N62_RS14465           1      HNH endonuclease  WP_207587062.1   \n",
      "2  J1N62_RS14465           1      HNH endonuclease  WP_207587062.1   \n",
      "3  J1N62_RS14465           1      HNH endonuclease  WP_207587062.1   \n",
      "4  J1N62_RS14470           1  hypothetical protein  WP_207587064.1   \n",
      "\n",
      "           db_xref                                       genome xrefs  \\\n",
      "0  GeneID:63213810  [BioProject:PRJNA224116, BioSample:SAMN1820337...   \n",
      "1  GeneID:63213810  [BioProject:PRJNA224116, BioSample:SAMN1820337...   \n",
      "2  GeneID:63213810  [BioProject:PRJNA224116, BioSample:SAMN1820337...   \n",
      "3  GeneID:63213810  [BioProject:PRJNA224116, BioSample:SAMN1820337...   \n",
      "4  GeneID:63213811  [BioProject:PRJNA224116, BioSample:SAMN1820337...   \n",
      "\n",
      "  EC_number targetname  accession         queryid  ...  clu description  \\\n",
      "0       NaN        HNH  pfam01844  WP_207587062.1  ...    0        HNH    \n",
      "1       NaN      HNH_5  pfam14279  WP_207587062.1  ...    0        HNH    \n",
      "2       NaN        HNH  pfam01844  WP_207587062.1  ...    0        HNH    \n",
      "3       NaN      HNH_5  pfam14279  WP_207587062.1  ...    0        HNH    \n",
      "4       NaN        NaN        NaN  WP_207587064.1  ...  NaN         NaN   \n",
      "\n",
      "  domain    arCOG-id functional-class gene-name  \\\n",
      "0   1844  arCOG10446                V         -   \n",
      "1  14279         NaN              NaN       NaN   \n",
      "2   1844  arCOG10446                V         -   \n",
      "3  14279         NaN              NaN       NaN   \n",
      "4    NaN         NaN              NaN       NaN   \n",
      "\n",
      "                            arCOG-annotation supercluster profile-cdd  \\\n",
      "0  Restriction endonuclease, McrA/HNH family     COG01403     cd00085   \n",
      "1                                        NaN          NaN         NaN   \n",
      "2  Restriction endonuclease, McrA/HNH family     COG01403     cd00085   \n",
      "3                                        NaN          NaN         NaN   \n",
      "4                                        NaN          NaN         NaN   \n",
      "\n",
      "  profile-tigrfam  \n",
      "0             NaN  \n",
      "1             NaN  \n",
      "2             NaN  \n",
      "3             NaN  \n",
      "4             NaN  \n",
      "\n",
      "[5 rows x 21 columns]\n",
      "halobonum.gb\n",
      "     locus_tag codon_start                            product  protein_id  \\\n",
      "0  HUG12_09255           3  DUF2182 domain-containing protein  QLG61896.1   \n",
      "1  HUG12_09255           3  DUF2182 domain-containing protein  QLG61896.1   \n",
      "2  HUG12_09260           1               hypothetical protein  QLG61897.1   \n",
      "3  HUG12_09265           1             IS5 family transposase  QLG61898.1   \n",
      "4  HUG12_09265           1             IS5 family transposase  QLG61898.1   \n",
      "\n",
      "                                       genome xrefs EC_number gene  \\\n",
      "0  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN   \n",
      "1  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN   \n",
      "2  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN   \n",
      "3  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN   \n",
      "4  [BioProject:PRJNA638230, BioSample:SAMN15183787]       NaN  NaN   \n",
      "\n",
      "    targetname  accession     queryid  ...  clu  \\\n",
      "0      DUF2182  pfam09948  QLG61896.1  ...    0   \n",
      "1      DUF2182  pfam09948  QLG61896.1  ...    0   \n",
      "2          NaN        NaN  QLG61897.1  ...  NaN   \n",
      "3    DDE_Tnp_1  pfam01609  QLG61898.1  ...    1   \n",
      "4  DDE_Tnp_1_3  pfam13612  QLG61898.1  ...    0   \n",
      "\n",
      "                                         description domain    arCOG-id  \\\n",
      "0  Predicted metal-binding integral membrane prot...   9948  arCOG10678   \n",
      "1  Predicted metal-binding integral membrane prot...   9948  arCOG10678   \n",
      "2                                                NaN    NaN         NaN   \n",
      "3                                   Transposase DDE    1609  arCOG09146   \n",
      "4                                   Transposase DDE   13612         NaN   \n",
      "\n",
      "  functional-class gene-name                arCOG-annotation supercluster  \\\n",
      "0                S         -  Metal-binding membrane protein     COG05486   \n",
      "1                S         -  Metal-binding membrane protein     COG05486   \n",
      "2              NaN       NaN                             NaN          NaN   \n",
      "3                X         -                 IS4 transposase     COG03385   \n",
      "4              NaN       NaN                             NaN          NaN   \n",
      "\n",
      "  profile-cdd profile-tigrfam  \n",
      "0         NaN             NaN  \n",
      "1         NaN             NaN  \n",
      "2         NaN             NaN  \n",
      "3         NaN             NaN  \n",
      "4         NaN             NaN  \n",
      "\n",
      "[5 rows x 21 columns]\n",
      "haloferax.gb\n",
      "         protein_id start  stop  \\\n",
      "0  LOEP01000011.1_1   497  1303   \n",
      "1  LOEP01000011.1_1   497  1303   \n",
      "2  LOEP01000011.1_1   497  1303   \n",
      "3  LOEP01000011.1_1   497  1303   \n",
      "4  LOEP01000011.1_2  3372  4208   \n",
      "\n",
      "                                       genome xrefs    targetname  accession  \\\n",
      "0  [BioProject:PRJNA294240, BioSample:SAMN04018175]  adh_short_C2  pfam13561   \n",
      "1  [BioProject:PRJNA294240, BioSample:SAMN04018175]     adh_short  pfam00106   \n",
      "2  [BioProject:PRJNA294240, BioSample:SAMN04018175]            KR  pfam08659   \n",
      "3  [BioProject:PRJNA294240, BioSample:SAMN04018175]     Epimerase  pfam01370   \n",
      "4  [BioProject:PRJNA294240, BioSample:SAMN04018175]  CTP_transf_3  pfam02348   \n",
      "\n",
      "            queryid       E-value clu                           description  \\\n",
      "0  LOEP01000011.1_1  2.400000e-65   0         Enoyl-(Acyl carrier protein)    \n",
      "1  LOEP01000011.1_1  2.800000e-49   0                          short chain    \n",
      "2  LOEP01000011.1_1  3.200000e-11   0                                   KR    \n",
      "3  LOEP01000011.1_1  1.300000e-05   1  NAD dependent epimerase/dehydratase    \n",
      "4  LOEP01000011.1_2  5.700000e-45   0                                         \n",
      "\n",
      "  domain    arCOG-id functional-class gene-name  \\\n",
      "0  13561  arCOG01262                I      FabG   \n",
      "1    106  arCOG04552                I      FabG   \n",
      "2   8659         NaN              NaN       NaN   \n",
      "3   1370  arCOG04762                M     Agl3    \n",
      "4   2348  arCOG04817                M      NeuA   \n",
      "\n",
      "                         arCOG-annotation supercluster profile-cdd  \\\n",
      "0       Short-chain alcohol dehydrogenase     COG01028     cd05233   \n",
      "1       Short-chain alcohol dehydrogenase     COG01028     cd05233   \n",
      "2                                     NaN          NaN         NaN   \n",
      "3             UDP-sulfoquinovose synthase     COG00451     cd05255   \n",
      "4  CMP-N-acetylneuraminic acid synthetase     COG01083     cd02518   \n",
      "\n",
      "  profile-tigrfam  \n",
      "0       TIGR01830  \n",
      "1       TIGR01830  \n",
      "2             NaN  \n",
      "3       TIGR01179  \n",
      "4       TIGR00466  \n",
      "halogeometricum.gb\n",
      "     locus_tag codon_start                            product  protein_id  \\\n",
      "0  G3I44_14265           2               hypothetical protein  QIB75350.1   \n",
      "1  G3I44_14270           1               hypothetical protein  QIB75351.1   \n",
      "2  G3I44_14275           1  DUF4258 domain-containing protein  QIB75352.1   \n",
      "3  G3I44_14275           1  DUF4258 domain-containing protein  QIB75352.1   \n",
      "4  G3I44_14280           1               hypothetical protein  QIB75353.1   \n",
      "\n",
      "                                       genome xrefs targetname  accession  \\\n",
      "0  [BioProject:PRJNA605299, BioSample:SAMN14053718]        NaN        NaN   \n",
      "1  [BioProject:PRJNA605299, BioSample:SAMN14053718]        NaN        NaN   \n",
      "2  [BioProject:PRJNA605299, BioSample:SAMN14053718]    DUF4258  pfam14076   \n",
      "3  [BioProject:PRJNA605299, BioSample:SAMN14053718]    DUF4258  pfam14076   \n",
      "4  [BioProject:PRJNA605299, BioSample:SAMN14053718]        NaN        NaN   \n",
      "\n",
      "      queryid   E-value  clu                  description domain    arCOG-id  \\\n",
      "0  QIB75350.1       NaN  NaN                          NaN    NaN         NaN   \n",
      "1  QIB75351.1       NaN  NaN                          NaN    NaN         NaN   \n",
      "2  QIB75352.1  0.000033    0  Domain of unknown function   14076  arCOG13027   \n",
      "3  QIB75352.1  0.000033    0  Domain of unknown function   14076  arCOG13027   \n",
      "4  QIB75353.1       NaN  NaN                          NaN    NaN         NaN   \n",
      "\n",
      "  functional-class gene-name         arCOG-annotation supercluster  \\\n",
      "0              NaN       NaN                      NaN          NaN   \n",
      "1              NaN       NaN                      NaN          NaN   \n",
      "2                S         -  Uncharacterized protein          NaN   \n",
      "3                S         -  Uncharacterized protein          NaN   \n",
      "4              NaN       NaN                      NaN          NaN   \n",
      "\n",
      "  profile-cdd profile-tigrfam  \n",
      "0         NaN             NaN  \n",
      "1         NaN             NaN  \n",
      "2         NaN             NaN  \n",
      "3         NaN             NaN  \n",
      "4         NaN             NaN  \n",
      "haloterigena_reg.gb\n",
      "     locus_tag codon_start                                            product  \\\n",
      "0  HT576_08605           2  twin-arginine translocase TatA/TatE family sub...   \n",
      "1  HT576_08605           2  twin-arginine translocase TatA/TatE family sub...   \n",
      "2  HT576_08615           1                            site-specific integrase   \n",
      "3  HT576_08615           1                            site-specific integrase   \n",
      "4  HT576_08615           1                            site-specific integrase   \n",
      "\n",
      "   protein_id                                      genome xrefs EC_number  \\\n",
      "0  NUB91080.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN   \n",
      "1  NUB91080.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN   \n",
      "2  NUB91081.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN   \n",
      "3  NUB91081.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN   \n",
      "4  NUB91081.1  [BioProject:PRJNA637493, BioSample:SAMN15143775]       NaN   \n",
      "\n",
      "        targetname  accession     queryid       E-value clu  \\\n",
      "0         TatA_B_E  pfam02416  NUB91080.1  1.900000e-14   1   \n",
      "1         TatA_B_E  pfam02416  NUB91080.1  1.900000e-14   1   \n",
      "2  Phage_int_SAM_1  pfam02899  NUB91081.1  3.100000e-07   1   \n",
      "3  Phage_int_SAM_5  pfam13102  NUB91081.1  5.700000e-05   0   \n",
      "4  Phage_int_SAM_1  pfam02899  NUB91081.1  3.100000e-07   1   \n",
      "\n",
      "                             description domain    arCOG-id functional-class  \\\n",
      "0                           mttA/Hcf106    2416  arCOG02694                U   \n",
      "1                           mttA/Hcf106    2416  arCOG02694                U   \n",
      "2  Phage integrase, N-terminal SAM-like    2899         NaN              NaN   \n",
      "3              Phage integrase SAM-like   13102         NaN              NaN   \n",
      "4  Phage integrase, N-terminal SAM-like    2899         NaN              NaN   \n",
      "\n",
      "  gene-name                                   arCOG-annotation supercluster  \\\n",
      "0      TatA  Sec-independent protein secretion pathway comp...     COG01826   \n",
      "1      TatA  Sec-independent protein secretion pathway comp...     COG01826   \n",
      "2       NaN                                                NaN          NaN   \n",
      "3       NaN                                                NaN          NaN   \n",
      "4       NaN                                                NaN          NaN   \n",
      "\n",
      "  profile-cdd profile-tigrfam  \n",
      "0     cd01915       TIGR01411  \n",
      "1     cd01915       TIGR01411  \n",
      "2         NaN             NaN  \n",
      "3         NaN             NaN  \n",
      "4         NaN             NaN  \n",
      "hnatans\n",
      "   genome_id                  genome_name   accession_x annotation  \\\n",
      "0   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "1   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "2   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "3   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "4   926690.3  Haloplanus natans DSM 17983  ATYM01000002     PATRIC   \n",
      "\n",
      "  feature_type              patric_id  refseq_locus_tag    start      end  \\\n",
      "0          CDS  fig|926690.3.peg.2651               NaN  1496172  1497365   \n",
      "1          CDS  fig|926690.3.peg.2652               NaN  1498710  1498907   \n",
      "2          CDS  fig|926690.3.peg.2653               NaN  1500237  1501934   \n",
      "3          CDS  fig|926690.3.peg.2654               NaN  1503001  1503423   \n",
      "4          CDS  fig|926690.3.peg.2655               NaN  1503499  1505130   \n",
      "\n",
      "  strand  ...  clu        description domain    arCOG-id functional-class  \\\n",
      "0      -  ...  NaN                NaN    NaN         NaN              NaN   \n",
      "1      -  ...  NaN                NaN    NaN         NaN              NaN   \n",
      "2      -  ...  NaN                NaN    NaN         NaN              NaN   \n",
      "3      +  ...  NaN                NaN    NaN         NaN              NaN   \n",
      "4      -  ...    1  Peptidase family    1435  arCOG07513                O   \n",
      "\n",
      "  gene-name                               arCOG-annotation supercluster  \\\n",
      "0       NaN                                            NaN          NaN   \n",
      "1       NaN                                            NaN          NaN   \n",
      "2       NaN                                            NaN          NaN   \n",
      "3       NaN                                            NaN          NaN   \n",
      "4      HtpX  Zn-dependent protease with chaperone function     COG00501   \n",
      "\n",
      "  profile-cdd  profile-tigrfam  \n",
      "0         NaN              NaN  \n",
      "1         NaN              NaN  \n",
      "2         NaN              NaN  \n",
      "3         NaN              NaN  \n",
      "4         NaN              NaN  \n",
      "\n",
      "[5 rows x 30 columns]\n"
     ]
    }
   ],
   "source": [
    "for key in featureDfs:\n",
    "    print(key)\n",
    "    if len( featureDfs[key]) > 0:\n",
    "        featureDfs[key] = featureDfs[key].merge(globaldf, left_index = True , right_on = 'queryid' , how='left')\n",
    "        for col in ['transl_table' , 'translation' , 'infile' , 'inference' , 'note' , 'profile-pfam' , 'score',\n",
    "       'bias', 'domain-E-value', 'domain-score', 'domain-bias', 'exp', 'reg', 'EC_number'\n",
    "       'clu', 'ov', 'env', 'dom', 'rep', 'inc' , 'accession2' ]:\n",
    "            try:\n",
    "                featureDfs[key] = featureDfs[key].drop( axis = 1, labels =[col]  )\n",
    "            except:\n",
    "                pass\n",
    "        featureDfs[key].index = np.arange( len(featureDfs[key]))\n",
    "        \n",
    "        \n",
    "        featureDfs[key].to_csv( key + 'arcog_features.csv')\n",
    "        print(featureDfs[key].head())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
