{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e2d20075-d14b-4390-b1b9-8dda2cda0572",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries\n",
    "import os\n",
    "import subprocess\n",
    "import rpy2\n",
    "import pandas as pd\n",
    "from Bio import Entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "174ecf8d-5f62-46b8-a380-62951a2d6e19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading in R\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fbcfb173-46d4-4709-83a8-ccba6f55832a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: ── \u001b[1mAttaching packages\u001b[22m ───────────────────────────────────────────────────────────────────────────────────── tidyverse 1.3.1 ──\n",
      "\n",
      "R[write to console]: \u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.3     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.2     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.6\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.1.3     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 1.4.0     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.1\n",
      "\n",
      "R[write to console]: ── \u001b[1mConflicts\u001b[22m ──────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘magrittr’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    set_names\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    extract\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘glue’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1m\u001b[1mColumn specification\u001b[1m\u001b[22m \u001b[36m──────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  .default = col_character(),\n",
      "  `scaffold length` = \u001b[32mcol_double()\u001b[39m,\n",
      "  LENGTH = \u001b[32mcol_double()\u001b[39m,\n",
      "  `SIGNAL PEPTIDE?` = \u001b[32mcol_double()\u001b[39m,\n",
      "  TMs = \u001b[32mcol_double()\u001b[39m,\n",
      "  `ECTO LENGTH` = \u001b[32mcol_double()\u001b[39m,\n",
      "  `Num CYS` = \u001b[32mcol_double()\u001b[39m,\n",
      "  `ECTO Isoelectric point` = \u001b[32mcol_double()\u001b[39m,\n",
      "  SOLID = \u001b[32mcol_double()\u001b[39m,\n",
      "  AQUEOUS = \u001b[32mcol_double()\u001b[39m,\n",
      "  `SALT?` = \u001b[32mcol_double()\u001b[39m,\n",
      "  pH = \u001b[32mcol_double()\u001b[39m,\n",
      "  T_Classified = \u001b[32mcol_double()\u001b[39m,\n",
      "  `ALT_DEPT (mts)` = \u001b[32mcol_double()\u001b[39m,\n",
      "  O2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m for the full column specifications.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -o ecological_data_tibble\n",
    "\n",
    "# loading libraries\n",
    "library(tidyverse)\n",
    "library(magrittr)\n",
    "library(glue)\n",
    "\n",
    "# load ecological categorical data\n",
    "## define column names\n",
    "ecological_data_columns = c('HEADER','scaffold ID_ORF','scaffold ID_ORF (SPADES)','CONFIDENCE','METAGENOMICS PROJECT',\n",
    "                               'TAXA','scaffold length','COMPLETE SEQUENCE','LENGTH','SIGNAL PEPTIDE?','TMs','ECTODOMAIN',\n",
    "                               'ECTO LENGTH','C','Num CYS','ECTO Isoelectric point','COMMENTS','BIOSAMPLE','MG NAME',\n",
    "                               'HABITAT_Detailed','Temperature_Detailed','elev mts','collec DATE','HABITAT','AUTHORS',\n",
    "                               'CONTACT','PAPER DOI','ISOLATION','SOLID','AQUEOUS','SALT?','pH','T_Classified',\n",
    "                               'ALT_DEPT (mts)','FILTER FRACTION','O2')\n",
    "\n",
    "ecological_data.tibble = readr::read_tsv('../data/metadata/modified_FsxAs-Kosher-Taxo-26-Abr-2021_Taxon_and_Kosher.tsv',\n",
    "                                           col_names = ecological_data_columns, \n",
    "                                           skip = 1)\n",
    "\n",
    "# performiing a little change in a too-long-tip\n",
    "ecological_data.tibble %<>% dplyr::mutate(HEADER = HEADER %>% str_replace_all(., ' \\\\(new.*\\\\)', '')) %>%\n",
    "        tidyr::separate_rows(., `BIOSAMPLE`, sep = ' y ')\n",
    "        \n",
    "# create variable to export\n",
    "ecological_data_tibble = ecological_data.tibble"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15f5fc9e-1d96-4180-8c56-f16885b261ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "import collections\n",
    "\n",
    "# trying for one biosample\n",
    "Entrez.email = 'mauricio.langleib@gmail.com'\n",
    "\n",
    "# create empty list to allocate to allocate rows\n",
    "geo_data_rows = []\n",
    "\n",
    "# loop over BioSample IDs to get geographical data\n",
    "for biosample_id in ecological_data_tibble['BIOSAMPLE'].to_list():\n",
    "    try:\n",
    "        # retrieve data from BioSample\n",
    "        # first get UID for given BioSample ID\n",
    "        handle = Entrez.esearch(db = 'biosample', term = biosample_id)\n",
    "        biosample_search_dict = Entrez.read(handle)\n",
    "        biosample_uid = biosample_search_dict['IdList']\n",
    "        # now search data for that UID\n",
    "        handle = Entrez.efetch(db = 'biosample', id = biosample_uid, type = 'text')\n",
    "        #print(handle.read())\n",
    "        #biosample_data_dict = Entrez.read(handle)\n",
    "        import xmltodict\n",
    "        biosample_dict = xmltodict.parse(handle)\n",
    "        # getting relevant geographical data\n",
    "        # setting defaults to NA\n",
    "        geo_loc_name = 'NA'\n",
    "        lat = 'NA'\n",
    "        lon = 'NA'\n",
    "        isolation_source = 'NA'\n",
    "        # parsing data\n",
    "        # first case: only one organism annotated\n",
    "        if type(biosample_dict['BioSampleSet']['BioSample']) == collections.OrderedDict:\n",
    "            for attribute in biosample_dict['BioSampleSet']['BioSample']['Attributes']['Attribute']:\n",
    "                if attribute['@attribute_name'] == 'geo_loc_name':\n",
    "                    geo_loc_name = attribute['#text']\n",
    "                if attribute['@attribute_name'] == 'lat_lon':\n",
    "                    lat = attribute['#text'].split(' ')[0] + ' ' + attribute['#text'].split(' ')[1]\n",
    "                    lon = attribute['#text'].split(' ')[2] + ' ' + attribute['#text'].split(' ')[3]\n",
    "                if attribute['@attribute_name'] == 'isolation_source':\n",
    "                    isolation_source = attribute['#text']\n",
    "            if 'Description' in biosample_dict['BioSampleSet']['BioSample'].keys():\n",
    "                if 'Title' in biosample_dict['BioSampleSet']['BioSample']['Description'].keys():\n",
    "                    title = biosample_dict['BioSampleSet']['BioSample']['Description']['Title']\n",
    "                else:\n",
    "                    title = 'NA'\n",
    "                if 'Organism' in biosample_dict['BioSampleSet']['BioSample']['Description'].keys():\n",
    "                    if '@taxonomy_id' in biosample_dict['BioSampleSet']['BioSample']['Description']['Organism']:\n",
    "                        taxa_id = biosample_dict['BioSampleSet']['BioSample']['Description']['Organism']['@taxonomy_id']\n",
    "                    else:\n",
    "                        taxa_id = 'NA'\n",
    "                    if 'OrganismName' in biosample_dict['BioSampleSet']['BioSample']['Description']['Organism']:\n",
    "                        organism = biosample_dict['BioSampleSet']['BioSample']['Description']['Organism']['OrganismName']\n",
    "                    else:\n",
    "                        organism = 'NA'\n",
    "            else:\n",
    "                title = 'NA'\n",
    "                taxa_id = 'NA'\n",
    "                organism = 'NA'\n",
    "            # append pandas DataFrame with data to <geo_data_rows>\n",
    "            geo_data_rows.append(pd.DataFrame.from_dict({'BioSample ID': [biosample_id], \n",
    "                                                         'Geographical location': [geo_loc_name], \n",
    "                                                         'Latitude': [str(lat)], \n",
    "                                                         'Longitude': [str(lon)],\n",
    "                                                         'Isolation source': [isolation_source],\n",
    "                                                         'Title': [title],\n",
    "                                                         'Taxa ID': [taxa_id],\n",
    "                                                         'Organism': [organism]}))\n",
    "        elif type(biosample_dict['BioSampleSet']['BioSample']) == list:\n",
    "            for item in biosample_dict['BioSampleSet']['BioSample']:\n",
    "                for attribute in item['Attributes']['Attribute']:\n",
    "                    if attribute['@attribute_name'] == 'geo_loc_name':\n",
    "                        geo_loc_name = attribute['#text']\n",
    "                    if attribute['@attribute_name'] == 'lat_lon':\n",
    "                        lat = attribute['#text'].split(' ')[0] + ' ' + attribute['#text'].split(' ')[1]\n",
    "                        lon = attribute['#text'].split(' ')[2] + ' ' + attribute['#text'].split(' ')[3]\n",
    "                    if attribute['@attribute_name'] == 'isolation_source':\n",
    "                        isolation_source = attribute['#text']\n",
    "                if 'Description' in item.keys():\n",
    "                    if 'Title' in item['Description'].keys():\n",
    "                        title = item['Description']['Title']\n",
    "                    else:\n",
    "                        title = 'NA'\n",
    "                    if 'Organism' in item['Description'].keys():\n",
    "                        if '@taxonomy_id' in item['Description']['Organism'].keys():\n",
    "                            taxa_id = item['Description']['Organism']['@taxonomy_id']\n",
    "                        else:\n",
    "                            taxa_id = 'NA'\n",
    "                        if 'OrganismName' in item['Description']['Organism'].keys():\n",
    "                            organism = item['Description']['Organism']['OrganismName']\n",
    "                        else:\n",
    "                            organism = 'NA'\n",
    "                else:\n",
    "                    title = 'NA'\n",
    "                    taxa_id = 'NA'\n",
    "                    organism = 'NA'\n",
    "                # append pandas DataFrame with data to <geo_data_rows>\n",
    "                geo_data_rows.append(pd.DataFrame.from_dict({'BioSample ID': [biosample_id], \n",
    "                                                             'Geographical location': [geo_loc_name], \n",
    "                                                             'Latitude': [str(lat)], \n",
    "                                                             'Longitude': [str(lon)],\n",
    "                                                             'Isolation source': [isolation_source],\n",
    "                                                             'Title': [title],\n",
    "                                                             'Taxa ID': [taxa_id],\n",
    "                                                             'Organism': [organism]}))               \n",
    "    except:\n",
    "        geo_data_rows.append(pd.DataFrame.from_dict({'BioSample ID': [biosample_id], \n",
    "                                                     'Geographical location': ['ERROR'], \n",
    "                                                     'Latitude': [str('ERROR')], \n",
    "                                                     'Longitude': [str('ERROR')],\n",
    "                                                     'Isolation source': ['ERROR'],\n",
    "                                                     'Title': ['ERROR'],\n",
    "                                                     'Taxa ID': ['ERROR'],\n",
    "                                                     'Organism': ['ERROR']}))\n",
    "        \n",
    "geo_data_table = pd.concat(geo_data_rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38a7f1a0-de74-4583-95a3-74a381545afe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save table\n",
    "geo_data_table.to_csv('../results/geodata_table.tsv', index = False, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e5830a2-d55f-4b1b-9340-7348dc9eaedc",
   "metadata": {},
   "outputs": [],
   "source": [
    "geo_data_table.query(\"`Geographical location` == 'ERROR'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e5e953e4-1d3f-4300-863c-f0081e791b52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioSample ID</th>\n",
       "      <th>Geographical location</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>Longitude</th>\n",
       "      <th>Isolation source</th>\n",
       "      <th>Title</th>\n",
       "      <th>Taxa ID</th>\n",
       "      <th>Organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRS2617844</td>\n",
       "      <td>USA:Emeryville, CA</td>\n",
       "      <td>37.840854 N</td>\n",
       "      <td>122.289843 W</td>\n",
       "      <td>Eden Landing Ponds, San Francisco, CA, USA</td>\n",
       "      <td>Aerobic enrichment media from Eden Landing Pon...</td>\n",
       "      <td>408172</td>\n",
       "      <td>marine metagenome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  BioSample ID Geographical location     Latitude     Longitude  \\\n",
       "0   SRS2617844    USA:Emeryville, CA  37.840854 N  122.289843 W   \n",
       "\n",
       "                             Isolation source  \\\n",
       "0  Eden Landing Ponds, San Francisco, CA, USA   \n",
       "\n",
       "                                               Title Taxa ID  \\\n",
       "0  Aerobic enrichment media from Eden Landing Pon...  408172   \n",
       "\n",
       "            Organism  \n",
       "0  marine metagenome  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geo_data_table.query(\"`BioSample ID` == 'SRS2617844'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a186f8dc-32b0-45f8-b7e0-ca027a2fb3a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import\n",
    "import requests\n",
    "biosample_id = 'SRS2617844'\n",
    "response = requests.get('https://www.ebi.ac.uk/metagenomics/api/v1/samples/{0}'.format(biosample_id))\n",
    "# getting request JSON data structure\n",
    "biosample_metadata_json = response.json()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e8822c2b-77c8-450f-9a71-865f1a3ac119",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'data': {'type': 'samples',\n",
       "  'id': 'SRS2617844',\n",
       "  'attributes': {'longitude': -122.2898,\n",
       "   'latitude': 37.8409,\n",
       "   'biosample': 'SAMN06268793',\n",
       "   'sample-metadata': [{'key': 'geographic location (longitude)',\n",
       "     'value': '-122.28984',\n",
       "     'unit': None},\n",
       "    {'key': 'geographic location (country and/or sea,region)',\n",
       "     'value': 'USA:Emeryville, CA',\n",
       "     'unit': None},\n",
       "    {'key': 'collection date', 'value': '2012-04-25', 'unit': None},\n",
       "    {'key': 'geographic location (latitude)',\n",
       "     'value': '37.840855',\n",
       "     'unit': None},\n",
       "    {'key': 'instrument model', 'value': 'Illumina HiSeq 2000', 'unit': None},\n",
       "    {'key': 'last update date', 'value': '2017-10-27', 'unit': None}],\n",
       "   'accession': 'SRS2617844',\n",
       "   'analysis-completed': '2018-10-25',\n",
       "   'collection-date': '2012-04-25',\n",
       "   'geo-loc-name': 'USA:Emeryville, CA',\n",
       "   'sample-desc': 'Aerobic enrichment media from Eden Landing Ponds, California, USA - A23 E3 (3)',\n",
       "   'environment-biome': None,\n",
       "   'environment-feature': None,\n",
       "   'environment-material': None,\n",
       "   'sample-name': 'Gp0054809',\n",
       "   'sample-alias': 'Gp0054809',\n",
       "   'host-tax-id': None,\n",
       "   'species': None,\n",
       "   'last-update': '2018-10-25T17:20:57'},\n",
       "  'relationships': {'biome': {'data': {'type': 'biomes',\n",
       "     'id': 'root:Environmental:Aquatic:Marine:Intertidal zone:Salt marsh'},\n",
       "    'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/biomes/root:Environmental:Aquatic:Marine:Intertidal%20zone:Salt%20marsh'}},\n",
       "   'studies': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/samples/SRS2617844/studies'},\n",
       "    'data': [],\n",
       "    'meta': {'count': 0}},\n",
       "   'runs': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/samples/SRS2617844/runs'}}},\n",
       "  'links': {'self': 'https://www.ebi.ac.uk/metagenomics/api/v1/samples/SRS2617844'}}}"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biosample_metadata_json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "23c4a438-46d4-483b-b643-3d2d88144846",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SAMN06268799',\n",
       " 'SAMN07630777',\n",
       " 'SAMN07630779',\n",
       " 'SAMN07630779',\n",
       " 'SAMN07631793',\n",
       " 'SAMN07630784',\n",
       " 'SAMN08777606',\n",
       " 'SAMN06267216',\n",
       " 'SAMN06267216',\n",
       " 'SAMN06267216',\n",
       " 'SAMN06267216',\n",
       " 'SAMN06267216',\n",
       " 'SAMN06267216',\n",
       " 'SAMN06267216',\n",
       " 'SAMN06267220',\n",
       " 'SAMN06267220',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06267222',\n",
       " 'SAMN06343954',\n",
       " 'SAMN06343954',\n",
       " 'SAMN06343954',\n",
       " 'SAMN06343954',\n",
       " 'SAMN06343954',\n",
       " 'SAMN06343954',\n",
       " 'SAMN06343954',\n",
       " 'SAMN06343954',\n",
       " 'SAMN06343954',\n",
       " 'SAMN06343954',\n",
       " 'SAMN06343954',\n",
       " 'SAMN06343954',\n",
       " 'SAMN03766243',\n",
       " 'SAMN03781032',\n",
       " 'SAMN06268369',\n",
       " 'SAMN10863950',\n",
       " 'SAMN09201407',\n",
       " 'SAMN06267292',\n",
       " 'SAMN05421986',\n",
       " 'SAMN03842451',\n",
       " 'SAMN05421978',\n",
       " 'SAMEA2619974',\n",
       " 'SRS3033589',\n",
       " 'SRR6823441',\n",
       " 'SRR6823441',\n",
       " 'SRR6823441',\n",
       " 'SRR6823441',\n",
       " 'SRR6823441',\n",
       " 'SRR6823441',\n",
       " 'SRS3033590',\n",
       " 'SAMD00080628',\n",
       " 'ERS524245',\n",
       " 'SRS2617844',\n",
       " 'SRS2617729',\n",
       " 'SRS2617842',\n",
       " 'SRS2617842',\n",
       " 'SRS2617736',\n",
       " 'SRS2617843',\n",
       " 'SRS2617738',\n",
       " 'SRS2617846',\n",
       " 'SRS2236490',\n",
       " 'SRS2617845',\n",
       " 'SAMN04018175',\n",
       " 'SAMN02800870',\n",
       " 'SAMN02471672',\n",
       " 'SAMN02440896',\n",
       " 'SAMN04124421',\n",
       " 'SAMN09214948',\n",
       " 'SAMN06269071',\n",
       " 'SAMN09215303',\n",
       " 'SAMN09215099',\n",
       " 'SAMN06269071',\n",
       " 'SAMN03217314',\n",
       " 'SAMN06264874',\n",
       " 'SAMN06264874',\n",
       " 'SAMN06264874',\n",
       " 'SAMN05518494',\n",
       " 'SAMN15143775',\n",
       " 'SAMN15183787',\n",
       " 'SAMN10872768',\n",
       " 'SAMN12405733',\n",
       " 'SAMN07631051',\n",
       " 'SAMN10967517',\n",
       " 'SAMN14053718']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ecological_data_tibble['BIOSAMPLE'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "b63dd37b-5eb1-4355-86c4-0311b9496a21",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[90m# A tibble: 1 x 8\u001b[39m\n",
      "  `BioSample ID` `Geographical loc… Latitude Longitude `Isolation sourc… Title  \n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m          \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m              \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m             \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \n",
      "\u001b[90m1\u001b[39m SAMEA2619974   NA                 NA       NA        NA                TARA_Y…\n",
      "\u001b[90m# … with 2 more variables: Taxa ID <chr>, Organism <chr>\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "%%R -i geo_data_table\n",
    "\n",
    "\n",
    "geo_data_table %<>% as_tibble() %>% dplyr::filter(`Longitude` != 'ERROR' & !is.na(Longitude))\n",
    "geo_data_table[54,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "e46f2863-5330-4a2d-b566-26ac4a2a8eee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1m\u001b[1mColumn specification\u001b[1m\u001b[22m \u001b[36m─────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  new_tip_label = \u001b[31mcol_character()\u001b[39m,\n",
      "  BIOSAMPLE = \u001b[31mcol_character()\u001b[39m,\n",
      "  `Metagenome Project ID` = \u001b[31mcol_character()\u001b[39m\n",
      ")\n",
      "\n",
      "\u001b[90m# A tibble: 608 x 12\u001b[39m\n",
      "   `BioSample ID` `Geographical lo… Latitude Longitude `Isolation sourc… Title  \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m          \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m             \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \n",
      "\u001b[90m 1\u001b[39m SAMN06268799   USA:Emeryville, …     37.8    -\u001b[31m122\u001b[39m\u001b[31m.\u001b[39m  Eden Landing Pon… Aerobi…\n",
      "\u001b[90m 2\u001b[39m SAMN07630777   Canada: Ontario       43.4     -\u001b[31m80\u001b[39m\u001b[31m.\u001b[39m\u001b[31m6\u001b[39m composite cistern Leacha…\n",
      "\u001b[90m 3\u001b[39m SAMN07630779   Canada: Ontario       43.4     -\u001b[31m80\u001b[39m\u001b[31m.\u001b[39m\u001b[31m6\u001b[39m aquifer           Ground…\n",
      "\u001b[90m 4\u001b[39m SAMN07630779   Canada: Ontario       43.4     -\u001b[31m80\u001b[39m\u001b[31m.\u001b[39m\u001b[31m6\u001b[39m aquifer           Ground…\n",
      "\u001b[90m 5\u001b[39m SAMN07630779   Canada: Ontario       43.4     -\u001b[31m80\u001b[39m\u001b[31m.\u001b[39m\u001b[31m6\u001b[39m aquifer           Ground…\n",
      "\u001b[90m 6\u001b[39m SAMN07630779   Canada: Ontario       43.4     -\u001b[31m80\u001b[39m\u001b[31m.\u001b[39m\u001b[31m6\u001b[39m aquifer           Ground…\n",
      "\u001b[90m 7\u001b[39m SAMN07631793   Sweden: Stordalen     68.4      19.0 Mire fen          Permaf…\n",
      "\u001b[90m 8\u001b[39m SAMN07630784   Sweden: Oskarsha…     57.4      16.7 Aspo Hard Rock L… Ground…\n",
      "\u001b[90m 9\u001b[39m SAMN06267216   Antarctica: Ace …    -\u001b[31m68\u001b[39m\u001b[31m.\u001b[39m\u001b[31m5\u001b[39m      78.2 Filter            Saline…\n",
      "\u001b[90m10\u001b[39m SAMN06267216   Antarctica: Ace …    -\u001b[31m68\u001b[39m\u001b[31m.\u001b[39m\u001b[31m5\u001b[39m      78.2 Filter            Saline…\n",
      "\u001b[90m# … with 598 more rows, and 6 more variables: Taxa ID <chr>, Organism <chr>,\u001b[39m\n",
      "\u001b[90m#   Latitude sign <chr>, Longitude sign <chr>, HEADER <chr>,\u001b[39m\n",
      "\u001b[90m#   BioSample ID2 <chr>\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "%%R -i geo_data_table\n",
    "\n",
    "# load libaries\n",
    "library(leaflet)\n",
    "library(htmlwidgets)\n",
    "library(IRdisplay)\n",
    "\n",
    "# tune a bit the tibble to work \n",
    "geo_data_table %<>% as_tibble() %>% dplyr::filter(`Longitude` != 'ERROR' & !is.na(Longitude) & Longitude != 'NA')\n",
    "geo_data_table %<>% dplyr::mutate(`Latitude sign` = `Latitude` %>% str_split(' ') %>% purrr::map_chr(2),\n",
    "                                  `Longitude sign` = `Longitude` %>% str_split(' ') %>% purrr::map_chr(2),\n",
    "                                  `Latitude` = `Latitude` %>% str_split(' ') %>% purrr::map_chr(1) %>% as.numeric(),\n",
    "                                  `Latitude` = case_when(`Latitude sign` == 'N' ~ `Latitude`,\n",
    "                                                         `Latitude sign` == 'S' ~ `Latitude`*-1),\n",
    "                                  `Longitude` = `Longitude` %>% str_split(' ') %>% purrr::map_chr(1) %>% as.numeric(),\n",
    "                                  `Longitude` = case_when(`Longitude sign` == 'E' ~ `Longitude`,\n",
    "                                                          `Longitude sign` == 'W' ~ `Longitude`*-1)) %>%\n",
    "                    dplyr::filter(!is.na(`Latitude`) & !is.na(`Longitude`))\n",
    "\n",
    "geo_data_table %<>%\n",
    "    dplyr::left_join(x = .,\n",
    "                    y = (ecological_data.tibble %>% dplyr::select(HEADER, BIOSAMPLE)),\n",
    "                    by = c('BioSample ID' = 'BIOSAMPLE'))\n",
    "\n",
    "geo_data_table %>%\n",
    "    group_by(`BioSample ID`) %>%\n",
    "    dplyr::mutate(HEADERs = paste(`HEADER`, collapse = ', ')) %>%\n",
    "    dplyr::select(-HEADER) %>%\n",
    "    unique()\n",
    "\n",
    "# load table and creating dictionary for MP IDs for BioSample IDs\n",
    "metagenome_project_ID_tips_and_biosample_corr_table = readr::read_tsv('../results/metagenome_project_ID_tips_and_biosample_corr_table.tsv')\n",
    "metagenome_projects.dict = metagenome_project_ID_tips_and_biosample_corr_table$`Metagenome Project ID`\n",
    "names(metagenome_projects.dict) = metagenome_project_ID_tips_and_biosample_corr_table$BIOSAMPLE\n",
    "\n",
    "library(tidyverse)\n",
    "geo_data_table %<>%\n",
    "    dplyr::mutate(`BioSample ID2` = `BioSample ID` %>% tidytidbits::lookup_chr(., dict = metagenome_projects.dict , default = identity))\n",
    "\n",
    "m <- leaflet() %>%\n",
    "  addTiles() %>%  # Add default OpenStreetMap map tiles\n",
    "  addMarkers(lng=174.768, lat=-36.852, popup=\"The birthplace of R\")\n",
    "\n",
    "saveWidget(m, 'demo.html', selfcontained = TRUE)\n",
    "\n",
    "k = leaflet::leaflet(geo_data_table) %>%\n",
    "  addTiles() %>%  # Add default OpenStreetMap map tiles\n",
    "  addMarkers(~Longitude, ~Latitude, label = geo_data_table$`BioSample ID2`, popup = glue('Biosample ID: {geo_data_table$`BioSample ID`} \\n Title:{geo_data_table$`Title`} \\n Geographical location: {geo_data_table$`Geographical location`} \\n Isolation source: {geo_data_table$`Isolation source`}'))\n",
    "\n",
    "saveWidget(k, 'demo2.html', selfcontained = TRUE)\n",
    "\n",
    "\n",
    "geo_data_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "00191fa5-565c-4861-8004-dcf43d2f8c67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"MP 1\"         \"MP 2\"         \"MP 3\"         \"MP 9\"         \"MP 4\"        \n",
      " [6] \"MP 11\"        \"MP 19\"        \"MP 6\"         \"MP 7\"         \"MP 22\"       \n",
      "[11] \"MP 16\"        \"MP 10\"        \"MP 28\"        \"MP 20\"        \"MP 12\"       \n",
      "[16] \"MP 14\"        \"MP 34\"        \"MP 18\"        \"SRS2617844\"   \"SRS2617729\"  \n",
      "[21] \"MP 44\"        \"SRS2617736\"   \"SRS2617843\"   \"SRS2617738\"   \"MP 49\"       \n",
      "[26] \"SRS2236490\"   \"SRS2617845\"   \"MP 40\"        \"MP 47\"        \"MP 38\"       \n",
      "[31] \"MP 35\"        \"MP 31\"        \"MP 17\"        \"SAMN03217314\" \"MP 32\"       \n",
      "[36] \"MP 36\"        \"MP 5\"         \"MP 21\"        \"MP 33\"        \"MP 24\"       \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "geo_data_table %>%\n",
    "    group_by(`BioSample ID`) %>%\n",
    "    dplyr::mutate(HEADERs = paste(`HEADER`, collapse = ', ')) %>%\n",
    "    dplyr::select(-HEADER) %>%\n",
    "    unique() %>%\n",
    "    .$`BioSample ID2`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "3d578dac-b011-42a1-8e60-0f73a3bc8735",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"1200\"\n",
       "            height=\"700\"\n",
       "            src=\"./demo2.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7fe477ba7a90>"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "\n",
    "IFrame(src='./demo2.html', width=1200, height=700)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "7173b8d3-ba82-461f-a42b-821b43d83909",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[90m# A tibble: 99 x 2\u001b[39m\n",
      "   HEADER                                                       BIOSAMPLE   \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                                                        \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m       \n",
      "\u001b[90m 1\u001b[39m 3300000868-JGI12330J12834-1000008-299010-8                   SAMN06268799\n",
      "\u001b[90m 2\u001b[39m 3300014206-Ga0172377-10000119-870930-129                     SAMN07630777\n",
      "\u001b[90m 3\u001b[39m 3300014208-Ga0172379-10000243-871512-158                     SAMN07630779\n",
      "\u001b[90m 4\u001b[39m 3300014208-Ga0172379-10001592-871560-40                      SAMN07630779\n",
      "\u001b[90m 5\u001b[39m 3300014494-Ga0182017-10003408-872394-14                      SAMN07631793\n",
      "\u001b[90m 6\u001b[39m 3300014613-Ga0180008-1001212-875221-12                       SAMN07630784\n",
      "\u001b[90m 7\u001b[39m 3300018015-Ga0187866-1000629-915963-9                        SAMN08777606\n",
      "\u001b[90m 8\u001b[39m AntAceMinimDraft_10_1070366.scaffolds.fasta_scaffold00069_38 SAMN06267216\n",
      "\u001b[90m 9\u001b[39m AntAceMinimDraft_10_1070366.scaffolds.fasta_scaffold00511_6  SAMN06267216\n",
      "\u001b[90m10\u001b[39m AntAceMinimDraft_10_1070366.scaffolds.fasta_scaffold03294_14 SAMN06267216\n",
      "\u001b[90m# … with 89 more rows\u001b[39m\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "ecological_data.tibble %>%\n",
    "    dplyr::select(HEADER, BIOSAMPLE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "eba0665c-1a07-404c-b0ca-19bf34722a53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File: /tmp/RtmpjD89pD/Rtxt64f27b644ae9\n",
      "leaflet                package:leaflet                 R Documentation\n",
      "\n",
      "\n",
      "\n",
      "Create a Leaflet map widget\n",
      "\n",
      "\n",
      "\n",
      "Description:\n",
      "\n",
      "\n",
      "\n",
      "     This function creates a Leaflet map widget using ‘htmlwidgets’.\n",
      "\n",
      "     The widget can be rendered on HTML pages generated from R\n",
      "\n",
      "     Markdown, Shiny, or other applications.\n",
      "\n",
      "\n",
      "\n",
      "Usage:\n",
      "\n",
      "\n",
      "\n",
      "     leaflet(\n",
      "\n",
      "       data = NULL,\n",
      "\n",
      "       width = NULL,\n",
      "\n",
      "       height = NULL,\n",
      "\n",
      "       padding = 0,\n",
      "\n",
      "       options = leafletOptions(),\n",
      "\n",
      "       elementId = NULL,\n",
      "\n",
      "       sizingPolicy = leafletSizingPolicy(padding = padding)\n",
      "\n",
      "     )\n",
      "\n",
      "     \n",
      "\n",
      "     leafletOptions(\n",
      "\n",
      "       minZoom = NULL,\n",
      "\n",
      "       maxZoom = NULL,\n",
      "\n",
      "       crs = leafletCRS(),\n",
      "\n",
      "       worldCopyJump = NULL,\n",
      "\n",
      "       preferCanvas = NULL,\n",
      "\n",
      "       ...\n",
      "\n",
      "     )\n",
      "\n",
      "     \n",
      "\n",
      "     leafletCRS(\n",
      "\n",
      "       crsClass = \"L.CRS.EPSG3857\",\n",
      "\n",
      "       code = NULL,\n",
      "\n",
      "       proj4def = NULL,\n",
      "\n",
      "       projectedBounds = NULL,\n",
      "\n",
      "       origin = NULL,\n",
      "\n",
      "       transformation = NULL,\n",
      "\n",
      "       scales = NULL,\n",
      "\n",
      "       resolutions = NULL,\n",
      "\n",
      "       bounds = NULL,\n",
      "\n",
      "       tileSize = NULL\n",
      "\n",
      "     )\n",
      "\n",
      "     \n",
      "\n",
      "Arguments:\n",
      "\n",
      "\n",
      "\n",
      "    data: a data object. Currently supported objects are matrices, data\n",
      "\n",
      "          frames, spatial objects from the ‘sp’ package\n",
      "\n",
      "          (‘SpatialPoints’, ‘SpatialPointsDataFrame’, ‘Polygon’,\n",
      "\n",
      "          ‘Polygons’, ‘SpatialPolygons’, ‘SpatialPolygonsDataFrame’,\n",
      "\n",
      "          ‘Line’, ‘Lines’, ‘SpatialLines’, and\n",
      "\n",
      "          ‘SpatialLinesDataFrame’), and spatial data frames from the\n",
      "\n",
      "          ‘sf’ package.\n",
      "\n",
      "\n",
      "\n",
      "   width: the width of the map\n",
      "\n",
      "\n",
      "\n",
      "  height: the height of the map\n",
      "\n",
      "\n",
      "\n",
      " padding: the padding of the map\n",
      "\n",
      "\n",
      "\n",
      " options: the map options\n",
      "\n",
      "\n",
      "\n",
      "elementId: Use an explicit element ID for the widget (rather than an\n",
      "\n",
      "          automatically generated one).\n",
      "\n",
      "\n",
      "\n",
      "sizingPolicy: htmlwidgets sizing policy object. Defaults to\n",
      "\n",
      "          ‘leafletSizingPolicy()’\n",
      "\n",
      "\n",
      "\n",
      " minZoom: Minimum zoom level of the map. Overrides any minZoom set on\n",
      "\n",
      "          map layers.\n",
      "\n",
      "\n",
      "\n",
      " maxZoom: Maximum zoom level of the map. This overrides any maxZoom set\n",
      "\n",
      "          on map layers.\n",
      "\n",
      "\n",
      "\n",
      "     crs: Coordinate Reference System to use. Don't change this if\n",
      "\n",
      "          you're not sure what it means.\n",
      "\n",
      "\n",
      "\n",
      "worldCopyJump: With this option enabled, the map tracks when you pan to\n",
      "\n",
      "          another \"copy\" of the world and seamlessly jumps to the\n",
      "\n",
      "          original one so that all overlays like markers and vector\n",
      "\n",
      "          layers are still visible.\n",
      "\n",
      "\n",
      "\n",
      "preferCanvas: Whether leaflet.js Paths should be rendered on a Canvas\n",
      "\n",
      "          renderer.\n",
      "\n",
      "\n",
      "\n",
      "     ...: other options used for leaflet.js map creation.\n",
      "\n",
      "\n",
      "\n",
      "crsClass: One of L.CRS.EPSG3857, L.CRS.EPSG4326, L.CRS.EPSG3395,\n",
      "\n",
      "          L.CRS.Simple, L.Proj.CRS\n",
      "\n",
      "\n",
      "\n",
      "    code: CRS identifier\n",
      "\n",
      "\n",
      "\n",
      "proj4def: Proj4 string\n",
      "\n",
      "\n",
      "\n",
      "projectedBounds: DEPRECATED! Use the bounds argument.\n",
      "\n",
      "\n",
      "\n",
      "  origin: Origin in projected coordinates, if set overrides\n",
      "\n",
      "          transformation option.\n",
      "\n",
      "\n",
      "\n",
      "transformation: to use when transforming projected coordinates into\n",
      "\n",
      "          pixel coordinates\n",
      "\n",
      "\n",
      "\n",
      "  scales: Scale factors (pixels per projection unit, for example\n",
      "\n",
      "          pixels/meter) for zoom levels; specify either scales or\n",
      "\n",
      "          resolutions, not both\n",
      "\n",
      "\n",
      "\n",
      "resolutions: factors (projection units per pixel, for example\n",
      "\n",
      "          meters/pixel) for zoom levels; specify either scales or\n",
      "\n",
      "          resolutions, not both\n",
      "\n",
      "\n",
      "\n",
      "  bounds: Bounds of the CRS, in projected coordinates; if defined,\n",
      "\n",
      "          Proj4Leaflet will use this in the getSize method, otherwise\n",
      "\n",
      "          defaulting to Leaflet's default CRS size\n",
      "\n",
      "\n",
      "\n",
      "tileSize: DEPRECATED! Specify the tilesize in the ‘tileOptions()’\n",
      "\n",
      "          argument.\n",
      "\n",
      "\n",
      "\n",
      "Details:\n",
      "\n",
      "\n",
      "\n",
      "     The ‘data’ argument is only needed if you are going to reference\n",
      "\n",
      "     variables in this object later in map layers. For example, ‘data’\n",
      "\n",
      "     can be a data frame containing columns ‘latitude’ and\n",
      "\n",
      "     ‘longtitude’, then we may add a circle layer to the map by\n",
      "\n",
      "     ‘leaflet(data) %>% addCircles(lat = ~latitude, lng =\n",
      "\n",
      "     ~longtitude)’, where the variables in the formulae will be\n",
      "\n",
      "     evaluated in the ‘data’.\n",
      "\n",
      "\n",
      "\n",
      "Value:\n",
      "\n",
      "\n",
      "\n",
      "     A HTML widget object, on which we can add graphics layers using\n",
      "\n",
      "     ‘%>%’ (see examples).\n",
      "\n",
      "\n",
      "\n",
      "Functions:\n",
      "\n",
      "\n",
      "\n",
      "        • ‘leafletOptions’: Options for map creation\n",
      "\n",
      "\n",
      "\n",
      "        • ‘leafletCRS’: class to create a custom CRS\n",
      "\n",
      "\n",
      "\n",
      "See Also:\n",
      "\n",
      "\n",
      "\n",
      "     ‘leafletCRS’ for creating a custom CRS.\n",
      "\n",
      "\n",
      "\n",
      "     See <URL: https://leafletjs.com/reference-1.3.4.html#map-option>\n",
      "\n",
      "     for details and more options.\n",
      "\n",
      "\n",
      "\n",
      "Examples:\n",
      "\n",
      "\n",
      "\n",
      "     # !formatR\n",
      "\n",
      "     library(leaflet)\n",
      "\n",
      "     m <- leaflet() %>% addTiles()\n",
      "\n",
      "     m  # a map with the default OSM tile layer\n",
      "\n",
      "     \n",
      "\n",
      "     \n",
      "\n",
      "     # set bounds\n",
      "\n",
      "     m %>% fitBounds(0, 40, 10, 50)\n",
      "\n",
      "     \n",
      "\n",
      "     # move the center to Snedecor Hall\n",
      "\n",
      "     m <- m %>% setView(-93.65, 42.0285, zoom = 17)\n",
      "\n",
      "     m\n",
      "\n",
      "     \n",
      "\n",
      "     # popup\n",
      "\n",
      "     m %>% addPopups(-93.65, 42.0285, \"Here is the <b>Department of Statistics</b>, ISU\")\n",
      "\n",
      "     rand_lng <- function(n = 10) rnorm(n, -93.65, .01)\n",
      "\n",
      "     rand_lat <- function(n = 10) rnorm(n, 42.0285, .01)\n",
      "\n",
      "     \n",
      "\n",
      "     # use automatic bounds derived from lng/lat data\n",
      "\n",
      "     m <- m %>% clearBounds()\n",
      "\n",
      "     \n",
      "\n",
      "     # popup\n",
      "\n",
      "     m %>% addPopups(rand_lng(), rand_lat(), \"Random popups\")\n",
      "\n",
      "     \n",
      "\n",
      "     # marker\n",
      "\n",
      "     m %>% addMarkers(rand_lng(), rand_lat())\n",
      "\n",
      "     m %>% addMarkers(\n",
      "\n",
      "       rand_lng(), rand_lat(), popup = paste(\"A random letter\", sample(LETTERS, 10))\n",
      "\n",
      "     )\n",
      "\n",
      "     \n",
      "\n",
      "     Rlogo <- file.path(R.home(\"doc\"), \"html\", \"logo.jpg\")\n",
      "\n",
      "     m %>% addMarkers(\n",
      "\n",
      "       174.7690922, -36.8523071, icon = list(\n",
      "\n",
      "         iconUrl = Rlogo, iconSize = c(100, 76)\n",
      "\n",
      "       ), popup = \"R was born here!\"\n",
      "\n",
      "     )\n",
      "\n",
      "     \n",
      "\n",
      "     m %>% addMarkers(rnorm(30, 175), rnorm(30, -37), icon = list(\n",
      "\n",
      "       iconUrl = Rlogo, iconSize = c(25, 19)\n",
      "\n",
      "     ))\n",
      "\n",
      "     \n",
      "\n",
      "     m %>% addMarkers(\n",
      "\n",
      "       c(-71.0382679, -122.1217866), c(42.3489054, 47.6763144), icon = list(\n",
      "\n",
      "         iconUrl = \"http://www.rstudio.com/wp-content/uploads/2014/03/blue-125.png\"\n",
      "\n",
      "       ), popup = c(\"RStudio @ Boston\", \"RStudio @ Seattle\")\n",
      "\n",
      "     )\n",
      "\n",
      "     \n",
      "\n",
      "     \n",
      "\n",
      "     # circle (units in metres)\n",
      "\n",
      "     m %>% addCircles(rand_lng(50), rand_lat(50), radius = runif(50, 50, 150))\n",
      "\n",
      "     \n",
      "\n",
      "     # circle marker (units in pixels)\n",
      "\n",
      "     m %>% addCircleMarkers(rand_lng(50), rand_lat(50), color = \"#ff0000\")\n",
      "\n",
      "     m %>% addCircleMarkers(rand_lng(100), rand_lat(100), radius = runif(100, 5, 15))\n",
      "\n",
      "     \n",
      "\n",
      "     # rectangle\n",
      "\n",
      "     m %>% addRectangles(\n",
      "\n",
      "       rand_lng(), rand_lat(), rand_lng(), rand_lat(),\n",
      "\n",
      "       color = \"red\", fill = FALSE, dashArray = \"5,5\", weight = 3\n",
      "\n",
      "     )\n",
      "\n",
      "     \n",
      "\n",
      "     # polyline\n",
      "\n",
      "     m %>% addPolylines(rand_lng(50), rand_lat(50))\n",
      "\n",
      "     \n",
      "\n",
      "     # polygon\n",
      "\n",
      "     m %>% addPolygons(rand_lng(), rand_lat(), layerId = \"foo\")\n",
      "\n",
      "     \n",
      "\n",
      "     # geoJSON\n",
      "\n",
      "     seattle_geojson <- list(\n",
      "\n",
      "       type = \"Feature\",\n",
      "\n",
      "       geometry = list(\n",
      "\n",
      "         type = \"MultiPolygon\",\n",
      "\n",
      "         coordinates = list(list(list(\n",
      "\n",
      "           c(-122.36075812146,  47.6759920119894),\n",
      "\n",
      "           c(-122.360781646764, 47.6668890126755),\n",
      "\n",
      "           c(-122.360782108665,  47.6614990696722),\n",
      "\n",
      "           c(-122.366199035722, 47.6614990696722),\n",
      "\n",
      "           c(-122.366199035722,  47.6592874248973),\n",
      "\n",
      "           c(-122.364582509469, 47.6576254522105),\n",
      "\n",
      "           c(-122.363887331445,  47.6569107302038),\n",
      "\n",
      "           c(-122.360865528129, 47.6538418253251),\n",
      "\n",
      "           c(-122.360866157644,  47.6535254473167),\n",
      "\n",
      "           c(-122.360866581103, 47.6533126275176),\n",
      "\n",
      "           c(-122.362526540691,  47.6541872926348),\n",
      "\n",
      "           c(-122.364442114483, 47.6551892850798),\n",
      "\n",
      "           c(-122.366077719797,  47.6560733960606),\n",
      "\n",
      "           c(-122.368818463838, 47.6579742346694),\n",
      "\n",
      "           c(-122.370115159943,  47.6588730808334),\n",
      "\n",
      "           c(-122.372295967029, 47.6604350102328),\n",
      "\n",
      "           c(-122.37381369088,  47.660582362063),\n",
      "\n",
      "           c(-122.375522972109, 47.6606413027949),\n",
      "\n",
      "           c(-122.376079703095,  47.6608793094619),\n",
      "\n",
      "           c(-122.376206315662, 47.6609242364243),\n",
      "\n",
      "           c(-122.377610811371,  47.6606160735197),\n",
      "\n",
      "           c(-122.379857378879, 47.6610306942278),\n",
      "\n",
      "           c(-122.382454873022,  47.6627496239169),\n",
      "\n",
      "           c(-122.385357955057, 47.6638573778241),\n",
      "\n",
      "           c(-122.386007328104,  47.6640865692306),\n",
      "\n",
      "           c(-122.387186331506, 47.6654326177161),\n",
      "\n",
      "           c(-122.387802656231,  47.6661492860294),\n",
      "\n",
      "           c(-122.388108244121, 47.6664548739202),\n",
      "\n",
      "           c(-122.389177800763,  47.6663784774359),\n",
      "\n",
      "           c(-122.390582858689, 47.6665072251861),\n",
      "\n",
      "           c(-122.390793942299,  47.6659699214511),\n",
      "\n",
      "           c(-122.391507906234, 47.6659200946229),\n",
      "\n",
      "           c(-122.392883050767,  47.6664166747017),\n",
      "\n",
      "           c(-122.392847210144, 47.6678696739431),\n",
      "\n",
      "           c(-122.392904778401,  47.6709016021624),\n",
      "\n",
      "           c(-122.39296705153, 47.6732047491624),\n",
      "\n",
      "           c(-122.393000803496,  47.6759322346303),\n",
      "\n",
      "           c(-122.37666945305, 47.6759896300663),\n",
      "\n",
      "           c(-122.376486363943,  47.6759891899754),\n",
      "\n",
      "           c(-122.366078869215, 47.6759641734893),\n",
      "\n",
      "           c(-122.36075812146,  47.6759920119894)\n",
      "\n",
      "         )))\n",
      "\n",
      "       ),\n",
      "\n",
      "       properties = list(\n",
      "\n",
      "         name = \"Ballard\",\n",
      "\n",
      "         population = 48000,\n",
      "\n",
      "         # You can inline styles if you want\n",
      "\n",
      "         style = list(\n",
      "\n",
      "           fillColor = \"yellow\",\n",
      "\n",
      "           weight = 2,\n",
      "\n",
      "           color = \"#000000\"\n",
      "\n",
      "         )\n",
      "\n",
      "       ),\n",
      "\n",
      "       id = \"ballard\"\n",
      "\n",
      "     )\n",
      "\n",
      "     m %>% setView(-122.36075812146, 47.6759920119894, zoom = 13) %>% addGeoJSON(seattle_geojson)\n",
      "\n",
      "     \n",
      "\n",
      "     \n",
      "\n",
      "     # use the Dark Matter layer from CartoDB\n",
      "\n",
      "     leaflet() %>% addTiles(\"http://{s}.basemaps.cartocdn.com/dark_all/{z}/{x}/{y}.png\",\n",
      "\n",
      "       attribution = paste(\n",
      "\n",
      "         \"&copy; <a href=\\\"http://openstreetmap.org\\\">OpenStreetMap</a> contributors\",\n",
      "\n",
      "         \"&copy; <a href=\\\"http://cartodb.com/attributions\\\">CartoDB</a>\"\n",
      "\n",
      "       )\n",
      "\n",
      "     ) %>% setView(-122.36, 47.67, zoom = 10)\n",
      "\n",
      "     \n",
      "\n",
      "     # provide a data frame to leaflet()\n",
      "\n",
      "     categories <- LETTERS[1:10]\n",
      "\n",
      "     df <- data.frame(\n",
      "\n",
      "       lat = rand_lat(100), lng = rand_lng(100), size = runif(100, 5, 20),\n",
      "\n",
      "       category = factor(sample(categories, 100, replace = TRUE), levels = categories),\n",
      "\n",
      "       value = rnorm(100)\n",
      "\n",
      "     )\n",
      "\n",
      "     m <- leaflet(df) %>% addTiles()\n",
      "\n",
      "     m %>% addCircleMarkers(~lng, ~lat, radius = ~size)\n",
      "\n",
      "     m %>% addCircleMarkers(~lng, ~lat, radius = runif(100, 4, 10), color = c(\"red\"))\n",
      "\n",
      "     \n",
      "\n",
      "     # Discrete colors using the \"RdYlBu\" colorbrewer palette, mapped to categories\n",
      "\n",
      "     RdYlBu <- colorFactor(\"RdYlBu\", domain = categories)\n",
      "\n",
      "     m %>% addCircleMarkers(~lng, ~lat, radius = ~size,\n",
      "\n",
      "       color = ~RdYlBu(category), fillOpacity = 0.5)\n",
      "\n",
      "     \n",
      "\n",
      "     # Continuous colors using the \"Greens\" colorbrewer palette, mapped to value\n",
      "\n",
      "     greens <- colorNumeric(\"Greens\", domain = NULL)\n",
      "\n",
      "     m %>% addCircleMarkers(~lng, ~lat, radius = ~size,\n",
      "\n",
      "       color = ~greens(value), fillOpacity = 0.5)\n",
      "\n",
      "     \n",
      "\n",
      "---\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "help(leaflet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "059d0f2d-d09c-424b-afb0-4df94a28884f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
