{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "70169b32-c3d7-4aa3-8a3b-71deb9defa30",
   "metadata": {},
   "source": [
    "# Overview\n",
    "subset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f4dc184-3356-42d5-908f-bb931af40b2a",
   "metadata": {},
   "source": [
    "## Loading data and creating files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "631dd653-cf1a-4742-ab56-851be4fd61ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries\n",
    "import os\n",
    "import glob\n",
    "import subprocess\n",
    "import toytree\n",
    "import toyplot.pdf\n",
    "import shutil\n",
    "from Bio import AlignIO, SeqIO\n",
    "\n",
    "# create directory for subsetted sets, and the subfolders for BED files and block_Files\n",
    "def create_dir(dir):\n",
    "    if not os.path.exists(dir):\n",
    "        os.mkdir(dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "afa43330-50d8-4ba2-9538-901a6ea3e66f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load xlsx pandas\n",
    "subset_clusters_table = pd.read_excel(io = '../results/plotting_ME_synteny/data/HGT_signature_ORFs_modified_210818.xlsx',\n",
    "                                      sheet_name = 'for plotting')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "21d2b487-3ba6-4f10-9006-98378bb80831",
   "metadata": {},
   "outputs": [],
   "source": [
    "# copying files\n",
    "hgs_to_split = ['CG_2', 'CG_17']\n",
    "for faa_file in glob.glob('../results/trying_to_catch_homologs/results/connected_groups_filtered/fastas/protein/*faa'):\n",
    "    hg_name = faa_file.rpartition('/')[2].split('.')[0]\n",
    "    if hg_name not in hgs_to_split:\n",
    "        shutil.copy(src = faa_file, dst = '../results/trying_to_catch_homologs/results/connected_groups_filtered_splitted_arCOGs/fastas/protein/')\n",
    "    if hg_name in hgs_to_split:\n",
    "        # get full list of record ids in FASTA\n",
    "        record_ids_fasta = [record.id for record in SeqIO.parse(faa_file, 'fasta')]\n",
    "        records_fasta = [record for record in SeqIO.parse(faa_file, 'fasta')]\n",
    "        #print('\\n----\\n')\n",
    "        #print('record_ids for', hg_name, 'at the beginning:', record_ids_fasta)\n",
    "        # grouping by specified column and create the subsetted files\n",
    "        for name, group in subset_clusters_table.query(\"`Cluster`.str.contains(@hg_name)\").groupby('Cluster'):\n",
    "            # get sequences for group, from table\n",
    "            group_seqs = group['Organism_ORF'].to_list() \n",
    "            # get record for sequences that enter in group\n",
    "            group_output = '../results/trying_to_catch_homologs/results/connected_groups_filtered_splitted_arCOGs/fastas/protein/{0}.faa'.format(name) \n",
    "            # create output name and save fasta\n",
    "            if not os.path.exists(group_output):\n",
    "                group_records = [record for record in records_fasta if record.id in group_seqs]\n",
    "                #print('group_records prior to exporting', group_records)\n",
    "                with open(group_output, 'w') as handle_fasta:\n",
    "                    SeqIO.write(group_records, handle_fasta, 'fasta')\n",
    "            # remove this ids from <record_ids_fasta>\n",
    "            group_records = [record for record in SeqIO.parse(faa_file, 'fasta') if record.id in group_seqs]\n",
    "            for record_id in [record.id for record in group_records]:\n",
    "                record_ids_fasta.remove(record_id)\n",
    "            #print('hg_name subsetted:', record_ids_fasta)\n",
    "        # getting those records not present in the list, and printing to output file\n",
    "        # getting output file name\n",
    "        if hg_name == 'CG_2':\n",
    "            output_fasta_unknowns = 'CG_2i'\n",
    "        if hg_name == 'CG_17':\n",
    "            output_fasta_unknowns = 'CG_17b'\n",
    "        # create output name\n",
    "        output_unknown_seqs = '../results/trying_to_catch_homologs/results/connected_groups_filtered_splitted_arCOGs/fastas/protein/{0}.faa'.format(output_fasta_unknowns)\n",
    "        # get those not present\n",
    "        seqids_unknowns = record_ids_fasta # because, I removed sequences as creating files\n",
    "        records_unknowns = [record for record in SeqIO.parse(faa_file, 'fasta') if record.id in seqids_unknowns]\n",
    "        # get to file\n",
    "        if not os.path.exists(output_unknown_seqs):\n",
    "            with open(output_unknown_seqs, 'w') as handle_fasta:\n",
    "                SeqIO.write(records_unknowns, handle_fasta, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b8bc346-c1bc-47ba-ae23-03c5ca4546ab",
   "metadata": {},
   "source": [
    "## Creating BED records"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50095fc5-f984-417c-a86b-ec0b57d20f9d",
   "metadata": {},
   "source": [
    "### Creating GFF3 for features annotated in PATRIC + ORFs in remote homologues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4329f6ae-f3aa-4dbd-8a2b-454f07903111",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load libraries\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "# creating complete GFF for each of the ME ORFs predictions\n",
    "# create directory to allocate results\n",
    "creating_dirs = ['../results/MEs_predicted_orfs_GFFs', \n",
    "                 '../results/MEs_predicted_orfs_GFFs/standard_nomenclature/',\n",
    "                 '../results/MEs_predicted_orfs_GFFs/simplified_nomenclature',\n",
    "                 '../results/plotting_ME_synteny',\n",
    "                 '../results/plotting_ME_synteny/data',\n",
    "                 '../results/plotting_ME_synteny/data/block_files',\n",
    "                 '../results/plotting_ME_synteny/data/GFFs',\n",
    "                 '../results/plotting_ME_synteny/data/GFFs_all_records/',\n",
    "                 '../results/plotting_ME_synteny/data/BED',\n",
    "                 '../results/plotting_ME_synteny/data/BED_all_records',\n",
    "                 '../results/plotting_ME_synteny/results']\n",
    "\n",
    "for dir in creating_dirs:\n",
    "    if not os.path.exists(dir):\n",
    "        os.mkdir(dir)\n",
    "\n",
    "# listing ORF prediction files and creating a list of all CDS that are in the remote homolog groups\n",
    "ORFs_in_remote_homologs_groups = []\n",
    "for remote_homologs_group_file in glob.glob('../results/trying_to_catch_homologs/results/connected_groups_filtered_splitted_arCOGs/fastas/protein/*faa'):\n",
    "    for record in SeqIO.parse(remote_homologs_group_file, 'fasta'):\n",
    "        ORFs_in_remote_homologs_groups.append(record.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "41580281-1f86-41ef-839b-a888ed2e233a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading table with new annotations\n",
    "import pandas as pd\n",
    "\n",
    "orf_annotation_table = pd.read_csv('../results/MEs_predicted_orfs_renamed/data/annotation_table_ORFs_renamed.tsv', sep = '\\t')\n",
    "orf_annotation_table.head()\n",
    "# perform little change\n",
    "orf_annotation_table_to_import = orf_annotation_table.drop(['Featured ID in Genome Annotation'], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e0369d35-5455-4c2f-b1d9-310f790e4105",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c1583084-0c53-4585-a0eb-e2502a0d09f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: ── \u001b[1mAttaching packages\u001b[22m ──────────────────────────────────────────────────────────────────────────────────────────────── tidyverse 1.3.1 ──\n",
      "\n",
      "R[write to console]: \u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.3     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.2     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.6\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.1.3     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 1.4.0     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.1\n",
      "\n",
      "R[write to console]: ── \u001b[1mConflicts\u001b[22m ─────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘magrittr’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    set_names\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    extract\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘glue’\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1m\u001b[1mColumn specification\u001b[1m\u001b[22m \u001b[36m─────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  `Genome ID` = \u001b[31mcol_character()\u001b[39m,\n",
      "  `ORF ID` = \u001b[31mcol_character()\u001b[39m,\n",
      "  `Featured ID in Genome Annotation` = \u001b[31mcol_character()\u001b[39m,\n",
      "  `getorf featured ID` = \u001b[31mcol_character()\u001b[39m,\n",
      "  contig = \u001b[31mcol_character()\u001b[39m,\n",
      "  strand = \u001b[31mcol_character()\u001b[39m,\n",
      "  START = \u001b[32mcol_double()\u001b[39m,\n",
      "  STOP = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i ORFs_in_remote_homologs_groups -o ME_ORFs_annot_table_simplecode\n",
    "\n",
    "library(tidyverse)\n",
    "library(magrittr)\n",
    "library(glue)\n",
    "\n",
    "# loading TSV with annotations for ME ORFs. with some modifications this can be actually converted into a GFF\n",
    "ME_ORFs_annot_table = readr::read_tsv('../results/MEs_predicted_orfs_renamed/data/annotation_table_ORFs_renamed.tsv')\n",
    "\n",
    "# saving version with simplified code\n",
    "ME_ORFs_annot_table_simplecode = ME_ORFs_annot_table %>%\n",
    "    # selecting features employed in the GFF file\n",
    "    dplyr::select(`Genome ID`, `ORF ID`, `contig`, `strand`, `START`, `STOP`) %>%\n",
    "    dplyr::rename(seqid = 'contig',\n",
    "                  start = 'START',\n",
    "                  end = 'STOP') %>%\n",
    "    dplyr::mutate(seqid = seqid %>% str_split(' ') %>% purrr::map_chr(1), source = 'getorf', phase = '1', type = 'CDS', score = 0.000000000000001, attributes = as.character(glue('ID={`ORF ID`};Name={`ORF ID`}'))) %>%\n",
    "    .[, c('Genome ID', 'seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes')]\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "60eb4f5a-dd37-4787-9785-c953892eb3d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1m\u001b[1mColumn specification\u001b[1m\u001b[22m \u001b[36m─────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "cols(\n",
      "  `Genome ID` = \u001b[31mcol_character()\u001b[39m,\n",
      "  `ORF ID` = \u001b[31mcol_character()\u001b[39m,\n",
      "  `Featured ID in Genome Annotation` = \u001b[31mcol_character()\u001b[39m,\n",
      "  `getorf featured ID` = \u001b[31mcol_character()\u001b[39m,\n",
      "  contig = \u001b[31mcol_character()\u001b[39m,\n",
      "  strand = \u001b[31mcol_character()\u001b[39m,\n",
      "  START = \u001b[32mcol_double()\u001b[39m,\n",
      "  STOP = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "[[1]]\n",
      "\u001b[90m# A tibble: 1,011 x 9\u001b[39m\n",
      "   seqid   source type  start   end score strand phase attributes               \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                    \n",
      "\u001b[90m 1\u001b[39m CP0104… getorf CDS    \u001b[4m3\u001b[24m001  \u001b[4m4\u001b[24m869 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_92_Nanoarchaeota_…\n",
      "\u001b[90m 2\u001b[39m CP0104… getorf CDS    \u001b[4m5\u001b[24m764  \u001b[4m6\u001b[24m360 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_120_Nanoarchaeota…\n",
      "\u001b[90m 3\u001b[39m CP0104… getorf CDS    \u001b[4m6\u001b[24m685  \u001b[4m6\u001b[24m981 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_128_Nanoarchaeota…\n",
      "\u001b[90m 4\u001b[39m CP0104… getorf CDS    \u001b[4m7\u001b[24m686  \u001b[4m8\u001b[24m144 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_154_Nanoarchaeota…\n",
      "\u001b[90m 5\u001b[39m CP0104… getorf CDS    \u001b[4m9\u001b[24m341 \u001b[4m1\u001b[24m\u001b[4m0\u001b[24m219 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_189_Nanoarchaeota…\n",
      "\u001b[90m 6\u001b[39m CP0104… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m0\u001b[24m198 \u001b[4m1\u001b[24m\u001b[4m1\u001b[24m766 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_211_Nanoarchaeota…\n",
      "\u001b[90m 7\u001b[39m CP0104… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m2\u001b[24m712 \u001b[4m1\u001b[24m\u001b[4m3\u001b[24m122 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_241_Nanoarchaeota…\n",
      "\u001b[90m 8\u001b[39m CP0104… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m4\u001b[24m216 \u001b[4m1\u001b[24m\u001b[4m4\u001b[24m374 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_270_Nanoarchaeota…\n",
      "\u001b[90m 9\u001b[39m CP0104… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m4\u001b[24m362 \u001b[4m1\u001b[24m\u001b[4m5\u001b[24m285 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_283_Nanoarchaeota…\n",
      "\u001b[90m10\u001b[39m CP0104… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m6\u001b[24m245 \u001b[4m1\u001b[24m\u001b[4m6\u001b[24m895 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_313_Nanoarchaeota…\n",
      "\u001b[90m# … with 1,001 more rows\u001b[39m\n",
      "\n",
      "[[2]]\n",
      "\u001b[90m# A tibble: 9 x 9\u001b[39m\n",
      "  seqid    source type  start   end score strand phase attributes               \n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                    \n",
      "\u001b[90m1\u001b[39m LKMP010… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m3\u001b[24m478 \u001b[4m2\u001b[24m\u001b[4m4\u001b[24m137 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_496_Nanohaloarcha…\n",
      "\u001b[90m2\u001b[39m LKMP010… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m2\u001b[24m639 \u001b[4m3\u001b[24m\u001b[4m4\u001b[24m075 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_672_Nanohaloarcha…\n",
      "\u001b[90m3\u001b[39m LKMP010… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m4\u001b[24m082 \u001b[4m3\u001b[24m\u001b[4m5\u001b[24m707 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_696_Nanohaloarcha…\n",
      "\u001b[90m4\u001b[39m LKMP010… getorf CDS   \u001b[4m4\u001b[24m\u001b[4m0\u001b[24m166 \u001b[4m4\u001b[24m\u001b[4m2\u001b[24m013 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_800_Nanohaloarcha…\n",
      "\u001b[90m5\u001b[39m LKMP010… getorf CDS   \u001b[4m4\u001b[24m\u001b[4m2\u001b[24m193 \u001b[4m4\u001b[24m\u001b[4m2\u001b[24m543 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_811_Nanohaloarcha…\n",
      "\u001b[90m6\u001b[39m LKMP010… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m6\u001b[24m109 \u001b[4m3\u001b[24m\u001b[4m6\u001b[24m600 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_947_Nanohaloarcha…\n",
      "\u001b[90m7\u001b[39m LKMP010… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m6\u001b[24m087 \u001b[4m2\u001b[24m\u001b[4m7\u001b[24m805 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_1150_Nanohaloarch…\n",
      "\u001b[90m8\u001b[39m LKMP010… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m9\u001b[24m628 \u001b[4m2\u001b[24m\u001b[4m1\u001b[24m415 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_1269_Nanohaloarch…\n",
      "\u001b[90m9\u001b[39m LKMP010… getorf CDS     999  \u001b[4m2\u001b[24m084 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_1578_Nanohaloarch…\n",
      "\n",
      "[[3]]\n",
      "\u001b[90m# A tibble: 12 x 9\u001b[39m\n",
      "   seqid    source type  start   end score strand phase attributes              \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                   \n",
      "\u001b[90m 1\u001b[39m QNAN010… getorf CDS     212  \u001b[4m1\u001b[24m318 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_20_Thermotogae_G…\n",
      "\u001b[90m 2\u001b[39m QNAN010… getorf CDS    \u001b[4m1\u001b[24m367  \u001b[4m1\u001b[24m576 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_24_Thermotogae_G…\n",
      "\u001b[90m 3\u001b[39m QNAN010… getorf CDS    \u001b[4m1\u001b[24m589  \u001b[4m3\u001b[24m391 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_50_Thermotogae_G…\n",
      "\u001b[90m 4\u001b[39m QNAN010… getorf CDS    \u001b[4m3\u001b[24m376  \u001b[4m3\u001b[24m669 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_56_Thermotogae_G…\n",
      "\u001b[90m 5\u001b[39m QNAN010… getorf CDS    \u001b[4m3\u001b[24m673  \u001b[4m4\u001b[24m815 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_74_Thermotogae_G…\n",
      "\u001b[90m 6\u001b[39m QNAN010… getorf CDS    \u001b[4m4\u001b[24m818  \u001b[4m5\u001b[24m573 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_87_Thermotogae_G…\n",
      "\u001b[90m 7\u001b[39m QNAN010… getorf CDS    \u001b[4m5\u001b[24m609  \u001b[4m5\u001b[24m851 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_93_Thermotogae_G…\n",
      "\u001b[90m 8\u001b[39m QNAN010… getorf CDS    \u001b[4m5\u001b[24m856  \u001b[4m6\u001b[24m188 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_98_Thermotogae_G…\n",
      "\u001b[90m 9\u001b[39m QNAN010… getorf CDS    \u001b[4m6\u001b[24m175  \u001b[4m6\u001b[24m750 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_109_Thermotogae_…\n",
      "\u001b[90m10\u001b[39m QNAN010… getorf CDS    \u001b[4m6\u001b[24m720  \u001b[4m6\u001b[24m923 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_111_Thermotogae_…\n",
      "\u001b[90m11\u001b[39m QNAN010… getorf CDS    \u001b[4m7\u001b[24m623  \u001b[4m7\u001b[24m958 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_133_Thermotogae_…\n",
      "\u001b[90m12\u001b[39m QNAN010… getorf CDS    \u001b[4m6\u001b[24m946  \u001b[4m7\u001b[24m515 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_151_Thermotogae_…\n",
      "\n",
      "[[4]]\n",
      "\u001b[90m# A tibble: 49 x 9\u001b[39m\n",
      "   seqid    source type  start   end score strand phase attributes              \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                   \n",
      "\u001b[90m 1\u001b[39m QNDS010… getorf CDS      15   308 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_5_Fermentibacter…\n",
      "\u001b[90m 2\u001b[39m QNDS010… getorf CDS    \u001b[4m2\u001b[24m862  \u001b[4m3\u001b[24m341 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_68_Fermentibacte…\n",
      "\u001b[90m 3\u001b[39m QNDS010… getorf CDS    \u001b[4m3\u001b[24m257  \u001b[4m3\u001b[24m742 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_73_Fermentibacte…\n",
      "\u001b[90m 4\u001b[39m QNDS010… getorf CDS    \u001b[4m3\u001b[24m726  \u001b[4m4\u001b[24m502 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_83_Fermentibacte…\n",
      "\u001b[90m 5\u001b[39m QNDS010… getorf CDS    \u001b[4m4\u001b[24m468  \u001b[4m4\u001b[24m824 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_86_Fermentibacte…\n",
      "\u001b[90m 6\u001b[39m QNDS010… getorf CDS    \u001b[4m4\u001b[24m824  \u001b[4m5\u001b[24m033 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_90_Fermentibacte…\n",
      "\u001b[90m 7\u001b[39m QNDS010… getorf CDS    \u001b[4m4\u001b[24m964  \u001b[4m5\u001b[24m332 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_95_Fermentibacte…\n",
      "\u001b[90m 8\u001b[39m QNDS010… getorf CDS    \u001b[4m5\u001b[24m340  \u001b[4m5\u001b[24m624 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_101_Fermentibact…\n",
      "\u001b[90m 9\u001b[39m QNDS010… getorf CDS    \u001b[4m5\u001b[24m608  \u001b[4m6\u001b[24m489 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_111_Fermentibact…\n",
      "\u001b[90m10\u001b[39m QNDS010… getorf CDS    \u001b[4m6\u001b[24m393  \u001b[4m6\u001b[24m710 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_116_Fermentibact…\n",
      "\u001b[90m# … with 39 more rows\u001b[39m\n",
      "\n",
      "[[5]]\n",
      "\u001b[90m# A tibble: 3 x 9\u001b[39m\n",
      "  seqid    source type  start   end score strand phase attributes               \n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                    \n",
      "\u001b[90m1\u001b[39m QMWF010… getorf CDS    \u001b[4m2\u001b[24m082  \u001b[4m2\u001b[24m894 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_82_Candidatus_Geo…\n",
      "\u001b[90m2\u001b[39m QMWF010… getorf CDS    \u001b[4m1\u001b[24m705  \u001b[4m1\u001b[24m998 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_90_Candidatus_Geo…\n",
      "\u001b[90m3\u001b[39m QMWF010… getorf CDS       2  \u001b[4m1\u001b[24m705 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_119_Candidatus_Ge…\n",
      "\n",
      "[[6]]\n",
      "\u001b[90m# A tibble: 2 x 9\u001b[39m\n",
      "  seqid    source type  start   end score strand phase attributes               \n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                    \n",
      "\u001b[90m1\u001b[39m QMXF010… getorf CDS      12   821 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_15_Candidatus_Bat…\n",
      "\u001b[90m2\u001b[39m QMXF010… getorf CDS     818  \u001b[4m2\u001b[24m272 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_40_Candidatus_Bat…\n",
      "\n",
      "[[7]]\n",
      "\u001b[90m# A tibble: 6 x 9\u001b[39m\n",
      "  seqid    source type  start   end score strand phase attributes               \n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                    \n",
      "\u001b[90m1\u001b[39m QMYS010… getorf CDS    \u001b[4m4\u001b[24m121  \u001b[4m4\u001b[24m450 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_102_Candidatus_Th…\n",
      "\u001b[90m2\u001b[39m QMYS010… getorf CDS    \u001b[4m4\u001b[24m440  \u001b[4m4\u001b[24m814 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_117_Candidatus_Th…\n",
      "\u001b[90m3\u001b[39m QMYS010… getorf CDS    \u001b[4m3\u001b[24m009  \u001b[4m4\u001b[24m118 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_141_Candidatus_Th…\n",
      "\u001b[90m4\u001b[39m QMYS010… getorf CDS    \u001b[4m1\u001b[24m993  \u001b[4m3\u001b[24m006 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_155_Candidatus_Th…\n",
      "\u001b[90m5\u001b[39m QMYS010… getorf CDS    \u001b[4m1\u001b[24m686  \u001b[4m1\u001b[24m976 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_161_Candidatus_Th…\n",
      "\u001b[90m6\u001b[39m QMYS010… getorf CDS      38  \u001b[4m1\u001b[24m696 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_187_Candidatus_Th…\n",
      "\n",
      "[[8]]\n",
      "\u001b[90m# A tibble: 13 x 9\u001b[39m\n",
      "   seqid    source type  start   end score strand phase attributes              \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                   \n",
      "\u001b[90m 1\u001b[39m DRBS010… getorf CDS     363  \u001b[4m1\u001b[24m469 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_27_Proteobacteri…\n",
      "\u001b[90m 2\u001b[39m DRBS010… getorf CDS    \u001b[4m1\u001b[24m494  \u001b[4m3\u001b[24m395 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_58_Proteobacteri…\n",
      "\u001b[90m 3\u001b[39m DRBS010… getorf CDS    \u001b[4m3\u001b[24m464  \u001b[4m4\u001b[24m429 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_76_Proteobacteri…\n",
      "\u001b[90m 4\u001b[39m DRBS010… getorf CDS    \u001b[4m4\u001b[24m593  \u001b[4m5\u001b[24m123 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_87_Proteobacteri…\n",
      "\u001b[90m 5\u001b[39m DRBS010… getorf CDS    \u001b[4m6\u001b[24m182  \u001b[4m6\u001b[24m541 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_112_Proteobacter…\n",
      "\u001b[90m 6\u001b[39m DRBS010… getorf CDS    \u001b[4m6\u001b[24m575  \u001b[4m7\u001b[24m045 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_121_Proteobacter…\n",
      "\u001b[90m 7\u001b[39m DRBS010… getorf CDS    \u001b[4m7\u001b[24m360  \u001b[4m7\u001b[24m959 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_139_Proteobacter…\n",
      "\u001b[90m 8\u001b[39m DRBS010… getorf CDS    \u001b[4m7\u001b[24m937 \u001b[4m1\u001b[24m\u001b[4m1\u001b[24m827 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_199_Proteobacter…\n",
      "\u001b[90m 9\u001b[39m DRBS010… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m1\u001b[24m845 \u001b[4m1\u001b[24m\u001b[4m2\u001b[24m156 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_205_Proteobacter…\n",
      "\u001b[90m10\u001b[39m DRBS010… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m2\u001b[24m316 \u001b[4m1\u001b[24m\u001b[4m2\u001b[24m609 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_216_Proteobacter…\n",
      "\u001b[90m11\u001b[39m DRBS010… getorf CDS    \u001b[4m5\u001b[24m764  \u001b[4m6\u001b[24m126 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_372_Proteobacter…\n",
      "\u001b[90m12\u001b[39m DRBS010… getorf CDS    \u001b[4m5\u001b[24m126  \u001b[4m5\u001b[24m782 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_383_Proteobacter…\n",
      "\u001b[90m13\u001b[39m DRBS010… getorf CDS      31   360 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_506_Proteobacter…\n",
      "\n",
      "[[9]]\n",
      "\u001b[90m# A tibble: 31 x 9\u001b[39m\n",
      "   seqid    source type  start   end score strand phase attributes              \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                   \n",
      "\u001b[90m 1\u001b[39m DQXS010… getorf CDS     212  \u001b[4m1\u001b[24m318 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_20_Altarchaeota_…\n",
      "\u001b[90m 2\u001b[39m DQXS010… getorf CDS    \u001b[4m1\u001b[24m367  \u001b[4m1\u001b[24m576 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_24_Altarchaeota_…\n",
      "\u001b[90m 3\u001b[39m DQXS010… getorf CDS    \u001b[4m1\u001b[24m589  \u001b[4m3\u001b[24m391 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_50_Altarchaeota_…\n",
      "\u001b[90m 4\u001b[39m DQXS010… getorf CDS    \u001b[4m3\u001b[24m376  \u001b[4m3\u001b[24m669 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_56_Altarchaeota_…\n",
      "\u001b[90m 5\u001b[39m DQXS010… getorf CDS    \u001b[4m3\u001b[24m673  \u001b[4m4\u001b[24m815 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_74_Altarchaeota_…\n",
      "\u001b[90m 6\u001b[39m DQXS010… getorf CDS    \u001b[4m4\u001b[24m818  \u001b[4m5\u001b[24m573 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_87_Altarchaeota_…\n",
      "\u001b[90m 7\u001b[39m DQXS010… getorf CDS    \u001b[4m5\u001b[24m578  \u001b[4m5\u001b[24m688 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_88_Altarchaeota_…\n",
      "\u001b[90m 8\u001b[39m DQXS010… getorf CDS    \u001b[4m5\u001b[24m609  \u001b[4m5\u001b[24m851 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_93_Altarchaeota_…\n",
      "\u001b[90m 9\u001b[39m DQXS010… getorf CDS    \u001b[4m5\u001b[24m856  \u001b[4m6\u001b[24m188 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_98_Altarchaeota_…\n",
      "\u001b[90m10\u001b[39m DQXS010… getorf CDS    \u001b[4m6\u001b[24m175  \u001b[4m6\u001b[24m750 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_109_Altarchaeota…\n",
      "\u001b[90m# … with 21 more rows\u001b[39m\n",
      "\n",
      "[[10]]\n",
      "\u001b[90m# A tibble: 3 x 9\u001b[39m\n",
      "  seqid    source type  start   end score strand phase attributes               \n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                    \n",
      "\u001b[90m1\u001b[39m DQZJ010… getorf CDS     246   512 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_8_Candidatus_Aeni…\n",
      "\u001b[90m2\u001b[39m DQZJ010… getorf CDS     561  \u001b[4m1\u001b[24m649 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_26_Candidatus_Aen…\n",
      "\u001b[90m3\u001b[39m DQZJ010… getorf CDS    \u001b[4m1\u001b[24m729  \u001b[4m3\u001b[24m345 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_58_Candidatus_Aen…\n",
      "\n",
      "[[11]]\n",
      "\u001b[90m# A tibble: 7 x 9\u001b[39m\n",
      "  seqid    source type  start   end score strand phase attributes               \n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                    \n",
      "\u001b[90m1\u001b[39m DTNV010… getorf CDS    \u001b[4m3\u001b[24m645  \u001b[4m3\u001b[24m905 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_81_Micrarchaeota_…\n",
      "\u001b[90m2\u001b[39m DTNV010… getorf CDS    \u001b[4m3\u001b[24m925  \u001b[4m4\u001b[24m320 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_86_Micrarchaeota_…\n",
      "\u001b[90m3\u001b[39m DTNV010… getorf CDS    \u001b[4m4\u001b[24m452  \u001b[4m4\u001b[24m775 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_96_Micrarchaeota_…\n",
      "\u001b[90m4\u001b[39m DTNV010… getorf CDS    \u001b[4m5\u001b[24m237  \u001b[4m5\u001b[24m653 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_127_Micrarchaeota…\n",
      "\u001b[90m5\u001b[39m DTNV010… getorf CDS    \u001b[4m4\u001b[24m849  \u001b[4m5\u001b[24m253 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_131_Micrarchaeota…\n",
      "\u001b[90m6\u001b[39m DTNV010… getorf CDS    \u001b[4m2\u001b[24m144  \u001b[4m3\u001b[24m193 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_175_Micrarchaeota…\n",
      "\u001b[90m7\u001b[39m DTNV010… getorf CDS     163  \u001b[4m2\u001b[24m016 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_195_Micrarchaeota…\n",
      "\n",
      "[[12]]\n",
      "\u001b[90m# A tibble: 40 x 9\u001b[39m\n",
      "   seqid    source type  start   end score strand phase attributes              \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                   \n",
      "\u001b[90m 1\u001b[39m DRYJ010… getorf CDS     110   703 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_12_Thermoproteot…\n",
      "\u001b[90m 2\u001b[39m DRYJ010… getorf CDS     716  \u001b[4m1\u001b[24m348 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_24_Thermoproteot…\n",
      "\u001b[90m 3\u001b[39m DRYJ010… getorf CDS    \u001b[4m1\u001b[24m382  \u001b[4m1\u001b[24m711 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_31_Thermoproteot…\n",
      "\u001b[90m 4\u001b[39m DRYJ010… getorf CDS    \u001b[4m1\u001b[24m701  \u001b[4m1\u001b[24m967 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_35_Thermoproteot…\n",
      "\u001b[90m 5\u001b[39m DRYJ010… getorf CDS    \u001b[4m1\u001b[24m913  \u001b[4m2\u001b[24m236 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_37_Thermoproteot…\n",
      "\u001b[90m 6\u001b[39m DRYJ010… getorf CDS    \u001b[4m2\u001b[24m361  \u001b[4m3\u001b[24m383 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_56_Thermoproteot…\n",
      "\u001b[90m 7\u001b[39m DRYJ010… getorf CDS    \u001b[4m3\u001b[24m430  \u001b[4m3\u001b[24m723 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_62_Thermoproteot…\n",
      "\u001b[90m 8\u001b[39m DRYJ010… getorf CDS    \u001b[4m3\u001b[24m723  \u001b[4m5\u001b[24m576 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_92_Thermoproteot…\n",
      "\u001b[90m 9\u001b[39m DRYJ010… getorf CDS    \u001b[4m5\u001b[24m576  \u001b[4m6\u001b[24m169 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_101_Thermoproteo…\n",
      "\u001b[90m10\u001b[39m DRYJ010… getorf CDS    \u001b[4m6\u001b[24m174  \u001b[4m6\u001b[24m404 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_105_Thermoproteo…\n",
      "\u001b[90m# … with 30 more rows\u001b[39m\n",
      "\n",
      "[[13]]\n",
      "\u001b[90m# A tibble: 56 x 9\u001b[39m\n",
      "   seqid    source type  start   end score strand phase attributes              \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                   \n",
      "\u001b[90m 1\u001b[39m VEPM010… getorf CDS    \u001b[4m1\u001b[24m428  \u001b[4m2\u001b[24m501 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_46_Halobacteriot…\n",
      "\u001b[90m 2\u001b[39m VEPM010… getorf CDS    \u001b[4m3\u001b[24m011  \u001b[4m3\u001b[24m847 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_72_Halobacteriot…\n",
      "\u001b[90m 3\u001b[39m VEPM010… getorf CDS    \u001b[4m4\u001b[24m713  \u001b[4m4\u001b[24m958 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_95_Halobacteriot…\n",
      "\u001b[90m 4\u001b[39m VEPM010… getorf CDS    \u001b[4m4\u001b[24m958  \u001b[4m6\u001b[24m400 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_116_Halobacterio…\n",
      "\u001b[90m 5\u001b[39m VEPM010… getorf CDS    \u001b[4m7\u001b[24m170  \u001b[4m8\u001b[24m090 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_150_Halobacterio…\n",
      "\u001b[90m 6\u001b[39m VEPM010… getorf CDS    \u001b[4m9\u001b[24m291  \u001b[4m9\u001b[24m656 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_179_Halobacterio…\n",
      "\u001b[90m 7\u001b[39m VEPM010… getorf CDS    \u001b[4m9\u001b[24m678 \u001b[4m1\u001b[24m\u001b[4m0\u001b[24m217 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_185_Halobacterio…\n",
      "\u001b[90m 8\u001b[39m VEPM010… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m0\u001b[24m283 \u001b[4m1\u001b[24m\u001b[4m2\u001b[24m343 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_220_Halobacterio…\n",
      "\u001b[90m 9\u001b[39m VEPM010… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m9\u001b[24m344 \u001b[4m1\u001b[24m\u001b[4m9\u001b[24m631 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_370_Halobacterio…\n",
      "\u001b[90m10\u001b[39m VEPM010… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m9\u001b[24m653 \u001b[4m1\u001b[24m\u001b[4m9\u001b[24m877 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_376_Halobacterio…\n",
      "\u001b[90m# … with 46 more rows\u001b[39m\n",
      "\n",
      "[[14]]\n",
      "\u001b[90m# A tibble: 2 x 9\u001b[39m\n",
      "  seqid    source type  start   end score strand phase attributes               \n",
      "  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                    \n",
      "\u001b[90m1\u001b[39m DTWW010… getorf CDS     379  \u001b[4m2\u001b[24m214 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_28_Candidatus_Mic…\n",
      "\u001b[90m2\u001b[39m DTWW010… getorf CDS    \u001b[4m2\u001b[24m280  \u001b[4m2\u001b[24m651 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_43_Candidatus_Mic…\n",
      "\n",
      "[[15]]\n",
      "\u001b[90m# A tibble: 22 x 9\u001b[39m\n",
      "   seqid    source type  start   end score strand phase attributes              \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                   \n",
      "\u001b[90m 1\u001b[39m JAADFW0… getorf CDS    \u001b[4m2\u001b[24m726  \u001b[4m3\u001b[24m106 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_69_Bacteroidota_…\n",
      "\u001b[90m 2\u001b[39m JAADFW0… getorf CDS    \u001b[4m3\u001b[24m096  \u001b[4m3\u001b[24m734 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_79_Bacteroidota_…\n",
      "\u001b[90m 3\u001b[39m JAADFW0… getorf CDS    \u001b[4m3\u001b[24m718  \u001b[4m4\u001b[24m116 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_88_Bacteroidota_…\n",
      "\u001b[90m 4\u001b[39m JAADFW0… getorf CDS    \u001b[4m4\u001b[24m106  \u001b[4m4\u001b[24m444 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_94_Bacteroidota_…\n",
      "\u001b[90m 5\u001b[39m JAADFW0… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m6\u001b[24m408 \u001b[4m1\u001b[24m\u001b[4m6\u001b[24m602 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_414_Bacteroidota…\n",
      "\u001b[90m 6\u001b[39m JAADFW0… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m5\u001b[24m516 \u001b[4m1\u001b[24m\u001b[4m6\u001b[24m382 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_428_Bacteroidota…\n",
      "\u001b[90m 7\u001b[39m JAADFW0… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m5\u001b[24m056 \u001b[4m1\u001b[24m\u001b[4m5\u001b[24m529 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_433_Bacteroidota…\n",
      "\u001b[90m 8\u001b[39m JAADFW0… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m4\u001b[24m553 \u001b[4m1\u001b[24m\u001b[4m5\u001b[24m065 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_442_Bacteroidota…\n",
      "\u001b[90m 9\u001b[39m JAADFW0… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m1\u001b[24m767 \u001b[4m1\u001b[24m\u001b[4m4\u001b[24m592 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_484_Bacteroidota…\n",
      "\u001b[90m10\u001b[39m JAADFW0… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m0\u001b[24m965 \u001b[4m1\u001b[24m\u001b[4m1\u001b[24m708 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m -      1     ID=ORF_496_Bacteroidota…\n",
      "\u001b[90m# … with 12 more rows\u001b[39m\n",
      "\n",
      "[[16]]\n",
      "\u001b[90m# A tibble: 115 x 9\u001b[39m\n",
      "   seqid    source type  start   end score strand phase attributes              \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                   \n",
      "\u001b[90m 1\u001b[39m NZ_CP07… getorf CDS    \u001b[4m1\u001b[24m599  \u001b[4m1\u001b[24m784 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_29_Halovivax_sp_…\n",
      "\u001b[90m 2\u001b[39m NZ_CP07… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m3\u001b[24m366 \u001b[4m1\u001b[24m\u001b[4m3\u001b[24m902 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_230_Halovivax_sp…\n",
      "\u001b[90m 3\u001b[39m NZ_CP07… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m3\u001b[24m933 \u001b[4m1\u001b[24m\u001b[4m4\u001b[24m415 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_239_Halovivax_sp…\n",
      "\u001b[90m 4\u001b[39m NZ_CP07… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m6\u001b[24m959 \u001b[4m1\u001b[24m\u001b[4m7\u001b[24m354 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_278_Halovivax_sp…\n",
      "\u001b[90m 5\u001b[39m NZ_CP07… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m2\u001b[24m659 \u001b[4m2\u001b[24m\u001b[4m3\u001b[24m489 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_371_Halovivax_sp…\n",
      "\u001b[90m 6\u001b[39m NZ_CP07… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m3\u001b[24m645 \u001b[4m2\u001b[24m\u001b[4m4\u001b[24m646 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_384_Halovivax_sp…\n",
      "\u001b[90m 7\u001b[39m NZ_CP07… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m4\u001b[24m691 \u001b[4m2\u001b[24m\u001b[4m6\u001b[24m757 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_404_Halovivax_sp…\n",
      "\u001b[90m 8\u001b[39m NZ_CP07… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m0\u001b[24m149 \u001b[4m3\u001b[24m\u001b[4m0\u001b[24m616 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_457_Halovivax_sp…\n",
      "\u001b[90m 9\u001b[39m NZ_CP07… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m0\u001b[24m849 \u001b[4m3\u001b[24m\u001b[4m1\u001b[24m166 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_468_Halovivax_sp…\n",
      "\u001b[90m10\u001b[39m NZ_CP07… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m1\u001b[24m216 \u001b[4m3\u001b[24m\u001b[4m1\u001b[24m554 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_474_Halovivax_sp…\n",
      "\u001b[90m# … with 105 more rows\u001b[39m\n",
      "\n",
      "[[17]]\n",
      "\u001b[90m# A tibble: 59 x 9\u001b[39m\n",
      "   seqid    source type  start   end score strand phase attributes              \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                   \n",
      "\u001b[90m 1\u001b[39m LOEP010… getorf CDS    \u001b[4m3\u001b[24m513  \u001b[4m3\u001b[24m665 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_79_Haloferax_sp_…\n",
      "\u001b[90m 2\u001b[39m LOEP010… getorf CDS    \u001b[4m3\u001b[24m266  \u001b[4m3\u001b[24m814 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_80_Haloferax_sp_…\n",
      "\u001b[90m 3\u001b[39m LOEP010… getorf CDS    \u001b[4m3\u001b[24m861  \u001b[4m4\u001b[24m196 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_87_Haloferax_sp_…\n",
      "\u001b[90m 4\u001b[39m LOEP010… getorf CDS    \u001b[4m4\u001b[24m244  \u001b[4m4\u001b[24m480 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_93_Haloferax_sp_…\n",
      "\u001b[90m 5\u001b[39m LOEP010… getorf CDS    \u001b[4m4\u001b[24m487  \u001b[4m4\u001b[24m780 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_97_Haloferax_sp_…\n",
      "\u001b[90m 6\u001b[39m LOEP010… getorf CDS    \u001b[4m4\u001b[24m807  \u001b[4m5\u001b[24m445 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_107_Haloferax_sp…\n",
      "\u001b[90m 7\u001b[39m LOEP010… getorf CDS    \u001b[4m8\u001b[24m369 \u001b[4m1\u001b[24m\u001b[4m3\u001b[24m414 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_208_Haloferax_sp…\n",
      "\u001b[90m 8\u001b[39m LOEP010… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m3\u001b[24m303 \u001b[4m1\u001b[24m\u001b[4m3\u001b[24m836 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_214_Haloferax_sp…\n",
      "\u001b[90m 9\u001b[39m LOEP010… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m0\u001b[24m068 \u001b[4m2\u001b[24m\u001b[4m0\u001b[24m352 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_329_Haloferax_sp…\n",
      "\u001b[90m10\u001b[39m LOEP010… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m0\u001b[24m374 \u001b[4m2\u001b[24m\u001b[4m2\u001b[24m269 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_347_Haloferax_sp…\n",
      "\u001b[90m# … with 49 more rows\u001b[39m\n",
      "\n",
      "[[18]]\n",
      "\u001b[90m# A tibble: 71 x 9\u001b[39m\n",
      "   seqid   source type  start   end score strand phase attributes               \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                    \n",
      "\u001b[90m 1\u001b[39m JNCS01… getorf CDS      27   173 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_3_Natrinema_altun…\n",
      "\u001b[90m 2\u001b[39m JNCS01… getorf CDS     280  \u001b[4m1\u001b[24m293 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_18_Natrinema_altu…\n",
      "\u001b[90m 3\u001b[39m JNCS01… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m6\u001b[24m564 \u001b[4m1\u001b[24m\u001b[4m7\u001b[24m007 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_255_Natrinema_alt…\n",
      "\u001b[90m 4\u001b[39m JNCS01… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m7\u001b[24m053 \u001b[4m1\u001b[24m\u001b[4m7\u001b[24m202 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_256_Natrinema_alt…\n",
      "\u001b[90m 5\u001b[39m JNCS01… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m1\u001b[24m671 \u001b[4m2\u001b[24m\u001b[4m2\u001b[24m378 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_316_Natrinema_alt…\n",
      "\u001b[90m 6\u001b[39m JNCS01… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m6\u001b[24m755 \u001b[4m2\u001b[24m\u001b[4m6\u001b[24m898 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_384_Natrinema_alt…\n",
      "\u001b[90m 7\u001b[39m JNCS01… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m1\u001b[24m039 \u001b[4m3\u001b[24m\u001b[4m1\u001b[24m230 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_441_Natrinema_alt…\n",
      "\u001b[90m 8\u001b[39m JNCS01… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m5\u001b[24m509 \u001b[4m3\u001b[24m\u001b[4m5\u001b[24m922 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_504_Natrinema_alt…\n",
      "\u001b[90m 9\u001b[39m JNCS01… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m6\u001b[24m994 \u001b[4m3\u001b[24m\u001b[4m7\u001b[24m350 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_521_Natrinema_alt…\n",
      "\u001b[90m10\u001b[39m JNCS01… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m8\u001b[24m732 \u001b[4m4\u001b[24m\u001b[4m0\u001b[24m747 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_567_Natrinema_alt…\n",
      "\u001b[90m# … with 61 more rows\u001b[39m\n",
      "\n",
      "[[19]]\n",
      "\u001b[90m# A tibble: 110 x 9\u001b[39m\n",
      "   seqid    source type  start   end score strand phase attributes              \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m    \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                   \n",
      "\u001b[90m 1\u001b[39m JABURA0… getorf CDS    \u001b[4m2\u001b[24m253  \u001b[4m2\u001b[24m759 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_35_Haloterrigena…\n",
      "\u001b[90m 2\u001b[39m JABURA0… getorf CDS    \u001b[4m2\u001b[24m769  \u001b[4m3\u001b[24m110 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_39_Haloterrigena…\n",
      "\u001b[90m 3\u001b[39m JABURA0… getorf CDS    \u001b[4m3\u001b[24m117  \u001b[4m3\u001b[24m713 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_43_Haloterrigena…\n",
      "\u001b[90m 4\u001b[39m JABURA0… getorf CDS    \u001b[4m3\u001b[24m625  \u001b[4m4\u001b[24m680 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_54_Haloterrigena…\n",
      "\u001b[90m 5\u001b[39m JABURA0… getorf CDS    \u001b[4m7\u001b[24m331  \u001b[4m8\u001b[24m041 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_112_Haloterrigen…\n",
      "\u001b[90m 6\u001b[39m JABURA0… getorf CDS    \u001b[4m8\u001b[24m269  \u001b[4m8\u001b[24m478 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_120_Haloterrigen…\n",
      "\u001b[90m 7\u001b[39m JABURA0… getorf CDS    \u001b[4m8\u001b[24m411 \u001b[4m1\u001b[24m\u001b[4m0\u001b[24m225 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_141_Haloterrigen…\n",
      "\u001b[90m 8\u001b[39m JABURA0… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m0\u001b[24m225 \u001b[4m1\u001b[24m\u001b[4m0\u001b[24m764 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_148_Haloterrigen…\n",
      "\u001b[90m 9\u001b[39m JABURA0… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m0\u001b[24m764 \u001b[4m1\u001b[24m\u001b[4m1\u001b[24m519 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_156_Haloterrigen…\n",
      "\u001b[90m10\u001b[39m JABURA0… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m1\u001b[24m595 \u001b[4m1\u001b[24m\u001b[4m1\u001b[24m909 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_160_Haloterrigen…\n",
      "\u001b[90m# … with 100 more rows\u001b[39m\n",
      "\n",
      "[[20]]\n",
      "\u001b[90m# A tibble: 244 x 9\u001b[39m\n",
      "   seqid  source type  start   end score strand phase attributes                \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                     \n",
      "\u001b[90m 1\u001b[39m CP058… getorf CDS     457  \u001b[4m1\u001b[24m752 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_15_Halobonum_sp_NJ…\n",
      "\u001b[90m 2\u001b[39m CP058… getorf CDS    \u001b[4m1\u001b[24m772  \u001b[4m2\u001b[24m554 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_20_Halobonum_sp_NJ…\n",
      "\u001b[90m 3\u001b[39m CP058… getorf CDS    \u001b[4m2\u001b[24m865  \u001b[4m3\u001b[24m455 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_33_Halobonum_sp_NJ…\n",
      "\u001b[90m 4\u001b[39m CP058… getorf CDS    \u001b[4m3\u001b[24m694  \u001b[4m4\u001b[24m089 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_39_Halobonum_sp_NJ…\n",
      "\u001b[90m 5\u001b[39m CP058… getorf CDS    \u001b[4m4\u001b[24m169  \u001b[4m4\u001b[24m831 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_44_Halobonum_sp_NJ…\n",
      "\u001b[90m 6\u001b[39m CP058… getorf CDS    \u001b[4m4\u001b[24m944  \u001b[4m5\u001b[24m885 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_55_Halobonum_sp_NJ…\n",
      "\u001b[90m 7\u001b[39m CP058… getorf CDS    \u001b[4m6\u001b[24m084  \u001b[4m6\u001b[24m245 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_60_Halobonum_sp_NJ…\n",
      "\u001b[90m 8\u001b[39m CP058… getorf CDS    \u001b[4m6\u001b[24m943  \u001b[4m7\u001b[24m419 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_75_Halobonum_sp_NJ…\n",
      "\u001b[90m 9\u001b[39m CP058… getorf CDS    \u001b[4m8\u001b[24m562  \u001b[4m8\u001b[24m963 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_102_Halobonum_sp_N…\n",
      "\u001b[90m10\u001b[39m CP058… getorf CDS    \u001b[4m8\u001b[24m995 \u001b[4m1\u001b[24m\u001b[4m0\u001b[24m536 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_114_Halobonum_sp_N…\n",
      "\u001b[90m# … with 234 more rows\u001b[39m\n",
      "\n",
      "[[21]]\n",
      "\u001b[90m# A tibble: 122 x 9\u001b[39m\n",
      "   seqid  source type  start   end score strand phase attributes                \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                     \n",
      "\u001b[90m 1\u001b[39m CP048… getorf CDS     543   782 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_16_Halogeometricum…\n",
      "\u001b[90m 2\u001b[39m CP048… getorf CDS    \u001b[4m1\u001b[24m509  \u001b[4m1\u001b[24m697 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_35_Halogeometricum…\n",
      "\u001b[90m 3\u001b[39m CP048… getorf CDS    \u001b[4m5\u001b[24m958  \u001b[4m6\u001b[24m470 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_114_Halogeometricu…\n",
      "\u001b[90m 4\u001b[39m CP048… getorf CDS    \u001b[4m6\u001b[24m403  \u001b[4m7\u001b[24m437 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_124_Halogeometricu…\n",
      "\u001b[90m 5\u001b[39m CP048… getorf CDS    \u001b[4m7\u001b[24m245  \u001b[4m7\u001b[24m583 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_125_Halogeometricu…\n",
      "\u001b[90m 6\u001b[39m CP048… getorf CDS    \u001b[4m7\u001b[24m589  \u001b[4m8\u001b[24m014 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_130_Halogeometricu…\n",
      "\u001b[90m 7\u001b[39m CP048… getorf CDS    \u001b[4m8\u001b[24m391  \u001b[4m8\u001b[24m765 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_144_Halogeometricu…\n",
      "\u001b[90m 8\u001b[39m CP048… getorf CDS    \u001b[4m8\u001b[24m765  \u001b[4m8\u001b[24m926 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_148_Halogeometricu…\n",
      "\u001b[90m 9\u001b[39m CP048… getorf CDS    \u001b[4m8\u001b[24m926  \u001b[4m9\u001b[24m144 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_150_Halogeometricu…\n",
      "\u001b[90m10\u001b[39m CP048… getorf CDS    \u001b[4m8\u001b[24m994  \u001b[4m9\u001b[24m332 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_154_Halogeometricu…\n",
      "\u001b[90m# … with 112 more rows\u001b[39m\n",
      "\n",
      "[[22]]\n",
      "\u001b[90m# A tibble: 88 x 9\u001b[39m\n",
      "   seqid   source type  start   end score strand phase attributes               \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                    \n",
      "\u001b[90m 1\u001b[39m ATYM01… getorf CDS    \u001b[4m1\u001b[24m643  \u001b[4m2\u001b[24m059 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_33_Haloplanus_nat…\n",
      "\u001b[90m 2\u001b[39m ATYM01… getorf CDS    \u001b[4m2\u001b[24m059  \u001b[4m2\u001b[24m319 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_37_Haloplanus_nat…\n",
      "\u001b[90m 3\u001b[39m ATYM01… getorf CDS    \u001b[4m5\u001b[24m553  \u001b[4m5\u001b[24m849 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_103_Haloplanus_na…\n",
      "\u001b[90m 4\u001b[39m ATYM01… getorf CDS    \u001b[4m8\u001b[24m221  \u001b[4m9\u001b[24m219 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_152_Haloplanus_na…\n",
      "\u001b[90m 5\u001b[39m ATYM01… getorf CDS    \u001b[4m9\u001b[24m920 \u001b[4m1\u001b[24m\u001b[4m0\u001b[24m939 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_177_Haloplanus_na…\n",
      "\u001b[90m 6\u001b[39m ATYM01… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m4\u001b[24m415 \u001b[4m1\u001b[24m\u001b[4m6\u001b[24m322 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_258_Haloplanus_na…\n",
      "\u001b[90m 7\u001b[39m ATYM01… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m7\u001b[24m192 \u001b[4m1\u001b[24m\u001b[4m7\u001b[24m338 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_278_Haloplanus_na…\n",
      "\u001b[90m 8\u001b[39m ATYM01… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m8\u001b[24m458 \u001b[4m1\u001b[24m\u001b[4m9\u001b[24m033 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_306_Haloplanus_na…\n",
      "\u001b[90m 9\u001b[39m ATYM01… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m0\u001b[24m981 \u001b[4m2\u001b[24m\u001b[4m2\u001b[24m696 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_358_Haloplanus_na…\n",
      "\u001b[90m10\u001b[39m ATYM01… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m2\u001b[24m696 \u001b[4m2\u001b[24m\u001b[4m3\u001b[24m199 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_362_Haloplanus_na…\n",
      "\u001b[90m# … with 78 more rows\u001b[39m\n",
      "\n",
      "[[23]]\n",
      "\u001b[90m# A tibble: 40 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga006571… getorf CDS     110   703 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_12_ga0065719_10…\n",
      "\u001b[90m 2\u001b[39m Ga006571… getorf CDS     716  \u001b[4m1\u001b[24m348 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_24_ga0065719_10…\n",
      "\u001b[90m 3\u001b[39m Ga006571… getorf CDS    \u001b[4m1\u001b[24m382  \u001b[4m1\u001b[24m711 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_31_ga0065719_10…\n",
      "\u001b[90m 4\u001b[39m Ga006571… getorf CDS    \u001b[4m1\u001b[24m701  \u001b[4m1\u001b[24m967 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_35_ga0065719_10…\n",
      "\u001b[90m 5\u001b[39m Ga006571… getorf CDS    \u001b[4m1\u001b[24m913  \u001b[4m2\u001b[24m236 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_37_ga0065719_10…\n",
      "\u001b[90m 6\u001b[39m Ga006571… getorf CDS    \u001b[4m2\u001b[24m361  \u001b[4m3\u001b[24m383 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_56_ga0065719_10…\n",
      "\u001b[90m 7\u001b[39m Ga006571… getorf CDS    \u001b[4m3\u001b[24m430  \u001b[4m3\u001b[24m723 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_62_ga0065719_10…\n",
      "\u001b[90m 8\u001b[39m Ga006571… getorf CDS    \u001b[4m3\u001b[24m723  \u001b[4m5\u001b[24m576 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_92_ga0065719_10…\n",
      "\u001b[90m 9\u001b[39m Ga006571… getorf CDS    \u001b[4m5\u001b[24m576  \u001b[4m6\u001b[24m169 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_101_ga0065719_1…\n",
      "\u001b[90m10\u001b[39m Ga006571… getorf CDS    \u001b[4m6\u001b[24m174  \u001b[4m6\u001b[24m404 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_105_ga0065719_1…\n",
      "\u001b[90m# … with 30 more rows\u001b[39m\n",
      "\n",
      "[[24]]\n",
      "\u001b[90m# A tibble: 49 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga007512… getorf CDS     114   434 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_6_ga0075120_100…\n",
      "\u001b[90m 2\u001b[39m Ga007512… getorf CDS     490   624 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_11_ga0075120_10…\n",
      "\u001b[90m 3\u001b[39m Ga007512… getorf CDS     620  \u001b[4m2\u001b[24m047 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_36_ga0075120_10…\n",
      "\u001b[90m 4\u001b[39m Ga007512… getorf CDS    \u001b[4m2\u001b[24m044  \u001b[4m2\u001b[24m691 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_46_ga0075120_10…\n",
      "\u001b[90m 5\u001b[39m Ga007512… getorf CDS    \u001b[4m2\u001b[24m694  \u001b[4m3\u001b[24m293 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_54_ga0075120_10…\n",
      "\u001b[90m 6\u001b[39m Ga007512… getorf CDS    \u001b[4m3\u001b[24m289  \u001b[4m3\u001b[24m945 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_65_ga0075120_10…\n",
      "\u001b[90m 7\u001b[39m Ga007512… getorf CDS    \u001b[4m3\u001b[24m986  \u001b[4m4\u001b[24m858 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_79_ga0075120_10…\n",
      "\u001b[90m 8\u001b[39m Ga007512… getorf CDS    \u001b[4m4\u001b[24m922  \u001b[4m5\u001b[24m164 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_84_ga0075120_10…\n",
      "\u001b[90m 9\u001b[39m Ga007512… getorf CDS    \u001b[4m5\u001b[24m143  \u001b[4m8\u001b[24m358 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_133_ga0075120_1…\n",
      "\u001b[90m10\u001b[39m Ga007512… getorf CDS    \u001b[4m8\u001b[24m413  \u001b[4m9\u001b[24m564 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_151_ga0075120_1…\n",
      "\u001b[90m# … with 39 more rows\u001b[39m\n",
      "\n",
      "[[25]]\n",
      "\u001b[90m# A tibble: 52 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga007512… getorf CDS      20   289 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_4_ga0075122_100…\n",
      "\u001b[90m 2\u001b[39m Ga007512… getorf CDS     289   477 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_8_ga0075122_100…\n",
      "\u001b[90m 3\u001b[39m Ga007512… getorf CDS     509  \u001b[4m1\u001b[24m129 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_17_ga0075122_10…\n",
      "\u001b[90m 4\u001b[39m Ga007512… getorf CDS    \u001b[4m1\u001b[24m129  \u001b[4m1\u001b[24m620 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_25_ga0075122_10…\n",
      "\u001b[90m 5\u001b[39m Ga007512… getorf CDS    \u001b[4m1\u001b[24m583  \u001b[4m1\u001b[24m789 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_28_ga0075122_10…\n",
      "\u001b[90m 6\u001b[39m Ga007512… getorf CDS    \u001b[4m1\u001b[24m817  \u001b[4m2\u001b[24m335 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_38_ga0075122_10…\n",
      "\u001b[90m 7\u001b[39m Ga007512… getorf CDS    \u001b[4m3\u001b[24m129  \u001b[4m3\u001b[24m341 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_59_ga0075122_10…\n",
      "\u001b[90m 8\u001b[39m Ga007512… getorf CDS    \u001b[4m9\u001b[24m760 \u001b[4m1\u001b[24m\u001b[4m0\u001b[24m950 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_230_ga0075122_1…\n",
      "\u001b[90m 9\u001b[39m Ga007512… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m0\u001b[24m946 \u001b[4m1\u001b[24m\u001b[4m1\u001b[24m404 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_237_ga0075122_1…\n",
      "\u001b[90m10\u001b[39m Ga007512… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m1\u001b[24m379 \u001b[4m1\u001b[24m\u001b[4m1\u001b[24m564 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_241_ga0075122_1…\n",
      "\u001b[90m# … with 42 more rows\u001b[39m\n",
      "\n",
      "[[26]]\n",
      "\u001b[90m# A tibble: 40 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga007512… getorf CDS      27   116 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_2_ga0075122_100…\n",
      "\u001b[90m 2\u001b[39m Ga007512… getorf CDS     123   248 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_6_ga0075122_100…\n",
      "\u001b[90m 3\u001b[39m Ga007512… getorf CDS     224   487 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_10_ga0075122_10…\n",
      "\u001b[90m 4\u001b[39m Ga007512… getorf CDS     512   655 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_14_ga0075122_10…\n",
      "\u001b[90m 5\u001b[39m Ga007512… getorf CDS     655  \u001b[4m1\u001b[24m191 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_20_ga0075122_10…\n",
      "\u001b[90m 6\u001b[39m Ga007512… getorf CDS    \u001b[4m1\u001b[24m160  \u001b[4m1\u001b[24m966 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_29_ga0075122_10…\n",
      "\u001b[90m 7\u001b[39m Ga007512… getorf CDS    \u001b[4m2\u001b[24m306  \u001b[4m4\u001b[24m153 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_61_ga0075122_10…\n",
      "\u001b[90m 8\u001b[39m Ga007512… getorf CDS    \u001b[4m4\u001b[24m143  \u001b[4m5\u001b[24m252 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_73_ga0075122_10…\n",
      "\u001b[90m 9\u001b[39m Ga007512… getorf CDS    \u001b[4m5\u001b[24m318  \u001b[4m5\u001b[24m629 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_79_ga0075122_10…\n",
      "\u001b[90m10\u001b[39m Ga007512… getorf CDS    \u001b[4m5\u001b[24m623  \u001b[4m6\u001b[24m327 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_87_ga0075122_10…\n",
      "\u001b[90m# … with 30 more rows\u001b[39m\n",
      "\n",
      "[[27]]\n",
      "\u001b[90m# A tibble: 33 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga007512… getorf CDS      17   283 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_4_ga0075122_100…\n",
      "\u001b[90m 2\u001b[39m Ga007512… getorf CDS    \u001b[4m1\u001b[24m950  \u001b[4m2\u001b[24m933 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_58_ga0075122_10…\n",
      "\u001b[90m 3\u001b[39m Ga007512… getorf CDS    \u001b[4m3\u001b[24m093  \u001b[4m4\u001b[24m886 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_87_ga0075122_10…\n",
      "\u001b[90m 4\u001b[39m Ga007512… getorf CDS    \u001b[4m4\u001b[24m886  \u001b[4m5\u001b[24m143 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_89_ga0075122_10…\n",
      "\u001b[90m 5\u001b[39m Ga007512… getorf CDS    \u001b[4m5\u001b[24m147  \u001b[4m6\u001b[24m172 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_103_ga0075122_1…\n",
      "\u001b[90m 6\u001b[39m Ga007512… getorf CDS    \u001b[4m6\u001b[24m176  \u001b[4m6\u001b[24m604 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_107_ga0075122_1…\n",
      "\u001b[90m 7\u001b[39m Ga007512… getorf CDS    \u001b[4m6\u001b[24m762  \u001b[4m7\u001b[24m478 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_123_ga0075122_1…\n",
      "\u001b[90m 8\u001b[39m Ga007512… getorf CDS    \u001b[4m7\u001b[24m482  \u001b[4m7\u001b[24m601 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_125_ga0075122_1…\n",
      "\u001b[90m 9\u001b[39m Ga007512… getorf CDS    \u001b[4m7\u001b[24m601  \u001b[4m7\u001b[24m906 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_132_ga0075122_1…\n",
      "\u001b[90m10\u001b[39m Ga007512… getorf CDS    \u001b[4m7\u001b[24m902 \u001b[4m1\u001b[24m\u001b[4m0\u001b[24m757 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_172_ga0075122_1…\n",
      "\u001b[90m# … with 23 more rows\u001b[39m\n",
      "\n",
      "[[28]]\n",
      "\u001b[90m# A tibble: 39 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga007897… getorf CDS    \u001b[4m6\u001b[24m702  \u001b[4m6\u001b[24m875 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_158_ga0078972_1…\n",
      "\u001b[90m 2\u001b[39m Ga007897… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m3\u001b[24m616 \u001b[4m1\u001b[24m\u001b[4m4\u001b[24m029 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_308_ga0078972_1…\n",
      "\u001b[90m 3\u001b[39m Ga007897… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m3\u001b[24m998 \u001b[4m1\u001b[24m\u001b[4m5\u001b[24m824 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_334_ga0078972_1…\n",
      "\u001b[90m 4\u001b[39m Ga007897… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m5\u001b[24m985 \u001b[4m1\u001b[24m\u001b[4m6\u001b[24m662 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_347_ga0078972_1…\n",
      "\u001b[90m 5\u001b[39m Ga007897… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m7\u001b[24m111 \u001b[4m1\u001b[24m\u001b[4m7\u001b[24m509 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_367_ga0078972_1…\n",
      "\u001b[90m 6\u001b[39m Ga007897… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m7\u001b[24m509 \u001b[4m1\u001b[24m\u001b[4m7\u001b[24m742 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_371_ga0078972_1…\n",
      "\u001b[90m 7\u001b[39m Ga007897… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m7\u001b[24m755 \u001b[4m1\u001b[24m\u001b[4m8\u001b[24m060 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_376_ga0078972_1…\n",
      "\u001b[90m 8\u001b[39m Ga007897… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m8\u001b[24m038 \u001b[4m1\u001b[24m\u001b[4m8\u001b[24m517 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_383_ga0078972_1…\n",
      "\u001b[90m 9\u001b[39m Ga007897… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m8\u001b[24m472 \u001b[4m1\u001b[24m\u001b[4m9\u001b[24m623 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_403_ga0078972_1…\n",
      "\u001b[90m10\u001b[39m Ga007897… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m9\u001b[24m610 \u001b[4m2\u001b[24m\u001b[4m0\u001b[24m242 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_410_ga0078972_1…\n",
      "\u001b[90m# … with 29 more rows\u001b[39m\n",
      "\n",
      "[[29]]\n",
      "\u001b[90m# A tibble: 31 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga011620… getorf CDS     212  \u001b[4m1\u001b[24m318 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_20_ga0116200_10…\n",
      "\u001b[90m 2\u001b[39m Ga011620… getorf CDS    \u001b[4m1\u001b[24m367  \u001b[4m1\u001b[24m576 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_24_ga0116200_10…\n",
      "\u001b[90m 3\u001b[39m Ga011620… getorf CDS    \u001b[4m1\u001b[24m589  \u001b[4m3\u001b[24m391 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_50_ga0116200_10…\n",
      "\u001b[90m 4\u001b[39m Ga011620… getorf CDS    \u001b[4m3\u001b[24m376  \u001b[4m3\u001b[24m669 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_56_ga0116200_10…\n",
      "\u001b[90m 5\u001b[39m Ga011620… getorf CDS    \u001b[4m3\u001b[24m673  \u001b[4m4\u001b[24m815 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_74_ga0116200_10…\n",
      "\u001b[90m 6\u001b[39m Ga011620… getorf CDS    \u001b[4m4\u001b[24m818  \u001b[4m5\u001b[24m573 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_87_ga0116200_10…\n",
      "\u001b[90m 7\u001b[39m Ga011620… getorf CDS    \u001b[4m5\u001b[24m578  \u001b[4m5\u001b[24m688 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_88_ga0116200_10…\n",
      "\u001b[90m 8\u001b[39m Ga011620… getorf CDS    \u001b[4m5\u001b[24m609  \u001b[4m5\u001b[24m851 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_93_ga0116200_10…\n",
      "\u001b[90m 9\u001b[39m Ga011620… getorf CDS    \u001b[4m5\u001b[24m856  \u001b[4m6\u001b[24m188 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_98_ga0116200_10…\n",
      "\u001b[90m10\u001b[39m Ga011620… getorf CDS    \u001b[4m6\u001b[24m175  \u001b[4m6\u001b[24m750 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_109_ga0116200_1…\n",
      "\u001b[90m# … with 21 more rows\u001b[39m\n",
      "\n",
      "[[30]]\n",
      "\u001b[90m# A tibble: 49 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga016431… getorf CDS      15   308 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_5_ga0164313_100…\n",
      "\u001b[90m 2\u001b[39m Ga016431… getorf CDS    \u001b[4m2\u001b[24m862  \u001b[4m3\u001b[24m341 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_68_ga0164313_10…\n",
      "\u001b[90m 3\u001b[39m Ga016431… getorf CDS    \u001b[4m3\u001b[24m257  \u001b[4m3\u001b[24m742 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_73_ga0164313_10…\n",
      "\u001b[90m 4\u001b[39m Ga016431… getorf CDS    \u001b[4m3\u001b[24m726  \u001b[4m4\u001b[24m502 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_83_ga0164313_10…\n",
      "\u001b[90m 5\u001b[39m Ga016431… getorf CDS    \u001b[4m4\u001b[24m468  \u001b[4m4\u001b[24m824 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_86_ga0164313_10…\n",
      "\u001b[90m 6\u001b[39m Ga016431… getorf CDS    \u001b[4m4\u001b[24m824  \u001b[4m5\u001b[24m033 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_90_ga0164313_10…\n",
      "\u001b[90m 7\u001b[39m Ga016431… getorf CDS    \u001b[4m4\u001b[24m964  \u001b[4m5\u001b[24m332 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_95_ga0164313_10…\n",
      "\u001b[90m 8\u001b[39m Ga016431… getorf CDS    \u001b[4m5\u001b[24m340  \u001b[4m5\u001b[24m624 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_101_ga0164313_1…\n",
      "\u001b[90m 9\u001b[39m Ga016431… getorf CDS    \u001b[4m5\u001b[24m608  \u001b[4m6\u001b[24m489 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_111_ga0164313_1…\n",
      "\u001b[90m10\u001b[39m Ga016431… getorf CDS    \u001b[4m6\u001b[24m393  \u001b[4m6\u001b[24m710 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_116_ga0164313_1…\n",
      "\u001b[90m# … with 39 more rows\u001b[39m\n",
      "\n",
      "[[31]]\n",
      "\u001b[90m# A tibble: 132 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga017237… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m2\u001b[24m465 \u001b[4m3\u001b[24m\u001b[4m2\u001b[24m626 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_749_ga0172377_1…\n",
      "\u001b[90m 2\u001b[39m Ga017237… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m2\u001b[24m630 \u001b[4m3\u001b[24m\u001b[4m3\u001b[24m136 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_757_ga0172377_1…\n",
      "\u001b[90m 3\u001b[39m Ga017237… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m3\u001b[24m132 \u001b[4m3\u001b[24m\u001b[4m3\u001b[24m782 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_766_ga0172377_1…\n",
      "\u001b[90m 4\u001b[39m Ga017237… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m3\u001b[24m786 \u001b[4m3\u001b[24m\u001b[4m4\u001b[24m007 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_768_ga0172377_1…\n",
      "\u001b[90m 5\u001b[39m Ga017237… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m4\u001b[24m011 \u001b[4m3\u001b[24m\u001b[4m4\u001b[24m259 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_774_ga0172377_1…\n",
      "\u001b[90m 6\u001b[39m Ga017237… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m4\u001b[24m275 \u001b[4m3\u001b[24m\u001b[4m4\u001b[24m508 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_779_ga0172377_1…\n",
      "\u001b[90m 7\u001b[39m Ga017237… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m4\u001b[24m531 \u001b[4m3\u001b[24m\u001b[4m4\u001b[24m803 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_786_ga0172377_1…\n",
      "\u001b[90m 8\u001b[39m Ga017237… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m4\u001b[24m884 \u001b[4m3\u001b[24m\u001b[4m5\u001b[24m963 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_802_ga0172377_1…\n",
      "\u001b[90m 9\u001b[39m Ga017237… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m5\u001b[24m974 \u001b[4m3\u001b[24m\u001b[4m6\u001b[24m885 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_818_ga0172377_1…\n",
      "\u001b[90m10\u001b[39m Ga017237… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m6\u001b[24m888 \u001b[4m3\u001b[24m\u001b[4m7\u001b[24m055 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_821_ga0172377_1…\n",
      "\u001b[90m# … with 122 more rows\u001b[39m\n",
      "\n",
      "[[32]]\n",
      "\u001b[90m# A tibble: 218 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga017237… getorf CDS       4   318 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_2_ga0172379_100…\n",
      "\u001b[90m 2\u001b[39m Ga017237… getorf CDS     173  \u001b[4m3\u001b[24m349 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_49_ga0172379_10…\n",
      "\u001b[90m 3\u001b[39m Ga017237… getorf CDS    \u001b[4m3\u001b[24m359  \u001b[4m3\u001b[24m625 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_56_ga0172379_10…\n",
      "\u001b[90m 4\u001b[39m Ga017237… getorf CDS    \u001b[4m3\u001b[24m821  \u001b[4m4\u001b[24m018 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_64_ga0172379_10…\n",
      "\u001b[90m 5\u001b[39m Ga017237… getorf CDS    \u001b[4m4\u001b[24m018  \u001b[4m5\u001b[24m310 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_87_ga0172379_10…\n",
      "\u001b[90m 6\u001b[39m Ga017237… getorf CDS    \u001b[4m5\u001b[24m259  \u001b[4m5\u001b[24m897 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_97_ga0172379_10…\n",
      "\u001b[90m 7\u001b[39m Ga017237… getorf CDS    \u001b[4m6\u001b[24m074  \u001b[4m6\u001b[24m409 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_104_ga0172379_1…\n",
      "\u001b[90m 8\u001b[39m Ga017237… getorf CDS    \u001b[4m6\u001b[24m501  \u001b[4m7\u001b[24m040 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_115_ga0172379_1…\n",
      "\u001b[90m 9\u001b[39m Ga017237… getorf CDS    \u001b[4m7\u001b[24m187  \u001b[4m8\u001b[24m485 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_139_ga0172379_1…\n",
      "\u001b[90m10\u001b[39m Ga017237… getorf CDS    \u001b[4m9\u001b[24m270 \u001b[4m1\u001b[24m\u001b[4m2\u001b[24m338 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_194_ga0172379_1…\n",
      "\u001b[90m# … with 208 more rows\u001b[39m\n",
      "\n",
      "[[33]]\n",
      "\u001b[90m# A tibble: 77 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga017237… getorf CDS      48   209 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_5_ga0172379_100…\n",
      "\u001b[90m 2\u001b[39m Ga017237… getorf CDS     222   428 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_7_ga0172379_100…\n",
      "\u001b[90m 3\u001b[39m Ga017237… getorf CDS     624  \u001b[4m1\u001b[24m139 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_21_ga0172379_10…\n",
      "\u001b[90m 4\u001b[39m Ga017237… getorf CDS    \u001b[4m1\u001b[24m195  \u001b[4m1\u001b[24m344 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_25_ga0172379_10…\n",
      "\u001b[90m 5\u001b[39m Ga017237… getorf CDS    \u001b[4m1\u001b[24m332  \u001b[4m1\u001b[24m862 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_31_ga0172379_10…\n",
      "\u001b[90m 6\u001b[39m Ga017237… getorf CDS    \u001b[4m2\u001b[24m179  \u001b[4m2\u001b[24m394 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_42_ga0172379_10…\n",
      "\u001b[90m 7\u001b[39m Ga017237… getorf CDS    \u001b[4m2\u001b[24m419  \u001b[4m2\u001b[24m943 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_53_ga0172379_10…\n",
      "\u001b[90m 8\u001b[39m Ga017237… getorf CDS    \u001b[4m2\u001b[24m960  \u001b[4m3\u001b[24m364 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_59_ga0172379_10…\n",
      "\u001b[90m 9\u001b[39m Ga017237… getorf CDS    \u001b[4m3\u001b[24m385  \u001b[4m4\u001b[24m815 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_79_ga0172379_10…\n",
      "\u001b[90m10\u001b[39m Ga017237… getorf CDS    \u001b[4m4\u001b[24m812  \u001b[4m5\u001b[24m531 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_90_ga0172379_10…\n",
      "\u001b[90m# … with 67 more rows\u001b[39m\n",
      "\n",
      "[[34]]\n",
      "\u001b[90m# A tibble: 84 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga020771… getorf CDS      15   968 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_5_ga0207715_100…\n",
      "\u001b[90m 2\u001b[39m Ga020771… getorf CDS    \u001b[4m1\u001b[24m014  \u001b[4m1\u001b[24m664 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_12_ga0207715_10…\n",
      "\u001b[90m 3\u001b[39m Ga020771… getorf CDS    \u001b[4m1\u001b[24m680  \u001b[4m2\u001b[24m366 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_19_ga0207715_10…\n",
      "\u001b[90m 4\u001b[39m Ga020771… getorf CDS    \u001b[4m4\u001b[24m092  \u001b[4m4\u001b[24m784 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_31_ga0207715_10…\n",
      "\u001b[90m 5\u001b[39m Ga020771… getorf CDS    \u001b[4m4\u001b[24m893  \u001b[4m5\u001b[24m453 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_38_ga0207715_10…\n",
      "\u001b[90m 6\u001b[39m Ga020771… getorf CDS    \u001b[4m5\u001b[24m502  \u001b[4m6\u001b[24m761 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_47_ga0207715_10…\n",
      "\u001b[90m 7\u001b[39m Ga020771… getorf CDS    \u001b[4m6\u001b[24m793  \u001b[4m7\u001b[24m443 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_50_ga0207715_10…\n",
      "\u001b[90m 8\u001b[39m Ga020771… getorf CDS    \u001b[4m7\u001b[24m443  \u001b[4m8\u001b[24m444 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_53_ga0207715_10…\n",
      "\u001b[90m 9\u001b[39m Ga020771… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m1\u001b[24m845 \u001b[4m1\u001b[24m\u001b[4m2\u001b[24m330 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_83_ga0207715_10…\n",
      "\u001b[90m10\u001b[39m Ga020771… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m2\u001b[24m992 \u001b[4m1\u001b[24m\u001b[4m3\u001b[24m717 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_93_ga0207715_10…\n",
      "\u001b[90m# … with 74 more rows\u001b[39m\n",
      "\n",
      "[[35]]\n",
      "\u001b[90m# A tibble: 150 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga020771… getorf CDS      85  \u001b[4m1\u001b[24m122 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_3_ga0207718_100…\n",
      "\u001b[90m 2\u001b[39m Ga020771… getorf CDS    \u001b[4m2\u001b[24m569  \u001b[4m2\u001b[24m655 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_18_ga0207718_10…\n",
      "\u001b[90m 3\u001b[39m Ga020771… getorf CDS    \u001b[4m7\u001b[24m216  \u001b[4m7\u001b[24m551 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_99_ga0207718_10…\n",
      "\u001b[90m 4\u001b[39m Ga020771… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m3\u001b[24m009 \u001b[4m1\u001b[24m\u001b[4m3\u001b[24m428 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_203_ga0207718_1…\n",
      "\u001b[90m 5\u001b[39m Ga020771… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m5\u001b[24m732 \u001b[4m1\u001b[24m\u001b[4m6\u001b[24m217 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_247_ga0207718_1…\n",
      "\u001b[90m 6\u001b[39m Ga020771… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m0\u001b[24m613 \u001b[4m2\u001b[24m\u001b[4m1\u001b[24m902 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_319_ga0207718_1…\n",
      "\u001b[90m 7\u001b[39m Ga020771… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m7\u001b[24m285 \u001b[4m2\u001b[24m\u001b[4m7\u001b[24m722 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_406_ga0207718_1…\n",
      "\u001b[90m 8\u001b[39m Ga020771… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m9\u001b[24m058 \u001b[4m2\u001b[24m\u001b[4m9\u001b[24m252 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_427_ga0207718_1…\n",
      "\u001b[90m 9\u001b[39m Ga020771… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m0\u001b[24m333 \u001b[4m3\u001b[24m\u001b[4m0\u001b[24m797 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_448_ga0207718_1…\n",
      "\u001b[90m10\u001b[39m Ga020771… getorf CDS   \u001b[4m3\u001b[24m\u001b[4m7\u001b[24m859 \u001b[4m3\u001b[24m\u001b[4m8\u001b[24m788 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_565_ga0207718_1…\n",
      "\u001b[90m# … with 140 more rows\u001b[39m\n",
      "\n",
      "[[36]]\n",
      "\u001b[90m# A tibble: 58 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga020771… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m8\u001b[24m807 \u001b[4m1\u001b[24m\u001b[4m9\u001b[24m235 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_352_ga0207719_1…\n",
      "\u001b[90m 2\u001b[39m Ga020771… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m9\u001b[24m894 \u001b[4m2\u001b[24m\u001b[4m0\u001b[24m307 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_371_ga0207719_1…\n",
      "\u001b[90m 3\u001b[39m Ga020771… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m0\u001b[24m320 \u001b[4m2\u001b[24m\u001b[4m1\u001b[24m468 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_385_ga0207719_1…\n",
      "\u001b[90m 4\u001b[39m Ga020771… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m1\u001b[24m468 \u001b[4m2\u001b[24m\u001b[4m4\u001b[24m224 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_412_ga0207719_1…\n",
      "\u001b[90m 5\u001b[39m Ga020771… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m4\u001b[24m170 \u001b[4m2\u001b[24m\u001b[4m5\u001b[24m198 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_424_ga0207719_1…\n",
      "\u001b[90m 6\u001b[39m Ga020771… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m5\u001b[24m147 \u001b[4m2\u001b[24m\u001b[4m7\u001b[24m114 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_441_ga0207719_1…\n",
      "\u001b[90m 7\u001b[39m Ga020771… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m7\u001b[24m114 \u001b[4m2\u001b[24m\u001b[4m7\u001b[24m527 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_444_ga0207719_1…\n",
      "\u001b[90m 8\u001b[39m Ga020771… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m7\u001b[24m566 \u001b[4m2\u001b[24m\u001b[4m7\u001b[24m862 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_447_ga0207719_1…\n",
      "\u001b[90m 9\u001b[39m Ga020771… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m8\u001b[24m601 \u001b[4m2\u001b[24m\u001b[4m9\u001b[24m857 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_476_ga0207719_1…\n",
      "\u001b[90m10\u001b[39m Ga020771… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m9\u001b[24m802 \u001b[4m3\u001b[24m\u001b[4m0\u001b[24m134 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_479_ga0207719_1…\n",
      "\u001b[90m# … with 48 more rows\u001b[39m\n",
      "\n",
      "[[37]]\n",
      "\u001b[90m# A tibble: 36 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga020773… getorf CDS       4   549 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_3_ga0207733_100…\n",
      "\u001b[90m 2\u001b[39m Ga020773… getorf CDS     549  \u001b[4m1\u001b[24m661 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_12_ga0207733_10…\n",
      "\u001b[90m 3\u001b[39m Ga020773… getorf CDS    \u001b[4m1\u001b[24m627  \u001b[4m1\u001b[24m857 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_14_ga0207733_10…\n",
      "\u001b[90m 4\u001b[39m Ga020773… getorf CDS    \u001b[4m1\u001b[24m784  \u001b[4m2\u001b[24m443 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_22_ga0207733_10…\n",
      "\u001b[90m 5\u001b[39m Ga020773… getorf CDS    \u001b[4m2\u001b[24m443  \u001b[4m2\u001b[24m982 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_28_ga0207733_10…\n",
      "\u001b[90m 6\u001b[39m Ga020773… getorf CDS    \u001b[4m5\u001b[24m720  \u001b[4m5\u001b[24m956 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_76_ga0207733_10…\n",
      "\u001b[90m 7\u001b[39m Ga020773… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m6\u001b[24m423 \u001b[4m1\u001b[24m\u001b[4m6\u001b[24m986 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_257_ga0207733_1…\n",
      "\u001b[90m 8\u001b[39m Ga020773… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m7\u001b[24m655 \u001b[4m1\u001b[24m\u001b[4m8\u001b[24m281 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_280_ga0207733_1…\n",
      "\u001b[90m 9\u001b[39m Ga020773… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m8\u001b[24m321 \u001b[4m1\u001b[24m\u001b[4m8\u001b[24m548 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_284_ga0207733_1…\n",
      "\u001b[90m10\u001b[39m Ga020773… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m8\u001b[24m494 \u001b[4m2\u001b[24m\u001b[4m0\u001b[24m233 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_307_ga0207733_1…\n",
      "\u001b[90m# … with 26 more rows\u001b[39m\n",
      "\n",
      "[[38]]\n",
      "\u001b[90m# A tibble: 184 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga020837… getorf CDS      27   596 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_10_ga0208375_10…\n",
      "\u001b[90m 2\u001b[39m Ga020837… getorf CDS     601   912 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_15_ga0208375_10…\n",
      "\u001b[90m 3\u001b[39m Ga020837… getorf CDS     917  \u001b[4m1\u001b[24m288 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_19_ga0208375_10…\n",
      "\u001b[90m 4\u001b[39m Ga020837… getorf CDS    \u001b[4m1\u001b[24m207  \u001b[4m1\u001b[24m374 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_21_ga0208375_10…\n",
      "\u001b[90m 5\u001b[39m Ga020837… getorf CDS    \u001b[4m1\u001b[24m356  \u001b[4m1\u001b[24m532 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_23_ga0208375_10…\n",
      "\u001b[90m 6\u001b[39m Ga020837… getorf CDS    \u001b[4m1\u001b[24m532  \u001b[4m1\u001b[24m690 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_25_ga0208375_10…\n",
      "\u001b[90m 7\u001b[39m Ga020837… getorf CDS    \u001b[4m1\u001b[24m648  \u001b[4m1\u001b[24m836 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_28_ga0208375_10…\n",
      "\u001b[90m 8\u001b[39m Ga020837… getorf CDS    \u001b[4m1\u001b[24m923  \u001b[4m2\u001b[24m033 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_32_ga0208375_10…\n",
      "\u001b[90m 9\u001b[39m Ga020837… getorf CDS    \u001b[4m2\u001b[24m033  \u001b[4m2\u001b[24m956 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_44_ga0208375_10…\n",
      "\u001b[90m10\u001b[39m Ga020837… getorf CDS    \u001b[4m2\u001b[24m956  \u001b[4m3\u001b[24m042 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_45_ga0208375_10…\n",
      "\u001b[90m# … with 174 more rows\u001b[39m\n",
      "\n",
      "[[39]]\n",
      "\u001b[90m# A tibble: 45 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga022265… getorf CDS      72   914 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_19_ga0222658_10…\n",
      "\u001b[90m 2\u001b[39m Ga022265… getorf CDS     914  \u001b[4m2\u001b[24m218 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_42_ga0222658_10…\n",
      "\u001b[90m 3\u001b[39m Ga022265… getorf CDS    \u001b[4m2\u001b[24m218  \u001b[4m2\u001b[24m427 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_45_ga0222658_10…\n",
      "\u001b[90m 4\u001b[39m Ga022265… getorf CDS    \u001b[4m2\u001b[24m482  \u001b[4m2\u001b[24m925 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_51_ga0222658_10…\n",
      "\u001b[90m 5\u001b[39m Ga022265… getorf CDS    \u001b[4m2\u001b[24m928  \u001b[4m3\u001b[24m443 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_61_ga0222658_10…\n",
      "\u001b[90m 6\u001b[39m Ga022265… getorf CDS    \u001b[4m3\u001b[24m455  \u001b[4m4\u001b[24m219 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_72_ga0222658_10…\n",
      "\u001b[90m 7\u001b[39m Ga022265… getorf CDS    \u001b[4m4\u001b[24m255  \u001b[4m8\u001b[24m028 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_127_ga0222658_1…\n",
      "\u001b[90m 8\u001b[39m Ga022265… getorf CDS    \u001b[4m7\u001b[24m860  \u001b[4m8\u001b[24m426 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_133_ga0222658_1…\n",
      "\u001b[90m 9\u001b[39m Ga022265… getorf CDS    \u001b[4m8\u001b[24m578  \u001b[4m8\u001b[24m937 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_144_ga0222658_1…\n",
      "\u001b[90m10\u001b[39m Ga022265… getorf CDS    \u001b[4m8\u001b[24m957  \u001b[4m9\u001b[24m466 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_156_ga0222658_1…\n",
      "\u001b[90m# … with 35 more rows\u001b[39m\n",
      "\n",
      "[[40]]\n",
      "\u001b[90m# A tibble: 65 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m Ga022266… getorf CDS     373   999 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_17_ga0222667_10…\n",
      "\u001b[90m 2\u001b[39m Ga022266… getorf CDS     999  \u001b[4m1\u001b[24m391 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_24_ga0222667_10…\n",
      "\u001b[90m 3\u001b[39m Ga022266… getorf CDS    \u001b[4m1\u001b[24m397  \u001b[4m1\u001b[24m909 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_30_ga0222667_10…\n",
      "\u001b[90m 4\u001b[39m Ga022266… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m2\u001b[24m942 \u001b[4m2\u001b[24m\u001b[4m4\u001b[24m588 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_552_ga0222667_1…\n",
      "\u001b[90m 5\u001b[39m Ga022266… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m4\u001b[24m587 \u001b[4m2\u001b[24m\u001b[4m5\u001b[24m684 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_569_ga0222667_1…\n",
      "\u001b[90m 6\u001b[39m Ga022266… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m5\u001b[24m684 \u001b[4m2\u001b[24m\u001b[4m5\u001b[24m944 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_575_ga0222667_1…\n",
      "\u001b[90m 7\u001b[39m Ga022266… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m5\u001b[24m938 \u001b[4m2\u001b[24m\u001b[4m6\u001b[24m666 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_588_ga0222667_1…\n",
      "\u001b[90m 8\u001b[39m Ga022266… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m6\u001b[24m663 \u001b[4m2\u001b[24m\u001b[4m6\u001b[24m977 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_591_ga0222667_1…\n",
      "\u001b[90m 9\u001b[39m Ga022266… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m6\u001b[24m977 \u001b[4m2\u001b[24m\u001b[4m7\u001b[24m117 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_595_ga0222667_1…\n",
      "\u001b[90m10\u001b[39m Ga022266… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m7\u001b[24m117 \u001b[4m2\u001b[24m\u001b[4m7\u001b[24m410 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_598_ga0222667_1…\n",
      "\u001b[90m# … with 55 more rows\u001b[39m\n",
      "\n",
      "[[41]]\n",
      "\u001b[90m# A tibble: 144 x 9\u001b[39m\n",
      "   seqid     source type  start   end score strand phase attributes             \n",
      "   \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m     \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<dbl>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m  \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m \u001b[3m\u001b[90m<chr>\u001b[39m\u001b[23m                  \n",
      "\u001b[90m 1\u001b[39m JGI12330… getorf CDS      27   173 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_3_jgi12330j1283…\n",
      "\u001b[90m 2\u001b[39m JGI12330… getorf CDS     243   491 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_9_jgi12330j1283…\n",
      "\u001b[90m 3\u001b[39m JGI12330… getorf CDS     484  \u001b[4m2\u001b[24m580 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_34_jgi12330j128…\n",
      "\u001b[90m 4\u001b[39m JGI12330… getorf CDS    \u001b[4m2\u001b[24m570  \u001b[4m3\u001b[24m172 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_41_jgi12330j128…\n",
      "\u001b[90m 5\u001b[39m JGI12330… getorf CDS    \u001b[4m7\u001b[24m162  \u001b[4m8\u001b[24m622 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_127_jgi12330j12…\n",
      "\u001b[90m 6\u001b[39m JGI12330… getorf CDS   \u001b[4m1\u001b[24m\u001b[4m7\u001b[24m993 \u001b[4m1\u001b[24m\u001b[4m8\u001b[24m364 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_314_jgi12330j12…\n",
      "\u001b[90m 7\u001b[39m JGI12330… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m2\u001b[24m553 \u001b[4m2\u001b[24m\u001b[4m3\u001b[24m896 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_385_jgi12330j12…\n",
      "\u001b[90m 8\u001b[39m JGI12330… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m6\u001b[24m789 \u001b[4m2\u001b[24m\u001b[4m7\u001b[24m262 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_429_jgi12330j12…\n",
      "\u001b[90m 9\u001b[39m JGI12330… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m8\u001b[24m142 \u001b[4m2\u001b[24m\u001b[4m8\u001b[24m294 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_443_jgi12330j12…\n",
      "\u001b[90m10\u001b[39m JGI12330… getorf CDS   \u001b[4m2\u001b[24m\u001b[4m8\u001b[24m328 \u001b[4m2\u001b[24m\u001b[4m8\u001b[24m933 1\u001b[90me\u001b[39m\u001b[31m-15\u001b[39m +      1     ID=ORF_454_jgi12330j12…\n",
      "\u001b[90m# … with 134 more rows\u001b[39m\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R -i ORFs_in_remote_homologs_groups,orf_annotation_table_to_import -o ME_ORFs_annot_table_simplecode\n",
    "\n",
    "# load libraries\n",
    "library(tidyverse)\n",
    "library(magrittr)\n",
    "library(glue)\n",
    "library(tidytidbits)\n",
    "\n",
    "# convert table to tibble and create dictionary\n",
    "orf_annotation_table_to_import %<>% as_tibble()\n",
    "getorf2new_id.dict = orf_annotation_table_to_import$`ORF ID`\n",
    "names(getorf2new_id.dict) = orf_annotation_table_to_import$`getorf featured ID` %>% str_split(' ') %>% purrr::map_chr(1)\n",
    "\n",
    "# loading TSV with annotations for ME ORFs. with some modifications this can be actually converted into a GFF\n",
    "ME_ORFs_annot_table = readr::read_tsv('../results/MEs_predicted_orfs_renamed/data/annotation_table_ORFs_renamed.tsv')\n",
    "\n",
    "# saving version with simplified code\n",
    "ME_ORFs_annot_table_simplecode = ME_ORFs_annot_table %>%\n",
    "    # selecting features employed in the GFF file\n",
    "    dplyr::select(`Genome ID`, `ORF ID`, `contig`, `strand`, `START`, `STOP`) %>%\n",
    "    dplyr::rename(seqid = 'contig',\n",
    "                  start = 'START',\n",
    "                  end = 'STOP') %>%\n",
    "    dplyr::mutate(seqid = seqid %>% str_split(' ') %>% purrr::map_chr(1), source = 'getorf', phase = '1', type = 'CDS', score = 0.000000000000001, attributes = as.character(glue('ID={`ORF ID`};Name={`ORF ID`}'))) %>%\n",
    "    .[, c('Genome ID', 'seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes')]\n",
    " \n",
    "ME_ORFs_annot_table_simplecode$`Genome ID` %<>% as.character()\n",
    "\n",
    "ORFs_in_remote_homologs_groups %<>% unlist() %>% tidytidbits::lookup_chr(., dict = getorf2new_id.dict, default = identity)\n",
    "ORFs_in_remote_homologs_groups\n",
    "\n",
    "# getting also the ORF ID of the feature IDs annotated by PATRIC\n",
    "PATRIC_featured_ids = ME_ORFs_annot_table %>% dplyr::filter(!is.na(`Featured ID in Genome Annotation`)) %>% .$`ORF ID` %>% unique()\n",
    "\n",
    "# saving for each ME in a different file\n",
    "ME_ORFs_annot_table_simplecode %>%\n",
    "    group_split(`Genome ID`) %>%\n",
    "    purrr::map(., ~{\n",
    "        # getting ME genome tag\n",
    "        tag = unique(.x$`Genome ID`)\n",
    "        \n",
    "        # saving TSV file with extension gff3\n",
    "        .x %>%\n",
    "            dplyr::select(-c(`Genome ID`)) %>%\n",
    "            dplyr::mutate(start2 = case_when(start > end ~ end, start < end ~ start), \n",
    "                          end2 = case_when(start > end ~ start, start < end ~ end)) %>%\n",
    "            dplyr::select(-c(start, end)) %>%\n",
    "            dplyr::rename(start = 'start2', end = 'end2') %>%\n",
    "            .[, c('seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes')] %>%\n",
    "            readr::write_tsv(., glue('../results/MEs_predicted_orfs_GFFs/simplified_nomenclature/ME.{tag}.ORFs.gff3'), col_names = F)\n",
    "        \n",
    "        # saving FASTA of remote homologs CDS\n",
    "        .x %>%\n",
    "            dplyr::mutate(orfid = attributes %>% str_split(';') %>% purrr::map_chr(1) %>% str_replace_all(., 'ID=', '')) %>%\n",
    "            dplyr::filter(orfid %in% ORFs_in_remote_homologs_groups) %>%\n",
    "            dplyr::select(-c(`Genome ID`, orfid)) %>%\n",
    "            dplyr::mutate(start2 = case_when(start > end ~ end, start < end ~ start), \n",
    "                          end2 = case_when(start > end ~ start, start < end ~ end)) %>%\n",
    "            dplyr::select(-c(start, end)) %>%\n",
    "            dplyr::rename(start = 'start2', end = 'end2') %>%\n",
    "            .[, c('seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes')] %>%\n",
    "            readr::write_tsv(., glue('../results/plotting_ME_synteny/data/GFFs/ME.{tag}.remote_homologues.ORFs.gff3'), col_names = F)\n",
    "        \n",
    "        # saving FASTA of features annotated by PATRIC + CDS coded in remote homologs (i.e. complete set of PATRIC + expanded annotation)\n",
    "        .x %>%\n",
    "            dplyr::mutate(orfid = attributes %>% str_split(';') %>% purrr::map_chr(1) %>% str_replace_all(., 'ID=', '')) %>%\n",
    "            dplyr::filter(orfid %in% c(ORFs_in_remote_homologs_groups, PATRIC_featured_ids)) %>%\n",
    "            dplyr::select(-c(`Genome ID`, orfid)) %>%\n",
    "            dplyr::mutate(start2 = case_when(start > end ~ end, start < end ~ start), \n",
    "                          end2 = case_when(start > end ~ start, start < end ~ end)) %>%\n",
    "            dplyr::select(-c(start, end)) %>%\n",
    "            dplyr::rename(start = 'start2', end = 'end2') %>%\n",
    "            .[, c('seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes')] %>%\n",
    "            readr::write_tsv(., glue('../results/plotting_ME_synteny/data/GFFs_all_records/ME.{tag}.all_records.ORFs.gff3'), col_names = F)\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "06fba9cf-43e3-48c4-bcf4-bdbd1a53d680",
   "metadata": {},
   "outputs": [],
   "source": [
    "# converting GFFs into BED format\n",
    "for ME_GFF_file in glob.glob('../results/plotting_ME_synteny/data/GFFs/*'):\n",
    "    # create BED file output name and store it in a variable\n",
    "    bed_output = ME_GFF_file.replace('GFFs', 'BED').replace('.gff3', '.bed')\n",
    "    # run external tool to create conversion\n",
    "    bed_conv_cmd = 'python -m jcvi.formats.gff bed --type=CDS --key=Name {0} -o {1}'.format(ME_GFF_file, bed_output).split(' ')\n",
    "    subprocess.run(bed_conv_cmd)\n",
    "    #print(ME_GFF_file)\n",
    "    #print(bed_output)\n",
    "\n",
    "# same for GFFs with all records\n",
    "for ME_GFF_file in glob.glob('../results/plotting_ME_synteny/data/GFFs_all_records/*'):\n",
    "    # create BED file output name and store it in a variable\n",
    "    bed_output = ME_GFF_file.replace('GFFs', 'BED').replace('.gff3', '.bed')\n",
    "    # run external tool to create conversion\n",
    "    bed_conv_cmd = 'python -m jcvi.formats.gff bed --type=CDS --key=Name {0} -o {1}'.format(ME_GFF_file, bed_output).split(' ')\n",
    "    subprocess.run(bed_conv_cmd)\n",
    "    #print(ME_GFF_file)\n",
    "    #print(bed_output)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c14b7c9-600b-4ef6-81f5-c10e7183d11a",
   "metadata": {},
   "source": [
    "## Generating miscellaneous files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3be2a71d-954c-41f4-aea9-ec78196a4ff6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools \n",
    "# create a small dictionary between species code and fasta id\n",
    "species2id_table = pd.read_csv('../data/genomes_and_taxonomy.csv', sep = ',')\n",
    "species2id_dict = {}\n",
    "\n",
    "for index, row in species2id_table.iterrows():\n",
    "    species2id_dict.update({row['Taxonomy']: row['Genome_id'].replace('GCA_', '')})\n",
    "\n",
    "# generating species pairs\n",
    "#[print(a) for a in itertools.combinations(species2id_table['Taxonomy'].to_list(), 2)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c072fc84-f80f-4841-93e7-60fbb5cc71f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first creating a dictionary to go from getorf IDs to novel ids\n",
    "orf_annotation_table_to_import.head()\n",
    "getorfid2new_id = {}\n",
    "\n",
    "for index, row in orf_annotation_table_to_import.iterrows():\n",
    "    getorfid = row['getorf featured ID'].split(' ')[0]\n",
    "    newid = row['ORF ID']\n",
    "    getorfid2new_id.update({getorfid: newid})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "94a99168-e203-4fde-9100-9955ee1c9c44",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading table with target organisms\n",
    "mobile_element_lengths_table_filtered = pd.read_csv('../results/target_organisms.tsv', sep ='\\t')\n",
    "mobile_element_lengths_table_filtered = mobile_element_lengths_table_filtered.drop(['Unnamed: 0', 'ME_file', 'Total length'], axis = 1)\n",
    "target_genomeids = mobile_element_lengths_table_filtered['Group tag'].to_list()\n",
    "\n",
    "# filtering species2id_table to organisms in this list\n",
    "species2id_table = species2id_table.query(\"`Genome_id` in @target_genomeids\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "262c70ab-1cab-4828-9b35-c01264b31ad6",
   "metadata": {},
   "source": [
    "### Modify BEDs to include dummy ORFs\n",
    "In order to be able to plot mobile elements in its full length, *dummy* ORFs are generated for each mobile element at its sequence start and end (of size 1bp) and included in the pipeline. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "97b977fe-7a43-480d-8ec5-042b74b311ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load function\n",
    "from Bio import SeqIO\n",
    "\n",
    "# loading all contigs from the mobile elements and parsing their length\n",
    "contig_lengths = {}\n",
    "for fna_file in glob.glob('../data/mobile_elements_sequences/*fna'):\n",
    "    records = [record for record in SeqIO.parse(fna_file, 'fasta')]\n",
    "    for record in records:\n",
    "        contig_lengths.update({record.id: len(record.seq)})\n",
    "        \n",
    "# get \n",
    "for bed_file in glob.glob('../results/plotting_ME_synteny/data/BED/*.remote_homologues.ORFs.bed'):\n",
    "    # working on non-empty files\n",
    "    if not os.stat(bed_file).st_size == 0:\n",
    "        #print(bed_file)\n",
    "        # load BED table\n",
    "        bed_table = pd.read_csv(bed_file, sep = '\\t', names = ['contig', 'start', 'end', 'orfid', 'evalue', 'strand'])\n",
    "        # get rid of dummy rows if already present\n",
    "        possible_dummies = [orfid for orfid in bed_table['orfid'].to_list() if 'dummy' in orfid]\n",
    "        bed_table = bed_table.query(\"`orfid` not in @possible_dummies\")\n",
    "        # get contig name\n",
    "        contig = bed_table.iloc[0]['contig']\n",
    "        # get contig length\n",
    "        contig_length = contig_lengths[contig]\n",
    "        # get dummy names by modifying name of first feature\n",
    "        orfid_list_start = bed_table.iloc[0]['orfid'].split('_')\n",
    "        orfid_list_end = bed_table.iloc[0]['orfid'].split('_')\n",
    "        orfid_list_start[1] = 'dummystart'\n",
    "        orfid_list_end[1] = 'dummyend'\n",
    "        orfid_start = '_'.join(orfid_list_start)\n",
    "        orfid_end = '_'.join(orfid_list_end)\n",
    "        # add dummy feature for start and end\n",
    "        dummy_start = pd.DataFrame.from_dict({'contig': [contig], \n",
    "                                              'start': [0], \n",
    "                                              'end': [1], \n",
    "                                              'orfid': [orfid_start],\n",
    "                                              'evalue': [1e-15],\n",
    "                                              'strand': ['+']})\n",
    "        dummy_end = pd.DataFrame.from_dict({'contig': [contig], \n",
    "                                            'start': [contig_length-1], \n",
    "                                            'end': [contig_length], \n",
    "                                            'orfid': [orfid_end],\n",
    "                                            'evalue': [1e-15],\n",
    "                                            'strand': ['+']})\n",
    "        # concatenate \n",
    "        bed_table_final = pd.concat([dummy_start, bed_table, dummy_end])\n",
    "        bed_table_final = bed_table_final.drop_duplicates()\n",
    "        # save table\n",
    "        bed_table_final.to_csv(bed_file, sep = '\\t', header = False, index = False)\n",
    "        \n",
    "# get \n",
    "for bed_file in glob.glob('../results/plotting_ME_synteny/data/BED_all_records/*.bed'):\n",
    "    # working on non-empty files\n",
    "    if not os.stat(bed_file).st_size == 0:\n",
    "        #print(bed_file)\n",
    "        # load BED table\n",
    "        bed_table = pd.read_csv(bed_file, sep = '\\t', names = ['contig', 'start', 'end', 'orfid', 'evalue', 'strand'])\n",
    "        # get rid of dummy rows if already present\n",
    "        possible_dummies = [orfid for orfid in bed_table['orfid'].to_list() if 'dummy' in orfid]\n",
    "        bed_table = bed_table.query(\"`orfid` not in @possible_dummies\")\n",
    "        # get contig name\n",
    "        contig = bed_table.iloc[0]['contig']\n",
    "        # get contig length\n",
    "        contig_length = contig_lengths[contig]\n",
    "        # get dummy names by modifying name of first feature\n",
    "        orfid_list_start = bed_table.iloc[0]['orfid'].split('_')\n",
    "        orfid_list_end = bed_table.iloc[0]['orfid'].split('_')\n",
    "        orfid_list_start[1] = 'dummystart'\n",
    "        orfid_list_end[1] = 'dummyend'\n",
    "        orfid_start = '_'.join(orfid_list_start)\n",
    "        orfid_end = '_'.join(orfid_list_end)\n",
    "        # add dummy feature for start and end\n",
    "        dummy_start = pd.DataFrame.from_dict({'contig': [contig], \n",
    "                                              'start': [0], \n",
    "                                              'end': [1], \n",
    "                                              'orfid': [orfid_start],\n",
    "                                              'evalue': [1e-15],\n",
    "                                              'strand': ['+']})\n",
    "        dummy_end = pd.DataFrame.from_dict({'contig': [contig], \n",
    "                                            'start': [contig_length-1], \n",
    "                                            'end': [contig_length], \n",
    "                                            'orfid': [orfid_end],\n",
    "                                            'evalue': [1e-15],\n",
    "                                            'strand': ['+']})\n",
    "        # concatenate \n",
    "        bed_table_final = pd.concat([dummy_start, bed_table, dummy_end])\n",
    "        bed_table_final = bed_table_final.drop_duplicates()\n",
    "        # save table\n",
    "        bed_table_final.to_csv(bed_file, sep = '\\t', header = False, index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17b29403-742f-4e6c-aa4e-f9cba6f2c685",
   "metadata": {},
   "source": [
    "### Concatenate to get all BEDs in one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1ba40ddf-360d-4bbd-bc81-fcccc2921878",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cat ../results/plotting_ME_synteny/data/BED_all_records/ME.017357405.1.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.60847.21.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.2743089.3.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.222984.5.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.2496101.3.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.1526048.3.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.ga0207733_100382.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.ga0207719_100190.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.926690.3.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.jgi12330j12834_1000008.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.ga0207718_100100.all_records.ORFs.bed > ../results/plotting_ME_synteny/data/BED/all_species_all_records.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2b62e8e8-ebc1-47b8-9445-f24b0a7aa5fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cat ../results/plotting_ME_synteny/data/BED/ME.60847.21.remote_homologues.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.2743089.3.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.222984.5.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.2496101.3.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.1526048.3.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.ga0207733_100382.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.ga0207719_100190.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.926690.3.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.jgi12330j12834_1000008.all_records.ORFs.bed \\\n",
    "../results/plotting_ME_synteny/data/BED_all_records/ME.ga0207718_100100.all_records.ORFs.bed > ../results/plotting_ME_synteny/data/BED_all_records/all_species_complete_records.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "000b56ca-db4e-4aac-a96f-bca7c4b5fb5b",
   "metadata": {},
   "source": [
    "## Creating block files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0cee50e-10ff-4068-835a-c223d2381242",
   "metadata": {},
   "source": [
    "### Create block files from homolog groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "853de0b3-55c4-4c38-9583-e249c6495ed9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a small dictionary between species code and fasta id\n",
    "species2id_table = pd.read_csv('../data/genomes_and_taxonomy.csv', sep = ',')\n",
    "species2id_dict = {}\n",
    "\n",
    "for index, row in species2id_table.iterrows():\n",
    "    species2id_dict.update({row['Taxonomy']: row['Genome_id'].replace('GCA_', '')})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ee0cd7d6-0245-4989-9dcb-386af4dc8914",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import\n",
    "import glob\n",
    "from Bio import SeqIO\n",
    "# loading table with new annotation\n",
    "import pandas as pd\n",
    "\n",
    "# create dictionary between species and column name in block file\n",
    "species2blockcol_dict = {'Halogeometricum_borinquense_strain_wsp4': 'Halogeometricum',\n",
    "                         'Halobonum_sp_NJ-3-1': 'Halobonum',\n",
    "                         'Natrinema_altunense_strain_AJ2': 'Natrinema',\n",
    "                         'Haloterrigena_sp_SYSU_A121-1': 'Haloterrigena',\n",
    "                         'Haloferax_sp_Q22': 'Haloferax',\n",
    "                         'ga0207733_100382': 'ga0207733_100382',\n",
    "                         'Halovivax_sp_KZCA124': 'Halovivax',\n",
    "                         'ga0207719_100190': 'ga0207719_100190',\n",
    "                         'Haloplanus_natans_DSM_17983': 'Haloplanus',\n",
    "                         'jgi12330j12834_1000008': 'jgi12330j12834_1000008',\n",
    "                         'ga0207718_100100': 'ga0207718_100100'}\n",
    "\n",
    "# create order of species in my configuration\n",
    "block_column_order = ['Halovivax', 'Halogeometricum', 'Halobonum', 'ga0207718_100100', 'jgi12330j12834_1000008', \n",
    "                                                             'Natrinema', 'Halovivax', 'Haloterrigena', 'Haloplanus', \n",
    "                                                             'Haloferax', 'ga0207719_100190', 'ga0207733_100382']\n",
    "\n",
    "orf_annotation_table = pd.read_csv('../results/MEs_predicted_orfs_renamed/data/annotation_table_ORFs_renamed.tsv', sep = '\\t')\n",
    "orf_annotation_table.head()\n",
    "# perform little change\n",
    "orf_annotation_table_to_import = orf_annotation_table.drop(['Featured ID in Genome Annotation'], axis = 1)\n",
    "\n",
    "# first creating a dictionary to go from getorf IDs to novel ids\n",
    "orf_annotation_table_to_import.head()\n",
    "getorfid2new_id = {}\n",
    "\n",
    "for index, row in orf_annotation_table_to_import.iterrows():\n",
    "    getorfid = row['getorf featured ID'].split(' ')[0]\n",
    "    newid = row['ORF ID']\n",
    "    getorfid2new_id.update({getorfid: newid})\n",
    "\n",
    "# create list to allocate rows\n",
    "homolog_groups_composition_rows = []\n",
    "\n",
    "# looping over groups\n",
    "for homolog_group_file in glob.glob('../results/trying_to_catch_homologs/results/connected_groups_filtered_splitted_arCOGs/fastas/protein/*faa'):\n",
    "    # parsing records in them and translating the records to new ids\n",
    "    records = [getorfid2new_id[record.id] for record in SeqIO.parse(homolog_group_file, 'fasta')]\n",
    "    # create a dictionary for relevant species, including an empty list for each\n",
    "    classifing_genes_by_species_dict = {'Halogeometricum_borinquense_strain_wsp4': [],\n",
    "                                   'Halobonum_sp_NJ-3-1': [],\n",
    "                                   'Natrinema_altunense_strain_AJ2': [],\n",
    "                                   'Haloterrigena_sp_SYSU_A121-1': [],\n",
    "                                   'Haloferax_sp_Q22':[] ,\n",
    "                                   'ga0207733_100382': [],\n",
    "                                   'Halovivax_sp_KZCA124': [],\n",
    "                                   'ga0207719_100190': [],\n",
    "                                   'Haloplanus_natans_DSM_17983': [],\n",
    "                                   'jgi12330j12834_1000008': [],\n",
    "                                   'ga0207718_100100': []}\n",
    "    for record_id in records:\n",
    "    # for each record... \n",
    "        # identify its species\n",
    "        species = '_'.join(record_id.split('_')[2:])\n",
    "        #print(species)\n",
    "        if species in list(species2blockcol_dict.keys()):\n",
    "            species_blockname = species2blockcol_dict[species]\n",
    "            # append to the relevant dictionary\n",
    "            classifing_genes_by_species_dict[species].append(record_id)\n",
    "    # create a pandas dataframe from a dictionary that includes entries from the same species collapsed with a ','\n",
    "    for key in classifing_genes_by_species_dict.keys():\n",
    "        if len(classifing_genes_by_species_dict[key]) > 0:\n",
    "            classifing_genes_by_species_dict[key] = [','.join(classifing_genes_by_species_dict[key])]\n",
    "        elif len(classifing_genes_by_species_dict[key]) == 0:\n",
    "            classifing_genes_by_species_dict[key] = ['.']\n",
    "    # add homolog group\n",
    "    classifing_genes_by_species_dict.update({'homolog_group': [homolog_group_file]})\n",
    "    gene_row = pd.DataFrame.from_dict(classifing_genes_by_species_dict)\n",
    "    # append to general list\n",
    "    homolog_groups_composition_rows.append(gene_row)\n",
    "\n",
    "# concatenate\n",
    "homolog_groups_composition_table = pd.concat(homolog_groups_composition_rows)\n",
    "\n",
    "# save table\n",
    "homolog_groups_composition_table.to_csv('../results/plotting_ME_synteny/data/homolog_groups_composition_table.tsv',\n",
    "                                        sep = '\\t',\n",
    "                                        index = False)\n",
    "\n",
    "# splitting rows in R..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cd6c3fe0-8fd3-4723-85b3-25fc7f48356b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "# listing ORF prediction files and creating a list of all CDS that are in the remote homolog groups\n",
    "ORFs_in_remote_homologs_groups = []\n",
    "for remote_homologs_group_file in glob.glob('../results/trying_to_catch_homologs/results/connected_groups_filtered_splitted_arCOGs/fastas/protein/*faa'):\n",
    "    for record in SeqIO.parse(remote_homologs_group_file, 'fasta'):\n",
    "        ORFs_in_remote_homologs_groups.append(record.id)\n",
    "\n",
    "# first creating a dictionary to go from getorf IDs to novel ids\n",
    "orf_annotation_table_to_import.head()\n",
    "getorfid2new_id = {}\n",
    "\n",
    "for index, row in orf_annotation_table_to_import.iterrows():\n",
    "    getorfid = row['getorf featured ID'].split(' ')[0]\n",
    "    newid = row['ORF ID']\n",
    "    getorfid2new_id.update({getorfid: newid})\n",
    "    \n",
    "# load concatenated BED with all features\n",
    "patric_and_orfs_bed_table = pd.read_csv('../results/plotting_ME_synteny/data/BED_all_records/all_species_complete_records.bed', sep = '\\t', names = ['contig', 'start', 'end', 'orfid', 'evalue', 'strand'])\n",
    "patric_and_orfs_orfids = patric_and_orfs_bed_table['orfid'].to_list()\n",
    "\n",
    "# remember that got list of features in homolog groups (ORFs_in_remote_homologs_groups)\n",
    "# translating to new id\n",
    "orfs_in_homolog_groups = [getorfid2new_id[id] for id in ORFs_in_remote_homologs_groups if 'dummy' not in id]\n",
    "\n",
    "# determine features that arent present in the list\n",
    "patric_and_orfs_orfids_nonhomologs = [id for id in patric_and_orfs_orfids if id not in orfs_in_homolog_groups]\n",
    "# create list to allocate rows of block\n",
    "dummy_block_entries_rows = []\n",
    "\n",
    "# create dictionary between species and column name in block file\n",
    "species2blockcol_dict = {'Halogeometricum_borinquense_strain_wsp4': 'Halogeometricum',\n",
    "                         'Halobonum_sp_NJ-3-1': 'Halobonum',\n",
    "                         'Natrinema_altunense_strain_AJ2': 'Natrinema',\n",
    "                         'Haloterrigena_sp_SYSU_A121-1': 'Haloterrigena',\n",
    "                         'Haloferax_sp_Q22': 'Haloferax',\n",
    "                         'ga0207733_100382': 'ga0207733_100382',\n",
    "                         'ga0207719_100190': 'ga0207719_100190',\n",
    "                         'Haloplanus_natans_DSM_17983': 'Haloplanus',\n",
    "                         'jgi12330j12834_1000008': 'jgi12330j12834_1000008',\n",
    "                         'ga0207718_100100': 'ga0207718_100100'}\n",
    "\n",
    "# create order of species in my configuration\n",
    "block_column_order = ['Halogeometricum', 'Halobonum', 'ga0207718_100100', 'jgi12330j12834_1000008', \n",
    "                                                             'Natrinema', 'Halovivax', 'Haloterrigena', 'Haloplanus', \n",
    "                                                             'Haloferax', 'ga0207719_100190', 'ga0207733_100382']\n",
    "# loop over\n",
    "for orfid in patric_and_orfs_orfids_nonhomologs:\n",
    "    # detect species\n",
    "    species = '_'.join(orfid.split('_')[2:])\n",
    "    species_blockname = species2blockcol_dict[species]\n",
    "    # get the other species\n",
    "    other_species = [species for species in block_column_order if species != species_blockname]\n",
    "    # create dictionary to allocate, as columns, the orfid (for the species) or a '.' (other species)\n",
    "    dummy_dict = {}\n",
    "    dummy_dict.update({species_blockname: [orfid]})\n",
    "    # create pandas data frame\n",
    "    for species in other_species:\n",
    "        dummy_dict.update({species: ['.']})\n",
    "    dummy_row = pd.DataFrame.from_dict(dummy_dict)\n",
    "    dummy_row = dummy_row[block_column_order]\n",
    "    # order and append to list\n",
    "    dummy_block_entries_rows.append(dummy_row)\n",
    "    \n",
    "# concatenate\n",
    "dummy_block_entries_table = pd.concat(dummy_block_entries_rows)\n",
    "# drop duplicates (although shouldn't exist)\n",
    "dummy_block_entries_table = dummy_block_entries_table.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5b483ab8-004e-47f8-add4-bc1d14dde487",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i homolog_groups_composition_table -o homolog_groups_composition_table_splitted\n",
    "\n",
    "# load libraries\n",
    "library(tidyverse)\n",
    "library(magrittr)\n",
    "library(glue)\n",
    "\n",
    "# convert to tibble()\n",
    "homolog_groups_composition_table %<>% as_tibble()\n",
    "homolog_groups_composition_table_splitted = homolog_groups_composition_table %>%\n",
    "    tidyr::separate_rows(., `Halogeometricum_borinquense_strain_wsp4`, sep = ',') %>%\n",
    "    tidyr::separate_rows(., `Halobonum_sp_NJ-3-1`, sep = ',') %>%\n",
    "    tidyr::separate_rows(., `Natrinema_altunense_strain_AJ2`, sep = ',') %>%\n",
    "    tidyr::separate_rows(., `Haloterrigena_sp_SYSU_A121-1`, sep = ',') %>%\n",
    "    tidyr::separate_rows(., `Haloferax_sp_Q22`, sep = ',') %>%\n",
    "    tidyr::separate_rows(., `ga0207733_100382`, sep = ',') %>%\n",
    "    tidyr::separate_rows(., `ga0207719_100190`, sep = ',') %>%\n",
    "    tidyr::separate_rows(., `Haloplanus_natans_DSM_17983`, sep = ',') %>%\n",
    "    tidyr::separate_rows(., `jgi12330j12834_1000008`, sep = ',') %>%\n",
    "    tidyr::separate_rows(., `ga0207718_100100`, sep = ',') %>%\n",
    "    tidyr::separate_rows(., `Halovivax_sp_KZCA124`, sep = ',') %>%\n",
    "    tidyr::replace_na(., list('Halogeometricum_borinquense_strain_wsp4' = '.',\n",
    "                              'Halobonum_sp_NJ-3-1' = '.',\n",
    "                              'Natrinema_altunense_strain_AJ2' = '.',\n",
    "                              'Haloterrigena_sp_SYSU_A121-1' = '.',\n",
    "                              'Haloferax_sp_Q22' = '.',\n",
    "                              'ga0207733_100382' = '.',\n",
    "                              'ga0207719_100190' = '.',\n",
    "                              'Haloplanus_natans_DSM_17983' = '.',\n",
    "                              'jgi12330j12834_1000008' = '.',\n",
    "                              'ga0207718_100100' = '.',\n",
    "                              'Halovivax_sp_KZCA124' = '.'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a9a53e7e-af72-4ce7-a164-a53b4c6337db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# order\n",
    "homolog_groups_composition_table_splitted = homolog_groups_composition_table_splitted[['homolog_group', 'Halogeometricum_borinquense_strain_wsp4', \n",
    "                                           'Halobonum_sp_NJ-3-1', 'ga0207718_100100', 'jgi12330j12834_1000008', \n",
    "                                           'Natrinema_altunense_strain_AJ2', 'Halovivax_sp_KZCA124', \n",
    "                                           'Haloterrigena_sp_SYSU_A121-1', 'Haloplanus_natans_DSM_17983', \n",
    "                                           'Haloferax_sp_Q22', 'ga0207719_100190', 'ga0207733_100382']]\n",
    "# save table\n",
    "homolog_groups_composition_table_splitted.to_csv('../results/plotting_ME_synteny/data/homolog_groups_composition_table.tsv', sep = '\\t', header = True, index = False)\n",
    "\n",
    "# get rid of homologous group table and save as block file\n",
    "homolog_groups_composition_table_splitted.drop('homolog_group', axis = 1).to_csv('../results/plotting_ME_synteny/data/block_files/all_species_noanchor.block', sep = '\\t', header = False, index = False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e20ac21b-e114-4813-8485-107716126263",
   "metadata": {},
   "source": [
    "### Create block files for features that dont present homology signal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "028230b3-e6eb-4d33-a565-becc6451251f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# listing ORF prediction files and creating a list of all CDS that are in the remote homolog groups\n",
    "ORFs_in_remote_homologs_groups = []\n",
    "for remote_homologs_group_file in glob.glob('../results/trying_to_catch_homologs/results/connected_groups_filtered_splitted_arCOGs/fastas/protein/*faa'):\n",
    "    for record in SeqIO.parse(remote_homologs_group_file, 'fasta'):\n",
    "        ORFs_in_remote_homologs_groups.append(record.id)\n",
    "\n",
    "# load concatenated BED with all features\n",
    "patric_and_orfs_bed_table = pd.read_csv('../results/plotting_ME_synteny/data/BED_all_records/all_species_complete_records.bed', sep = '\\t', names = ['contig', 'start', 'end', 'orfid', 'evalue', 'strand'])\n",
    "patric_and_orfs_orfids = patric_and_orfs_bed_table['orfid'].to_list()\n",
    "\n",
    "# remember that got list of features in homolog groups (ORFs_in_remote_homologs_groups)\n",
    "# translating to new id\n",
    "orfs_in_homolog_groups = [getorfid2new_id[id] for id in ORFs_in_remote_homologs_groups if 'dummy' not in id]\n",
    "\n",
    "# determine features that arent present in the list\n",
    "patric_and_orfs_orfids_nonhomologs = [id for id in patric_and_orfs_orfids if id not in orfs_in_homolog_groups]\n",
    "# create list to allocate rows of block\n",
    "dummy_block_entries_rows = []\n",
    "\n",
    "# create dictionary between species and column name in block file\n",
    "species2blockcol_dict = {'Halogeometricum_borinquense_strain_wsp4': 'Halogeometricum',\n",
    "                         'Halobonum_sp_NJ-3-1': 'Halobonum',\n",
    "                         'Natrinema_altunense_strain_AJ2': 'Natrinema',\n",
    "                         'Haloterrigena_sp_SYSU_A121-1': 'Haloterrigena',\n",
    "                         'Haloferax_sp_Q22': 'Haloferax',\n",
    "                         'ga0207733_100382': 'ga0207733_100382',\n",
    "                         'ga0207719_100190': 'ga0207719_100190',\n",
    "                         'Haloplanus_natans_DSM_17983': 'Haloplanus',\n",
    "                         'jgi12330j12834_1000008': 'jgi12330j12834_1000008',\n",
    "                         'ga0207718_100100': 'ga0207718_100100'}\n",
    "\n",
    "# create order of species in my configuration\n",
    "block_column_order = ['Halogeometricum', 'Halobonum', 'ga0207718_100100', 'jgi12330j12834_1000008', \n",
    "                                                             'Natrinema', 'Halovivax', 'Haloterrigena', 'Haloplanus', \n",
    "                                                             'Haloferax', 'ga0207719_100190', 'ga0207733_100382']\n",
    "# loop over\n",
    "for orfid in patric_and_orfs_orfids_nonhomologs:\n",
    "    # detect species\n",
    "    species = '_'.join(orfid.split('_')[2:])\n",
    "    species_blockname = species2blockcol_dict[species]\n",
    "    # get the other species\n",
    "    other_species = [species for species in block_column_order if species != species_blockname]\n",
    "    # create dictionary to allocate, as columns, the orfid (for the species) or a '.' (other species)\n",
    "    dummy_dict = {}\n",
    "    dummy_dict.update({species_blockname: [orfid]})\n",
    "    # create pandas data frame\n",
    "    for species in other_species:\n",
    "        dummy_dict.update({species: ['.']})\n",
    "    dummy_row = pd.DataFrame.from_dict(dummy_dict)\n",
    "    dummy_row = dummy_row[block_column_order]\n",
    "    # order and append to list\n",
    "    dummy_block_entries_rows.append(dummy_row)\n",
    "    \n",
    "# concatenate\n",
    "dummy_block_entries_table = pd.concat(dummy_block_entries_rows)\n",
    "# drop duplicates (although shouldn't exist)\n",
    "dummy_block_entries_table = dummy_block_entries_table.drop_duplicates()\n",
    "\n",
    "# save into a file\n",
    "dummy_block_entries_table.to_csv('../results/plotting_ME_synteny/data/block_files/dummy_features_no_homologues.block', sep = '\\t', index = False, header = False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d356ffb-c09a-4e2c-8edd-eff6045673d7",
   "metadata": {},
   "source": [
    "### Concatenating to create final block file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ac80030e-b0d4-4110-b2fa-b669c96595d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cat ../results/plotting_ME_synteny/data/block_files/all_species_noanchor.block ../results/plotting_ME_synteny/data/block_files/dummy_features_no_homologues.block > ../results/plotting_ME_synteny/data/block_files/all_species_with_anchors_full_features.block"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "601db6b1-1429-47b2-ba58-8609ad7b6592",
   "metadata": {},
   "source": [
    "### Subsetting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "a54c6446-1978-49c8-92a6-a51ff79a7f22",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get table\n",
    "homolog_groups_composition_table = pd.read_csv('../results/plotting_ME_synteny/data/homolog_groups_composition_table.tsv',\n",
    "                                        sep = '\\t')\n",
    "\n",
    "# subset for target species\n",
    "# convert selected species to blocknames column names with <species2blockcol_dict>\n",
    "selected_block_columns = [species2blockcol_dict[species] for species in selection_of_species]\n",
    "\n",
    "homolog_groups_composition_table_subset = homolog_groups_composition_table[selection_of_species]\n",
    "homolog_groups_composition_table_subset = homolog_groups_composition_table_subset.drop_duplicates()\n",
    "\n",
    "# splitting in R\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "228d6ea3-86a1-42e7-8bfd-f167a82f8abf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The rpy2.ipython extension is already loaded. To reload it, use:\n",
      "  %reload_ext rpy2.ipython\n"
     ]
    }
   ],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "47250806-d449-42df-b045-12c7da6ed56a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i homolog_groups_composition_table_subset -o homolog_groups_composition_table_subset_splitted\n",
    "\n",
    "# load libraries\n",
    "library(tidyverse)\n",
    "library(magrittr)\n",
    "library(glue)\n",
    "\n",
    "# convert to tibble()\n",
    "homolog_groups_composition_table_subset %<>% as_tibble()\n",
    "homolog_groups_composition_table_subset_splitted = homolog_groups_composition_table_subset %>%\n",
    "    tidyr::separate_rows(., `Halogeometricum_borinquense_strain_wsp4`, sep = ',') %>%\n",
    "    tidyr::separate_rows(., `Natrinema_altunense_strain_AJ2`, sep = ',') %>%\n",
    "    tidyr::separate_rows(., `Haloferax_sp_Q22`, sep = ',') %>%\n",
    "    tidyr::separate_rows(., `Haloplanus_natans_DSM_17983`, sep = ',') %>%\n",
    "    tidyr::separate_rows(., `jgi12330j12834_1000008`, sep = ',') %>%\n",
    "    tidyr::replace_na(., list('Halogeometricum_borinquense_strain_wsp4' = '.',\n",
    "                              'Natrinema_altunense_strain_AJ2' = '.',\n",
    "                              'Haloferax_sp_Q22' = '.',\n",
    "                              'Haloplanus_natans_DSM_17983' = '.',\n",
    "                              'jgi12330j12834_1000008' = '.'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "49f63a59-20b8-44f6-93ef-6b7058b15b76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Halogeometricum_borinquense_strain_wsp4</th>\n",
       "      <th>jgi12330j12834_1000008</th>\n",
       "      <th>Natrinema_altunense_strain_AJ2</th>\n",
       "      <th>Haloplanus_natans_DSM_17983</th>\n",
       "      <th>Haloferax_sp_Q22</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>.</td>\n",
       "      <td>ORF_2889_jgi12330j12834_1000008</td>\n",
       "      <td>.</td>\n",
       "      <td>ORF_2192_Haloplanus_natans_DSM_17983</td>\n",
       "      <td>ORF_1207_Haloferax_sp_Q22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ORF_1098_Halogeometricum_borinquense_strain_wsp4</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>.</td>\n",
       "      <td>ORF_2530_jgi12330j12834_1000008</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>.</td>\n",
       "      <td>ORF_1532_jgi12330j12834_1000008</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>ORF_124_Halogeometricum_borinquense_strain_wsp4</td>\n",
       "      <td>ORF_1496_jgi12330j12834_1000008</td>\n",
       "      <td>ORF_1255_Natrinema_altunense_strain_AJ2</td>\n",
       "      <td>ORF_1736_Haloplanus_natans_DSM_17983</td>\n",
       "      <td>ORF_1454_Haloferax_sp_Q22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>ORF_124_Halogeometricum_borinquense_strain_wsp4</td>\n",
       "      <td>ORF_1496_jgi12330j12834_1000008</td>\n",
       "      <td>ORF_1255_Natrinema_altunense_strain_AJ2</td>\n",
       "      <td>ORF_2439_Haloplanus_natans_DSM_17983</td>\n",
       "      <td>ORF_1454_Haloferax_sp_Q22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>ORF_1919_Halogeometricum_borinquense_strain_wsp4</td>\n",
       "      <td>ORF_1496_jgi12330j12834_1000008</td>\n",
       "      <td>ORF_1255_Natrinema_altunense_strain_AJ2</td>\n",
       "      <td>ORF_1736_Haloplanus_natans_DSM_17983</td>\n",
       "      <td>ORF_1454_Haloferax_sp_Q22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>ORF_1919_Halogeometricum_borinquense_strain_wsp4</td>\n",
       "      <td>ORF_1496_jgi12330j12834_1000008</td>\n",
       "      <td>ORF_1255_Natrinema_altunense_strain_AJ2</td>\n",
       "      <td>ORF_2439_Haloplanus_natans_DSM_17983</td>\n",
       "      <td>ORF_1454_Haloferax_sp_Q22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ORF_521_Natrinema_altunense_strain_AJ2</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>143 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Halogeometricum_borinquense_strain_wsp4  \\\n",
       "1                                                   .   \n",
       "2    ORF_1098_Halogeometricum_borinquense_strain_wsp4   \n",
       "3                                                   .   \n",
       "4                                                   .   \n",
       "5                                                   .   \n",
       "..                                                ...   \n",
       "139   ORF_124_Halogeometricum_borinquense_strain_wsp4   \n",
       "140   ORF_124_Halogeometricum_borinquense_strain_wsp4   \n",
       "141  ORF_1919_Halogeometricum_borinquense_strain_wsp4   \n",
       "142  ORF_1919_Halogeometricum_borinquense_strain_wsp4   \n",
       "143                                                 .   \n",
       "\n",
       "              jgi12330j12834_1000008           Natrinema_altunense_strain_AJ2  \\\n",
       "1    ORF_2889_jgi12330j12834_1000008                                        .   \n",
       "2                                  .                                        .   \n",
       "3                                  .                                        .   \n",
       "4    ORF_2530_jgi12330j12834_1000008                                        .   \n",
       "5    ORF_1532_jgi12330j12834_1000008                                        .   \n",
       "..                               ...                                      ...   \n",
       "139  ORF_1496_jgi12330j12834_1000008  ORF_1255_Natrinema_altunense_strain_AJ2   \n",
       "140  ORF_1496_jgi12330j12834_1000008  ORF_1255_Natrinema_altunense_strain_AJ2   \n",
       "141  ORF_1496_jgi12330j12834_1000008  ORF_1255_Natrinema_altunense_strain_AJ2   \n",
       "142  ORF_1496_jgi12330j12834_1000008  ORF_1255_Natrinema_altunense_strain_AJ2   \n",
       "143                                .   ORF_521_Natrinema_altunense_strain_AJ2   \n",
       "\n",
       "              Haloplanus_natans_DSM_17983           Haloferax_sp_Q22  \n",
       "1    ORF_2192_Haloplanus_natans_DSM_17983  ORF_1207_Haloferax_sp_Q22  \n",
       "2                                       .                          .  \n",
       "3                                       .                          .  \n",
       "4                                       .                          .  \n",
       "5                                       .                          .  \n",
       "..                                    ...                        ...  \n",
       "139  ORF_1736_Haloplanus_natans_DSM_17983  ORF_1454_Haloferax_sp_Q22  \n",
       "140  ORF_2439_Haloplanus_natans_DSM_17983  ORF_1454_Haloferax_sp_Q22  \n",
       "141  ORF_1736_Haloplanus_natans_DSM_17983  ORF_1454_Haloferax_sp_Q22  \n",
       "142  ORF_2439_Haloplanus_natans_DSM_17983  ORF_1454_Haloferax_sp_Q22  \n",
       "143                                     .                          .  \n",
       "\n",
       "[143 rows x 5 columns]"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "homolog_groups_composition_table_subset_splitted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ccc376a6-1c9f-4818-84e6-e6da02847c81",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'homolog_groups_composition_table_subset_splitted' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-26-6de6fb4394e9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mhomolog_groups_composition_table_subset_splitted\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../results/plotting_ME_synteny/data/block_files/subset_species_no_anchors_only_homologues.block'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'\\t'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'homolog_groups_composition_table_subset_splitted' is not defined"
     ]
    }
   ],
   "source": [
    "homolog_groups_composition_table_subset_splitted.to_csv('../results/plotting_ME_synteny/data/block_files/subset_species_no_anchors_only_homologues.block', sep = '\\t', index = False, header = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "34253751-8749-4fd8-8c96-f7c938841c37",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create rows for each species\n",
    "dummystart_block_halovivax = pd.DataFrame.from_dict({'Halovivax': ['ORF_dummystart_Halovivax_sp_KZCA124'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummyend_block_halovivax = pd.DataFrame.from_dict({'Halovivax': ['ORF_dummyend_Halovivax_sp_KZCA124'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummystart_block_halogeometricum = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['ORF_dummystart_Halogeometricum_borinquense_strain_wsp4'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummyend_block_halogeometricum = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['ORF_dummyend_Halogeometricum_borinquense_strain_wsp4'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummystart_block_halobonum = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['ORF_dummystart_Halobonum_sp_NJ-3-1'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummyend_block_halobonum = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['ORF_dummyend_Halobonum_sp_NJ-3-1'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummystart_block_natrinema = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['ORF_dummystart_Natrinema_altunense_strain_AJ2'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummyend_block_natrinema = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['ORF_dummyend_Natrinema_altunense_strain_AJ2'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummystart_block_haloterrigena = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['ORF_dummystart_Haloterrigena_sp_SYSU_A121-1'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummyend_block_haloterrigena = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['ORF_dummyend_Haloterrigena_sp_SYSU_A121-1'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummystart_block_haloferax = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['ORF_dummystart_Haloferax_sp_Q22'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummyend_block_haloferax = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['ORF_dummyend_Haloferax_sp_Q22'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummystart_block_ga0207733_100382 = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['ORF_dummystart_ga0207733_100382'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummyend_block_ga0207733_100382 = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['ORF_dummyend_ga0207733_100382'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummystart_block_ga0207719_100190 = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['ORF_dummystart_ga0207719_100190'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummyend_block_ga0207719_100190 = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['ORF_dummyend_ga0207719_100190'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummystart_block_haloplanus = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['ORF_dummystart_Haloplanus_natans_DSM_17983'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummyend_block_haloplanus = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['ORF_dummyend_Haloplanus_natans_DSM_17983'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummystart_block_jgi12330j12834_1000008 = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['ORF_dummystart_jgi12330j12834_1000008'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummyend_block_jgi12330j12834_1000008 = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['ORF_dummyend_jgi12330j12834_1000008'], 'ga0207718_100100': ['.']})\n",
    "\n",
    "dummystart_block_ga0207718_100100 = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['ORF_dummystart_ga0207718_100100']})\n",
    "\n",
    "dummyend_block_ga0207718_100100 = pd.DataFrame.from_dict({'Halovivax': ['.'], 'Halogeometricum': ['.'],'Halobonum': ['.'], 'Natrinema': ['.'], 'Haloterrigena': ['.'],\n",
    "                                                      'Haloferax': ['.'], 'ga0207733_100382': ['.'], 'ga0207719_100190': ['.'], 'Haloplanus': ['.'],\n",
    "                                                      'jgi12330j12834_1000008': ['.'], 'ga0207718_100100': ['ORF_dummyend_ga0207718_100100']})\n",
    "\n",
    "# concatenate\n",
    "block_column_order = ['Halogeometricum', 'Halobonum', 'ga0207718_100100', 'jgi12330j12834_1000008', \n",
    "                                                             'Natrinema', 'Halovivax', 'Haloterrigena', 'Haloplanus', \n",
    "                                                             'Haloferax', 'ga0207719_100190', 'ga0207733_100382']\n",
    "\n",
    "dummy_block_table = pd.concat([dummystart_block_halovivax[block_column_order],\n",
    "           dummyend_block_halovivax[block_column_order],\n",
    "           dummystart_block_halogeometricum[block_column_order], dummyend_block_halogeometricum[block_column_order],\n",
    "           dummystart_block_halobonum[block_column_order], dummyend_block_halobonum[block_column_order],\n",
    "           dummystart_block_natrinema[block_column_order], dummyend_block_natrinema[block_column_order],\n",
    "           dummystart_block_haloterrigena[block_column_order], dummyend_block_haloterrigena[block_column_order],\n",
    "           dummystart_block_haloferax[block_column_order], dummyend_block_haloferax[block_column_order],\n",
    "           dummystart_block_ga0207733_100382[block_column_order], dummyend_block_ga0207733_100382[block_column_order],\n",
    "           dummystart_block_ga0207719_100190[block_column_order], dummyend_block_ga0207719_100190[block_column_order],\n",
    "           dummystart_block_haloplanus[block_column_order], dummyend_block_haloplanus[block_column_order],\n",
    "           dummystart_block_jgi12330j12834_1000008[block_column_order], dummyend_block_jgi12330j12834_1000008[block_column_order],\n",
    "           dummystart_block_ga0207718_100100[block_column_order], dummyend_block_ga0207718_100100[block_column_order]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "76a3f879-9e8a-4a29-8bf2-f207973d321c",
   "metadata": {},
   "outputs": [],
   "source": [
    "selection_of_species = ['Halogeometricum_borinquense_strain_wsp4', \n",
    "                  'jgi12330j12834_1000008', \n",
    "                  'Natrinema_altunense_strain_AJ2', \n",
    "                  'Haloplanus_natans_DSM_17983', \n",
    "                  'Haloferax_sp_Q22']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "df53b7e7-e874-4d51-b89d-4fd31c634a1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load table\n",
    "dummy_block_entries_table = pd.read_csv('../results/plotting_ME_synteny/data/block_files/dummy_features_no_homologues.block', \n",
    "                                        sep = '\\t', names = ['Halogeometricum', 'Halobonum', 'ga0207718_100100', 'jgi12330j12834_1000008', \n",
    "                                                             'Natrinema', 'Halovivax', 'Haloterrigena', 'Haloplanus', \n",
    "                                                             'Haloferax', 'ga0207719_100190', 'ga0207733_100382'])\n",
    "\n",
    "# subset\n",
    "# convert selected species to blocknames column names with <species2blockcol_dict>\n",
    "selection_of_species = ['Halogeometricum_borinquense_strain_wsp4', \n",
    "                  'jgi12330j12834_1000008', \n",
    "                  'Natrinema_altunense_strain_AJ2', \n",
    "                  'Haloplanus_natans_DSM_17983', \n",
    "                  'Haloferax_sp_Q22']\n",
    "selected_block_columns = [species2blockcol_dict[species] for species in selection_of_species]\n",
    "\n",
    "dummy_block_entries_table_subset = dummy_block_entries_table[selected_block_columns]\n",
    "\n",
    "# now subsetting for dummy blocks and joining\n",
    "dummy_block_table_subset = dummy_block_table[selected_block_columns]\n",
    "dummy_block_entries_table_subset = pd.concat([dummy_block_entries_table_subset, dummy_block_table_subset])\n",
    "\n",
    "# now dropping duplicates\n",
    "dummy_block_entries_table_subset = dummy_block_entries_table_subset.drop_duplicates()\n",
    "\n",
    "# save to file\n",
    "dummy_block_entries_table_subset.to_csv('../results/plotting_ME_synteny/data/block_files/dummy_features_no_homologues_subset.block', sep = '\\t', index = False, header = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "88590eeb-1458-4986-937f-78061ea5d8e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cat ../results/plotting_ME_synteny/data/block_files/subset_species_no_anchors_only_homologues.block \\\n",
    "../results/plotting_ME_synteny/data/block_files/dummy_features_no_homologues_subset.block > ../results/plotting_ME_synteny/data/block_files/all_species_with_anchors_full_features_subsetted.block "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ffca4d6-6b93-4d18-a855-347b0f13be56",
   "metadata": {},
   "source": [
    "## Layout file employed\n",
    "Layout file specifying all graphical positions for mobile elements is in path *../results/plotting_ME_synteny/data/layout_files/all_species_full_with_anchor_full_features.layout*.\n",
    "___\n",
    "\n",
    "Please find attached below its content"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32507542-2ef4-4f0b-b297-168fe71ee432",
   "metadata": {},
   "source": [
    "```\n",
    "# x,   y, rotation,     ha,     va, color, ratio,            label\n",
    "0.27, 0.9,        0, right, center,     #fc8d62,    1, Halogeometricum borinquense strain wsp4\n",
    "0.35, 0.8,        180, right, center,      #fc8d62,     1, Halobonum sp NJ-3-1\n",
    "0.47, 0.7,        180, right, center,      #fc8d62,    1, ga0207718_100100\n",
    "0.455, 0.6,         180, right,   center,  #fc8d62,   1, jgi12330j12834_1000008\n",
    "0.45, 0.5,        180, right, center,      #fc8d62,    1, Natrinema altunense strain AJ2\n",
    "0.44, 0.4,        0, right, center,      #fc8d62,     1,  Halovivax sp KZCA124\n",
    "0.40, 0.32,        0, right, center,      #fc8d62,     1, Haloterrigena sp SYSU A121-1\n",
    "0.437, 0.24,        180, right, center,      #fc8d62,     1, Haloplanus natans DSM 17983\n",
    "0.475, 0.175,        0, right, center,      #fc8d62,     1, Haloferax sp Q22 \n",
    "0.55, 0.105,        180, right, center,      #fc8d62,     1, ga0207719_100190\n",
    "0.538, 0.05,        180, right, center,      #fc8d62,     1, ga0207733_100382\n",
    "# edges\n",
    "e, 0, 1\n",
    "e, 1, 2\n",
    "e, 2, 3\n",
    "e, 3, 4\n",
    "e, 4, 5\n",
    "e, 5, 6\n",
    "e, 6, 7\n",
    "e, 7, 8\n",
    "e, 8, 9\n",
    "e, 9, 10\n",
    "```\n",
    "\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "862fcc96-d037-423e-b79d-ce0da751e7f2",
   "metadata": {},
   "source": [
    "## Plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "d26f0086-f1c4-47b6-9c47-4629587f1fba",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[19:28:01] DEBUG    Load file `../results/plotting_ME_synteny/data/BE base.py:37\n",
      "                    D/all_species_all_records.bed`                              \n",
      "           DEBUG    Load file `../results/plotting_ME_synteny/data/bl base.py:37\n",
      "                    ock_files/all_species_with_anchors_full_features.           \n",
      "                    block`                                                      \n",
      "           DEBUG    Load file `../results/plotting_ME_synteny/data/la base.py:37\n",
      "                    yout_files/all_species_full_with_anchor_full_feat           \n",
      "                    ures.layout`                                                \n",
      "Column 0: ORF_dummystart_Halogeometricum_borinquense_strain_wsp4 - ORF_dummyend_Halogeometricum_borinquense_strain_wsp4 (CP048739:1-89772)\n",
      "  CP048739 .. 124 (779) features .. +\n",
      "Column 1: ORF_dummystart_Halobonum_sp_NJ-3-1 - ORF_dummyend_Halobonum_sp_NJ-3-1 (CP058579:1-172009)\n",
      "  CP058579 .. 246 (991) features .. +\n",
      "Column 2: ORF_dummystart_ga0207718_100100 - ORF_dummyend_ga0207718_100100 (Ga0207718_100100:1-108523)\n",
      "  Ga0207718_100100 .. 152 (913) features .. -\n",
      "Column 3: ORF_dummystart_jgi12330j12834_1000008 - ORF_dummyend_jgi12330j12834_1000008 (JGI12330J12834_1000008:1-97919)\n",
      "  JGI12330J12834_1000008 .. 146 (906) features .. -\n",
      "Column 4: ORF_dummystart_Natrinema_altunense_strain_AJ2 - ORF_dummyend_Natrinema_altunense_strain_AJ2 (JNCS01000001:1-94988)\n",
      "  JNCS01000001 .. 73 (843) features .. -\n",
      "Column 5: ORF_2831_Halovivax_sp_KZCA124 - ORF_1536_Halovivax_sp_KZCA124 (NZ_CP071597.1:1858-90300)\n",
      "  NZ_CP071597.1 .. 107 (792) features .. -\n",
      "Column 6: ORF_dummystart_Haloterrigena_sp_SYSU_A121-1 - ORF_dummyend_Haloterrigena_sp_SYSU_A121-1 (JABURA010000001:1-98810)\n",
      "  JABURA010000001 .. 112 (859) features .. -\n",
      "Column 7: ORF_dummystart_Haloplanus_natans_DSM_17983 - ORF_dummyend_Haloplanus_natans_DSM_17983 (ATYM01000002:1-113161)\n",
      "  ATYM01000002 .. 90 (845) features .. -\n",
      "Column 8: ORF_dummystart_Haloferax_sp_Q22 - ORF_dummyend_Haloferax_sp_Q22 (LOEP01000012:1-56511)\n",
      "  LOEP01000012 .. 61 (812) features .. +\n",
      "Column 9: ORF_1390_ga0207719_100190 - ORF_dummyend_ga0207719_100190 (Ga0207719_100190:1-48632)\n",
      "  Ga0207719_100190 .. 60 (810) features .. -\n",
      "Column 10: ORF_dummystart_ga0207733_100382 - ORF_dummyend_ga0207733_100382 (Ga0207733_100382:1-30902)\n",
      "  Ga0207733_100382 .. 38 (620) features .. +\n",
      "           DEBUG    A total of 2 orthogroups formed               synteny.py:266\n",
      "[19:28:07] DEBUG    Figure saved to `../results/plotting_ME_synteny/ base.py:295\n",
      "                    data/block_files/all_species_with_anchors_full_f            \n",
      "                    eatures.pdf` (2000px x 1000px)                              \n",
      "[19:28:08] DEBUG    Load file `../results/plotting_ME_synteny/data/BE base.py:37\n",
      "                    D/all_species_all_records.bed`                              \n",
      "           DEBUG    Load file `../results/plotting_ME_synteny/data/bl base.py:37\n",
      "                    ock_files/all_species_with_anchors_full_features.           \n",
      "                    block`                                                      \n",
      "           DEBUG    Load file `../results/plotting_ME_synteny/data/la base.py:37\n",
      "                    yout_files/all_species_full_with_anchor_full_feat           \n",
      "                    ures.layout`                                                \n",
      "Column 0: ORF_dummystart_Halogeometricum_borinquense_strain_wsp4 - ORF_dummyend_Halogeometricum_borinquense_strain_wsp4 (CP048739:1-89772)\n",
      "  CP048739 .. 124 (779) features .. +\n",
      "Column 1: ORF_dummystart_Halobonum_sp_NJ-3-1 - ORF_dummyend_Halobonum_sp_NJ-3-1 (CP058579:1-172009)\n",
      "  CP058579 .. 246 (991) features .. +\n",
      "Column 2: ORF_dummystart_ga0207718_100100 - ORF_dummyend_ga0207718_100100 (Ga0207718_100100:1-108523)\n",
      "  Ga0207718_100100 .. 152 (913) features .. -\n",
      "Column 3: ORF_dummystart_jgi12330j12834_1000008 - ORF_dummyend_jgi12330j12834_1000008 (JGI12330J12834_1000008:1-97919)\n",
      "  JGI12330J12834_1000008 .. 146 (906) features .. -\n",
      "Column 4: ORF_dummystart_Natrinema_altunense_strain_AJ2 - ORF_dummyend_Natrinema_altunense_strain_AJ2 (JNCS01000001:1-94988)\n",
      "  JNCS01000001 .. 73 (843) features .. -\n",
      "Column 5: ORF_2831_Halovivax_sp_KZCA124 - ORF_1536_Halovivax_sp_KZCA124 (NZ_CP071597.1:1858-90300)\n",
      "  NZ_CP071597.1 .. 107 (792) features .. -\n",
      "Column 6: ORF_dummystart_Haloterrigena_sp_SYSU_A121-1 - ORF_dummyend_Haloterrigena_sp_SYSU_A121-1 (JABURA010000001:1-98810)\n",
      "  JABURA010000001 .. 112 (859) features .. -\n",
      "Column 7: ORF_dummystart_Haloplanus_natans_DSM_17983 - ORF_dummyend_Haloplanus_natans_DSM_17983 (ATYM01000002:1-113161)\n",
      "  ATYM01000002 .. 90 (845) features .. -\n",
      "Column 8: ORF_dummystart_Haloferax_sp_Q22 - ORF_dummyend_Haloferax_sp_Q22 (LOEP01000012:1-56511)\n",
      "  LOEP01000012 .. 61 (812) features .. +\n",
      "Column 9: ORF_1390_ga0207719_100190 - ORF_dummyend_ga0207719_100190 (Ga0207719_100190:1-48632)\n",
      "  Ga0207719_100190 .. 60 (810) features .. -\n",
      "Column 10: ORF_dummystart_ga0207733_100382 - ORF_dummyend_ga0207733_100382 (Ga0207733_100382:1-30902)\n",
      "  Ga0207733_100382 .. 38 (620) features .. +\n",
      "           DEBUG    A total of 2 orthogroups formed               synteny.py:266\n",
      "[19:28:13] DEBUG    Figure saved to `../results/plotting_ME_synteny/ base.py:295\n",
      "                    data/block_files/all_species_with_anchors_full_f            \n",
      "                    eatures.svg` (2000px x 1000px)                              \n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "python -m jcvi.graphics.synteny ../results/plotting_ME_synteny/data/block_files/all_species_with_anchors_full_features.block ../results/plotting_ME_synteny/data/BED/all_species_all_records.bed ../results/plotting_ME_synteny/data/layout_files/all_species_full_with_anchor_full_features.layout --glyphcolor=orthogroup --figsize=20x10 --dpi=100\n",
    "mv ../results/plotting_ME_synteny/data/block_files/all_species_with_anchors_full_features.pdf ../results/plotting_ME_synteny/results/all_species_with_anchors_full_features.pdf\n",
    "\n",
    "python -m jcvi.graphics.synteny ../results/plotting_ME_synteny/data/block_files/all_species_with_anchors_full_features.block ../results/plotting_ME_synteny/data/BED/all_species_all_records.bed ../results/plotting_ME_synteny/data/layout_files/all_species_full_with_anchor_full_features.layout --glyphcolor=orthogroup --figsize=20x10 --dpi=100 --format=svg\n",
    "mv ../results/plotting_ME_synteny/data/block_files/all_species_with_anchors_full_features.svg ../results/plotting_ME_synteny/results/all_species_with_anchors_full_features.svg"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
